--- sqlite-3.28.0/src/alter.c	2026-01-09 16:11:01
+++ src/alter.c	2025-09-24 15:21:58
@@ -19,6 +19,7 @@
 ** ALTER TABLE logic from the build.
 */
 #ifndef SQLITE_OMIT_ALTERTABLE
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
 
 /*
 ** Parameter zName is the name of a table that is about to be altered
@@ -473,6 +474,7 @@
   sqlite3SrcListDelete(db, pSrc);
   return;
 }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Parameter pTab is the subject of an ALTER TABLE ... RENAME COLUMN
@@ -507,6 +509,7 @@
 # define isRealTable(x,y) (0)
 #endif
 
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
 /*
 ** Handles the following parser reduction:
 **
@@ -1639,4 +1642,5 @@
   };
   sqlite3InsertBuiltinFuncs(aAlterTableFuncs, ArraySize(aAlterTableFuncs));
 }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif  /* SQLITE_ALTER_TABLE */
