--- sqlite-3.28.0/src/analyze.c	2026-01-09 16:11:01
+++ src/analyze.c	2025-12-03 15:49:33
@@ -142,6 +142,11 @@
 #ifndef SQLITE_OMIT_ANALYZE
 #include "sqliteInt.h"
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <logmsg.h>
+int is_comdb2_index_disableskipscan(const char *);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 #if defined(SQLITE_ENABLE_STAT4)
 # define IsStat4     1
 # define IsStat3     0
@@ -156,7 +161,19 @@
 #endif
 #define IsStat34    (IsStat3+IsStat4)  /* 1 for STAT3 or STAT4. 0 otherwise */
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 /*
+** The number of samples of an index that SQLite takes in order to
+** construct a histogram of the table content when running ANALYZE
+** and with SQLITE_ENABLE_STAT2
+*/
+#define SQLITE_INDEX_SAMPLES 10
+
+static __thread int skip4;
+int64_t analyze_get_nrecs( int iTable );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
+/*
 ** This routine generates code that opens the sqlite_statN tables.
 ** The sqlite_stat1 table is always relevant.  sqlite_stat2 is now
 ** obsolete.  sqlite_stat3 and sqlite_stat4 are only opened when
@@ -207,6 +224,27 @@
   /* Create new statistic tables if they do not exist, or clear them
   ** if they do already exist.
   */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( db->isExpert==0 ){
+    skip4 = 0; /* TODO: Flag on connection? */
+    for(i=0; i<ArraySize(aTable); i++){
+      const char *zTab = aTable[i].zName;
+      Table *pStat;
+      if( (pStat = sqlite3FindTable(db, zTab, NULL))==0 ){
+        if( zTab[11]=='1' || zTab[11]=='3' ){
+          /* do nothing */
+        }else if( zTab[11]=='4' ){
+          skip4 = 1;
+        }
+      }else{
+        aRoot[i] = pStat->tnum;
+        sqlite3VdbeAddOp4Int(v, OP_OpenWrite, iStatCur+i, aRoot[i], iDb, 3);
+        sqlite3VdbeChangeP5(v, 0);
+        VdbeComment((v, zTab));
+      }
+    }
+  }else{ /* db->isExpert==0 */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   for(i=0; i<ArraySize(aTable); i++){
     const char *zTab = aTable[i].zName;
     Table *pStat;
@@ -239,8 +277,12 @@
         sqlite3NestedParse(pParse, "DELETE FROM %Q.%s", pDb->zDbSName, zTab);
 #endif
       }else{
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        sqlite3NestedParse(pParse, "DELETE FROM %Q.%s", pDb->zDbSName, zTab);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         /* The sqlite_stat[134] table already exists.  Delete all rows. */
         sqlite3VdbeAddOp2(v, OP_Clear, aRoot[i], iDb);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       }
     }
   }
@@ -252,6 +294,9 @@
     sqlite3VdbeChangeP5(v, aCreateTbl[i]);
     VdbeComment((v, aTable[i].zName));
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  } /* db->isExpert==0 */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
 /*
@@ -261,6 +306,10 @@
 # define SQLITE_STAT4_SAMPLES 24
 #endif
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+# define PACKEDROWSIZE 1024
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Three SQL functions - stat_init(), stat_push(), and stat_get() -
 ** share an instance of the following structure to hold their state
@@ -273,21 +322,32 @@
   tRowcnt *anDLt;                 /* sqlite_stat4.nDLt */
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
   tRowcnt *anLt;                  /* sqlite_stat4.nLt */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   union {
     i64 iRowid;                     /* Rowid in main table of the key */
     u8 *aRowid;                     /* Key for WITHOUT ROWID tables */
   } u;
   u32 nRowid;                     /* Sizeof aRowid[] */
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   u8 isPSample;                   /* True if a periodic sample */
   int iCol;                       /* If !isPSample, the reason for inclusion */
   u32 iHash;                      /* Tiebreaker hash */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  u8 packedRow[PACKEDROWSIZE];
+  u32 nPackedRow;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
 };                                                    
 struct Stat4Accum {
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  i64 nActualRow;           /* Number of rows in ??? */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   tRowcnt nRow;             /* Number of rows in the entire table */
   tRowcnt nPSample;         /* How often to do a periodic sample */
   int nCol;                 /* Number of columns in index + pk/rowid */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   int nKeyCol;              /* Number of index columns w/o the pk/rowid */
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   int mxSample;             /* Maximum number of samples to accumulate */
   Stat4Sample current;      /* Current row as a Stat4Sample */
   u32 iPrn;                 /* Pseudo-random number used for sampling */
@@ -304,17 +364,30 @@
 */
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
 static void sampleClear(sqlite3 *db, Stat4Sample *p){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  p->nPackedRow = 0;
+  memset(p->packedRow, 0, sizeof(p->packedRow));
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( db!=0 );
   if( p->nRowid ){
     sqlite3DbFree(db, p->u.aRowid);
     p->nRowid = 0;
   }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 #endif
 
 /* Initialize the BLOB value of a ROWID
 */
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+static void sampleSetPackedRow(sqlite3 *db, Stat4Sample *p, int n, const u8 *pData){
+  assert( db!=0 );
+  assert( n<PACKEDROWSIZE );
+  p->nPackedRow = n;
+  memcpy(p->packedRow, pData, n);
+}
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 static void sampleSetRowid(sqlite3 *db, Stat4Sample *p, int n, const u8 *pData){
   assert( db!=0 );
   if( p->nRowid ) sqlite3DbFree(db, p->u.aRowid);
@@ -326,17 +399,20 @@
     p->nRowid = 0;
   }
 }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
 
 /* Initialize the INTEGER value of a ROWID.
 */
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
 static void sampleSetRowidInt64(sqlite3 *db, Stat4Sample *p, i64 iRowid){
   assert( db!=0 );
   if( p->nRowid ) sqlite3DbFree(db, p->u.aRowid);
   p->nRowid = 0;
   p->u.iRowid = iRowid;
 }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
 
 
@@ -351,11 +427,15 @@
   memcpy(pTo->anEq, pFrom->anEq, sizeof(tRowcnt)*p->nCol);
   memcpy(pTo->anLt, pFrom->anLt, sizeof(tRowcnt)*p->nCol);
   memcpy(pTo->anDLt, pFrom->anDLt, sizeof(tRowcnt)*p->nCol);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  sampleSetPackedRow(p->db, pTo, pFrom->nPackedRow, pFrom->packedRow);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( pFrom->nRowid ){
     sampleSetRowid(p->db, pTo, pFrom->nRowid, pFrom->u.aRowid);
   }else{
     sampleSetRowidInt64(p->db, pTo, pFrom->u.iRowid);
   }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 #endif
 
@@ -373,6 +453,21 @@
   sqlite3DbFree(p->db, p);
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+static inline int numRowsToNumSamplesEst(u64 n)
+{
+  int s;
+  if     (n <     10000) s =    24;
+  else if(n <     50000) s =    32;
+  else if(n <    100000) s =    48;
+  else if(n <   1000000) s =  2*32;
+  else if(n <  10000000) s =  4*32;
+  else if(n < 100000000) s =  8*32;
+  else                   s = 16*32;
+  return s;
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Implementation of the stat_init(N,K,C) SQL function. The three parameters
 ** are:
@@ -403,22 +498,39 @@
 ){
   Stat4Accum *p;
   int nCol;                       /* Number of columns in index being sampled */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   int nKeyCol;                    /* Number of key columns */
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   int nColUp;                     /* nCol rounded up for alignment */
   int n;                          /* Bytes of space to allocate */
   sqlite3 *db;                    /* Database connection */
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
   int mxSample = SQLITE_STAT4_SAMPLES;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  u64 nRows = sqlite3_value_int64(argv[1]);
+  if( sqlite3_gbl_tunables.stat4_samples_multiplier > 0 ){
+      mxSample = numRowsToNumSamplesEst(nRows) * sqlite3_gbl_tunables.stat4_samples_multiplier;
+  }
+  if( sqlite3_gbl_tunables.stat4_extra_samples > 0 ){
+      mxSample += sqlite3_gbl_tunables.stat4_extra_samples;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
 
   /* Decode the three function arguments */
   UNUSED_PARAMETER(argc);
   nCol = sqlite3_value_int(argv[0]);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( nCol>1 ); /* >1 because it includes the rowid column */
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( nCol>0 );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   nColUp = sizeof(tRowcnt)<8 ? (nCol+1)&~1 : nCol;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   nKeyCol = sqlite3_value_int(argv[1]);
   assert( nKeyCol<=nCol );
   assert( nKeyCol>0 );
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   /* Allocate the space required for the Stat4Accum object */
   n = sizeof(*p) 
@@ -439,8 +551,13 @@
 
   p->db = db;
   p->nRow = 0;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  p->nActualRow = sqlite3_value_int64(argv[2]);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   p->nCol = nCol;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   p->nKeyCol = nKeyCol;
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   p->current.anDLt = (tRowcnt*)&p[1];
   p->current.anEq = &p->current.anDLt[nColUp];
 
@@ -451,9 +568,17 @@
 
     p->iGet = -1;
     p->mxSample = mxSample;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    p->nPSample = (tRowcnt)(nRows/(mxSample/3+1) + 1);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     p->nPSample = (tRowcnt)(sqlite3_value_int64(argv[2])/(mxSample/3+1) + 1);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     p->current.anLt = &p->current.anEq[nColUp];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    p->iPrn = 0x689e962d*(u32)nCol ^ 0xd0944565*(u32)sqlite3_value_int(argv[1]);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     p->iPrn = 0x689e962d*(u32)nCol ^ 0xd0944565*(u32)sqlite3_value_int(argv[2]);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   
     /* Set up the Stat4Accum.a[] and aBest[] arrays */
     p->a = (struct Stat4Sample*)&p->current.anLt[nColUp];
@@ -604,7 +729,11 @@
     sampleClear(p->db, pMin);
     memmove(pMin, &pMin[1], sizeof(p->a[0])*(p->nSample-p->iMin-1));
     pSample = &p->a[p->nSample-1];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    pSample->nPackedRow = 0;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     pSample->nRowid = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     pSample->anEq = anEq;
     pSample->anDLt = anDLt;
     pSample->anLt = anLt;
@@ -638,6 +767,19 @@
         iMin = i;
       }
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( iMin==-1 ){
+      /* Number of records has increased N times (N > ~200%) since statInit(),
+         however we don't know what N is. Assume that N is 2. It takes log3(N)
+         iterations to get the right sample frequency. Note that This approach
+         will make the sampling distribution more left skewed. */
+      p->nPSample *= 3;
+      for(i=0; i<p->mxSample; i+=3){
+        p->a[i].isPSample = 0;
+      }
+      goto find_new_min;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( iMin>=0 );
     p->iMin = iMin;
   }
@@ -739,7 +881,11 @@
 
   UNUSED_PARAMETER( argc );
   UNUSED_PARAMETER( context );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( p->nCol>1 );        /* Includes rowid field */
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( p->nCol>0 );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( iChng<p->nCol );
 
   if( p->nRow==0 ){
@@ -764,12 +910,17 @@
   }
   p->nRow++;
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  sampleSetPackedRow(p->db, &p->current, sqlite3_value_bytes(argv[2]),
+                                         sqlite3_value_blob(argv[2]));
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( sqlite3_value_type(argv[2])==SQLITE_INTEGER ){
     sampleSetRowidInt64(p->db, &p->current, sqlite3_value_int64(argv[2]));
   }else{
     sampleSetRowid(p->db, &p->current, sqlite3_value_bytes(argv[2]),
                                        sqlite3_value_blob(argv[2]));
   }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   p->current.iHash = p->iPrn = p->iPrn*1103515245 + 12345;
 #endif
 
@@ -807,11 +958,19 @@
   {0}
 };
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 #define STAT_GET_STAT1 0          /* "stat" column of stat1 table */
+#define STAT_GET_NEQ   1          /* "neq" column of stat[34] entry */
+#define STAT_GET_NLT   2          /* "nlt" column of stat[34] entry */
+#define STAT_GET_NDLT  3          /* "ndlt" column of stat[34] entry */
+#define STAT_GET_ROW   4
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+#define STAT_GET_STAT1 0          /* "stat" column of stat1 table */
 #define STAT_GET_ROWID 1          /* "rowid" column of stat[34] entry */
 #define STAT_GET_NEQ   2          /* "neq" column of stat[34] entry */
 #define STAT_GET_NLT   3          /* "nlt" column of stat[34] entry */
 #define STAT_GET_NDLT  4          /* "ndlt" column of stat[34] entry */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Implementation of the stat_get(P,J) SQL function.  This routine is
@@ -842,10 +1001,17 @@
   /* STAT3 and STAT4 have a parameter on this routine. */
   int eCall = sqlite3_value_int(argv[1]);
   assert( argc==2 );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( eCall==STAT_GET_STAT1 || eCall==STAT_GET_NEQ
+       || eCall==STAT_GET_ROW   || eCall==STAT_GET_NLT
+       || eCall==STAT_GET_NDLT
+  );
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( eCall==STAT_GET_STAT1 || eCall==STAT_GET_NEQ 
        || eCall==STAT_GET_ROWID || eCall==STAT_GET_NLT
        || eCall==STAT_GET_NDLT 
   );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( eCall==STAT_GET_STAT1 )
 #else
   assert( argc==1 );
@@ -875,12 +1041,30 @@
     char *z;
     int i;
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    char *zRet = sqlite3MallocZero( p->nCol*25 );
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     char *zRet = sqlite3MallocZero( (p->nKeyCol+1)*25 );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( zRet==0 ){
       sqlite3_result_error_nomem(context);
       return;
     }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    {
+      u64 nRow = p->nActualRow > 0 ? p->nActualRow : p->nRow;
+      /* Never let the estimated number of rows be less than 1 */
+      sqlite3_snprintf(24, zRet, "%llu", MAX(nRow, 1));
+      z = zRet + sqlite3Strlen30(zRet);
+      for(i=0; i<(p->nCol-1); i++){
+        u64 nDistinct = p->current.anDLt[i] + 1;
+        u64 iVal = (nRow + nDistinct - 1) / nDistinct;
+        sqlite3_snprintf(24, z, " %llu", iVal);
+        z += sqlite3Strlen30(z);
+      }
+    }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3_snprintf(24, zRet, "%llu", (u64)p->nRow);
     z = zRet + sqlite3Strlen30(zRet);
     for(i=0; i<p->nKeyCol; i++){
@@ -890,11 +1074,66 @@
       z += sqlite3Strlen30(z);
       assert( p->current.anEq[i] );
     }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( z[0]=='\0' && z>zRet );
 
     sqlite3_result_text(context, zRet, -1, sqlite3_free);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    return; /* end of sqlite_stat1 */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  {
+    Stat4Sample *pS = NULL;
+    tRowcnt *aCnt = NULL;
+    int scale = 1;
+    int iGet;
+    if( p->iGet<0 ){
+      samplePushPrevious(p, 0);
+      p->iGet = 0;
+    }
+    iGet = p->iGet;
+    if( iGet>=p->nSample ){
+      sqlite3_result_null(context);
+      return;
+    }
+    switch( eCall ){
+      case STAT_GET_NEQ:  aCnt = p->a[iGet].anEq; break;
+      case STAT_GET_NLT:  aCnt = p->a[iGet].anLt; break;
+      case STAT_GET_NDLT: aCnt = p->a[iGet].anDLt; break;
+      case STAT_GET_ROW:
+        pS = p->a + iGet;
+        sqlite3_result_blob(context, pS->packedRow, pS->nPackedRow,
+                            SQLITE_TRANSIENT);
+        ++p->iGet;
+        break;
+    }
+    if( p->nActualRow>0 ){
+      scale = (int)((p->nActualRow / (double)p->nRow) + 0.5);
+    }
+    if( eCall!=STAT_GET_ROW ){
+      char *zRet = sqlite3MallocZero( p->nCol*25 );
+      if( zRet==0 ){
+        sqlite3_result_error_nomem(context);
+      }else{
+        int i;
+        char *z = zRet;
+        for(i=0; i<p->nCol; i++){
+          if( aCnt[i] > 2 ){
+            sqlite3_snprintf(24, z, "%llu ", (u64)aCnt[i] * scale);
+          }else{
+            sqlite3_snprintf(24, z, "%llu ", (u64)aCnt[i]);
+          }
+          z += sqlite3Strlen30(z);
+        }
+        assert( z[0]=='\0' && z>zRet );
+        z[-1] = '\0';
+        sqlite3_result_text(context, zRet, -1, sqlite3_free);
+      }
+    }
+  }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   else if( eCall==STAT_GET_ROWID ){
     if( p->iGet<0 ){
       samplePushPrevious(p, 0);
@@ -942,6 +1181,7 @@
       }
     }
   }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif /* SQLITE_ENABLE_STAT3_OR_STAT4 */
 #ifndef SQLITE_DEBUG
   UNUSED_PARAMETER( argc );
@@ -991,14 +1231,20 @@
   int iTabCur;                 /* Table cursor */
   Vdbe *v;                     /* The virtual machine being built up */
   int i;                       /* Loop counter */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   int jZeroRows = -1;          /* Jump from here if number of rows is zero */
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   int iDb;                     /* Index of database containing pTab */
   u8 needTableCnt = 1;         /* True to count the table */
   int regNewRowid = iMem++;    /* Rowid for the inserted record */
   int regStat4 = iMem++;       /* Register to hold Stat4Accum object */
   int regChng = iMem++;        /* Index of changed index field */
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  int regSampleRow = iMem++;   /* Register to hold sample row */
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   int regRowid = iMem++;       /* Rowid argument passed to stat_push() */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
   int regTemp = iMem++;        /* Temporary use register */
   int regTabname = iMem++;     /* Register containing table name */
@@ -1025,7 +1271,9 @@
   assert( sqlite3BtreeHoldsAllMutexes(db) );
   iDb = sqlite3SchemaToIndex(db, pTab->pSchema);
   assert( iDb>=0 );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
   assert( sqlite3SchemaMutexHeld(db, iDb, 0) );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifndef SQLITE_OMIT_AUTHORIZATION
   if( sqlite3AuthCheck(pParse, SQLITE_ANALYZE, pTab->zName, 0,
       db->aDb[iDb].zDbSName ) ){
@@ -1065,7 +1313,22 @@
 
     if( pOnlyIdx && pOnlyIdx!=pIdx ) continue;
     if( pIdx->pPartIdxWhere==0 ) needTableCnt = 0;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    nCol = pIdx->nKeyCol;
+    for(i=0; i < nCol; ++i){
+      if( strcmp(pIdx->azColl[i], "DATACOPY")==0 ){
+        nCol = i;
+        break;
+      }
+    }
+    nColTest = nCol;
     if( !HasRowid(pTab) && IsPrimaryKeyIndex(pIdx) ){
+      zIdxName = pTab->zName;
+    }else{
+      zIdxName = pIdx->zName;
+    }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+    if( !HasRowid(pTab) && IsPrimaryKeyIndex(pIdx) ){
       nCol = pIdx->nKeyCol;
       zIdxName = pTab->zName;
       nColTest = nCol - 1;
@@ -1074,6 +1337,7 @@
       zIdxName = pIdx->zName;
       nColTest = pIdx->uniqNotNull ? pIdx->nKeyCol-1 : nCol-1;
     }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
     /* Populate the register containing the index name. */
     sqlite3VdbeLoadString(v, regIdxname, zIdxName);
@@ -1133,11 +1397,27 @@
     **
     ** The third argument is only used for STAT3 and STAT4
     */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
+    /* we only care about our shared tables */
+    if( iDb==0 ){
+      i64 actualCount = analyze_get_nrecs(pIdx->tnum);
+      sqlite3VdbeAddOp4Dup8(v, OP_Int64, 0, regStat4+3, 0,
+                            (const u8*)&actualCount, P4_INT64);
+    }else{
+      /* TODO: Is there a better default value here? */
+      sqlite3VdbeAddOp2(v, OP_Integer, 0, regStat4+3);
+    }
+#endif
+    sqlite3VdbeAddOp2(v, OP_Integer, nCol+1, regStat4+1);
+    sqlite3VdbeAddOp2(v, OP_Count, iIdxCur, regStat4+2);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+#ifdef SQLITE_ENABLE_STAT3_OR_STAT4
     sqlite3VdbeAddOp2(v, OP_Count, iIdxCur, regStat4+3);
 #endif
     sqlite3VdbeAddOp2(v, OP_Integer, nCol, regStat4+1);
     sqlite3VdbeAddOp2(v, OP_Integer, pIdx->nKeyCol, regStat4+2);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3VdbeAddOp4(v, OP_Function0, 0, regStat4+1, regStat4,
                      (char*)&statInitFuncdef, P4_FUNCDEF);
     sqlite3VdbeChangeP5(v, 2+IsStat34);
@@ -1216,6 +1496,7 @@
     **   Next csr
     **   if !eof(csr) goto next_row;
     */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
     assert( regRowid==(regStat4+2) );
     if( HasRowid(pTab) ){
@@ -1234,13 +1515,25 @@
       sqlite3ReleaseTempRange(pParse, regKey, pPk->nKeyCol);
     }
 #endif
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( regChng==(regStat4+1) );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#ifdef SQLITE_ENABLE_STAT3_OR_STAT4
+    assert( regSampleRow==(regStat4+2) );
+    sqlite3VdbeAddOp3(v, OP_MakeRecord, regPrev, nCol, regSampleRow);
+#endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3VdbeAddOp4(v, OP_Function0, 1, regStat4, regTemp,
                      (char*)&statPushFuncdef, P4_FUNCDEF);
     sqlite3VdbeChangeP5(v, 2+IsStat34);
     sqlite3VdbeAddOp2(v, OP_Next, iIdxCur, addrNextRow); VdbeCoverage(v);
 
     /* Add the entry to the stat1 table. */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( sqlite3_gbl_tunables.analyze_empty_tables ){
+      sqlite3VdbeJumpHere(v, addrRewind);
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     callStatGet(v, regStat4, STAT_GET_STAT1, regStat1);
     assert( "BBB"[0]==SQLITE_AFF_TEXT );
     sqlite3VdbeAddOp4(v, OP_MakeRecord, regTabname, 3, regTemp, "BBB", 0);
@@ -1253,26 +1546,50 @@
 
     /* Add the entries to the stat3 or stat4 table. */
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( skip4==0 ){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     {
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       int regEq = regStat1;
       int regLt = regStat1+1;
       int regDLt = regStat1+2;
       int regSample = regStat1+3;
       int regCol = regStat1+4;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
       int regSampleRowid = regCol + nCol;
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
       int addrNext;
       int addrIsNull;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
       u8 seekOp = HasRowid(pTab) ? OP_NotExists : OP_NotFound;
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 
       pParse->nMem = MAX(pParse->nMem, regCol+nCol);
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( sqlite3_gbl_tunables.analyze_empty_tables ){
+        addrRewind = sqlite3VdbeAddOp1(v, OP_Rewind, iIdxCur);
+        VdbeCoverage(v);
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
       addrNext = sqlite3VdbeCurrentAddr(v);
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
       callStatGet(v, regStat4, STAT_GET_ROWID, regSampleRowid);
       addrIsNull = sqlite3VdbeAddOp1(v, OP_IsNull, regSampleRowid);
       VdbeCoverage(v);
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
       callStatGet(v, regStat4, STAT_GET_NEQ, regEq);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      addrIsNull = sqlite3VdbeAddOp1(v, OP_IsNull, regEq);
+      VdbeCoverage(v);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       callStatGet(v, regStat4, STAT_GET_NLT, regLt);
       callStatGet(v, regStat4, STAT_GET_NDLT, regDLt);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      callStatGet(v, regStat4, STAT_GET_ROW, regSample);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3VdbeAddOp4Int(v, seekOp, iTabCur, addrNext, regSampleRowid, 0);
       VdbeCoverage(v);
 #ifdef SQLITE_ENABLE_STAT3
@@ -1283,19 +1600,32 @@
       }
       sqlite3VdbeAddOp3(v, OP_MakeRecord, regCol, nCol, regSample);
 #endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3VdbeAddOp3(v, OP_MakeRecord, regTabname, 6, regTemp);
       sqlite3VdbeAddOp2(v, OP_NewRowid, iStatCur+1, regNewRowid);
       sqlite3VdbeAddOp3(v, OP_Insert, iStatCur+1, regTemp, regNewRowid);
       sqlite3VdbeAddOp2(v, OP_Goto, 1, addrNext); /* P1==1 for end-of-loop */
       sqlite3VdbeJumpHere(v, addrIsNull);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( sqlite3_gbl_tunables.analyze_empty_tables ){
+        sqlite3VdbeJumpHere(v, addrRewind);
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
 #endif /* SQLITE_ENABLE_STAT3_OR_STAT4 */
 
     /* End of analysis */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( !sqlite3_gbl_tunables.analyze_empty_tables ){
+      sqlite3VdbeJumpHere(v, addrRewind);
+    }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3VdbeJumpHere(v, addrRewind);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
 
 
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   /* Create a single sqlite_stat1 entry containing NULL as the index
   ** name and the row count as the content.
   */
@@ -1314,6 +1644,9 @@
 #endif
     sqlite3VdbeJumpHere(v, jZeroRows);
   }
+#else /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
+  UNUSED_PARAMETER(needTableCnt);
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
 
@@ -1374,7 +1707,9 @@
     openStatTable(pParse, iDb, iStatCur, pTab->zName, "tbl");
   }
   analyzeOneTable(pParse, pTab, pOnlyIdx, iStatCur,pParse->nMem+1,pParse->nTab);
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   loadAnalysis(pParse, iDb);
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
 /*
@@ -1397,7 +1732,9 @@
   Table *pTab;
   Index *pIdx;
   Token *pTableName;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   Vdbe *v;
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   /* Read the database schema. If an error occurs, leave an error message
   ** and code in pParse and return NULL. */
@@ -1432,9 +1769,11 @@
       }
     }
   }
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   if( db->nSqlExec==0 && (v = sqlite3GetVdbe(pParse))!=0 ){
     sqlite3VdbeAddOp0(v, OP_Expire);
   }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
 /*
@@ -1477,13 +1816,20 @@
     }
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
     if( aOut ) aOut[i] = v;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
     if( aLog ) aLog[i] = sqlite3LogEst(v);
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 #else
     assert( aOut==0 );
     UNUSED_PARAMETER(aOut);
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
     assert( aLog!=0 );
     aLog[i] = sqlite3LogEst(v);
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( aLog ) aLog[i] = sqlite3LogEst(v);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( *z==' ' ) z++;
   }
 #ifndef SQLITE_ENABLE_STAT3_OR_STAT4
@@ -1493,6 +1839,18 @@
 #endif
     pIndex->bUnordered = 0;
     pIndex->noSkipScan = 0;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /* assign noSkipScan, only if not foreign table */ 
+    if( pIndex->pTable ) {
+      if( pIndex->pTable->iDb==0 ){
+        pIndex->noSkipScan = is_comdb2_index_disableskipscan(
+                pIndex->pTable->zName
+                );
+      } else if( pIndex->pTable->iDb > 1 ){ /* foreign */
+        pIndex->noSkipScan = 1;
+      }
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     while( z[0] ){
       if( sqlite3_strglob("unordered*", z)==0 ){
         pIndex->bUnordered = 1;
@@ -1535,7 +1893,11 @@
   if( argv==0 || argv[0]==0 || argv[2]==0 ){
     return 0;
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  pTable = sqlite3FindTableCheckOnly(pInfo->db, argv[0], pInfo->zDatabase);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pTable = sqlite3FindTable(pInfo->db, argv[0], pInfo->zDatabase);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( pTable==0 ){
     return 0;
   }
@@ -1568,6 +1930,13 @@
       pTable->nRowLogEst = pIndex->aiRowLogEst[0];
       pTable->tabFlags |= TF_HasStat1;
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /*
+   * stale entries in sqlite_stat1 will have pIndex null
+   * and result in a potential bad estimate for table
+   * if stale index entry was processed last
+   */
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }else{
     Index fakeIdx;
     fakeIdx.szIdxRow = pTable->szTabRow;
@@ -1576,6 +1945,7 @@
 #endif
     decodeIntArray((char*)z, 1, 0, &pTable->nRowLogEst, &fakeIdx);
     pTable->szTabRow = fakeIdx.szIdxRow;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     pTable->tabFlags |= TF_HasStat1;
   }
 
@@ -1590,11 +1960,28 @@
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
   if( pIdx->aSample ){
     int j;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    for(j=0; j<pIdx->nAlloc; j++){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     for(j=0; j<pIdx->nSample; j++){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       IndexSample *p = &pIdx->aSample[j];
       sqlite3DbFree(db, p->p);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      sqlite3DbFree(db, p->anEq); p->anEq = 0;
+      sqlite3DbFree(db, p->anLt); p->anLt = 0;
+      sqlite3DbFree(db, p->anDLt); p->anDLt = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    sqlite3DbFree(db, pIdx->aAvgEq); pIdx->aAvgEq = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3DbFree(db, pIdx->aSample);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    pIdx->nAlloc = 0;
+    pIdx->nSample = 0;
+    pIdx->aSample = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   if( db && db->pnBytesFreed==0 ){
     pIdx->nSample = 0;
@@ -1614,15 +2001,25 @@
 static void initAvgEq(Index *pIdx){
   if( pIdx ){
     IndexSample *aSample = pIdx->aSample;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( aSample==0 ) return;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     IndexSample *pFinal = &aSample[pIdx->nSample-1];
     int iCol;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    int nCol = pIdx->nSampleCol;
+    {
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     int nCol = 1;
     if( pIdx->nSampleCol>1 ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       /* If this is stat4 data, then calculate aAvgEq[] values for all
       ** sample columns except the last. The last is always set to 1, as
       ** once the trailing PK fields are considered all index keys are
       ** unique.  */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
       nCol = pIdx->nSampleCol-1;
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
       pIdx->aAvgEq[nCol] = 1;
     }
     for(iCol=0; iCol<nCol; iCol++){
@@ -1677,12 +2074,17 @@
 ){
   Index *pIdx = sqlite3FindIndex(db, zName, zDb);
   if( pIdx==0 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    Table *pTab = sqlite3FindTableCheckOnly(db, zName, zDb);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     Table *pTab = sqlite3FindTable(db, zName, zDb);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( pTab && !HasRowid(pTab) ) pIdx = sqlite3PrimaryKeyIndex(pTab);
   }
   return pIdx;
 }
 
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
 /*
 ** Load the content from either the sqlite_stat4 or sqlite_stat3 table 
 ** into the relevant Index.aSample[] arrays.
@@ -1756,6 +2158,7 @@
     }
     pSpace = (tRowcnt*)&pIdx->aSample[nSample];
     pIdx->aAvgEq = pSpace; pSpace += nIdxCol;
+    pIdx->pTable->tabFlags |= TF_HasStat4;
     for(i=0; i<nSample; i++){
       pIdx->aSample[i].anEq = pSpace; pSpace += nIdxCol;
       pIdx->aSample[i].anLt = pSpace; pSpace += nIdxCol;
@@ -1817,12 +2220,174 @@
   if( rc==SQLITE_OK ) initAvgEq(pPrevIdx);
   return rc;
 }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <vdbecompare.c>
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Load content from the sqlite_stat4 and sqlite_stat3 tables into 
 ** the Index.aSample[] arrays of all indices.
 */
 static int loadStat4(sqlite3 *db, const char *zDb){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  int rc;                   /* Result codes from subroutines */
+  sqlite3_stmt *pStmt = 0;  /* An SQL statement being run */
+  char *zSql;
+  const char *zSql2 = "SELECT idx,neq,nlt,ndlt,sample,tbl FROM %Q.sqlite_stat4"
+                      " WHERE tbl not like 'cdb2.%%.sav';";
+
+  if( sqlite3FindTableCheckOnly(db, "sqlite_stat4", zDb)==0 ){
+    return SQLITE_OK;
+  }
+  zSql = sqlite3MPrintf(db, zSql2, zDb);
+  if( !zSql ) return SQLITE_NOMEM_BKPT;
+  rc = sqlite3_prepare(db, zSql, -1, &pStmt, 0);
+  sqlite3DbFree(db, zSql);
+  if( rc!=SQLITE_OK ) return rc;
+
+  while( sqlite3_step(pStmt)==SQLITE_ROW ){
+    char *zTable;
+    char *zIndex;
+    Index *pIdx;
+    int nCol, iPos, nAlloc;
+    IndexSample *pSample;
+
+    zTable = (char *)sqlite3_column_text(pStmt, 5);
+    if( strncmp("cdb2.", zTable, 5)==0 ) continue;
+    zIndex = (char *)sqlite3_column_text(pStmt, 0);
+    if( zIndex==0 ) continue;
+    pIdx = findIndexOrPrimaryKey(db, zIndex, zDb);
+    if( pIdx==0 ) continue;
+    if( pIdx->pKeyInfo==0 ){
+      Parse parse;
+      int j;
+      memset(&parse, 0, sizeof(Parse));
+      parse.db = db;
+      pIdx->pKeyInfo = sqlite3KeyInfoOfIndex(&parse, pIdx);
+      if( pIdx->pKeyInfo==0 ) continue;
+      for(j=0; j<pIdx->nKeyCol; j++){
+        if( pIdx->pKeyInfo->aColl[j] ) break; /* DATACOPY */
+      }
+      pIdx->nSampleCol = j;
+    }
+    nCol = pIdx->nSampleCol;
+    nAlloc = pIdx->nAlloc;
+    if( pIdx->nSample>=nAlloc ){
+      int i, bOom;
+      int nSize = (nCol + 1) * sizeof(tRowcnt);
+      if( nAlloc==0 ){
+        nAlloc = MAX(pIdx->nSample, SQLITE_STAT4_SAMPLES);
+        assert( pIdx->aAvgEq==0 );
+        pIdx->aAvgEq = sqlite3DbMallocZero(db, nSize);
+        if( pIdx->aAvgEq==0 ){
+          sqlite3_finalize(pStmt);
+          return SQLITE_NOMEM_BKPT;
+        }
+      }else{
+        nAlloc = pIdx->nSample * 2;
+      }
+      assert( nAlloc>=pIdx->nSample );
+      pIdx->aSample = sqlite3DbRealloc(db, pIdx->aSample,
+                                       nAlloc * sizeof(IndexSample));
+      if( pIdx->aSample==0 ){
+        sqlite3_finalize(pStmt);
+        return SQLITE_NOMEM_BKPT;
+      }
+      /* Zero out the new slots */
+      memset(&pIdx->aSample[pIdx->nSample], 0,
+             (nAlloc - pIdx->nSample) * sizeof(IndexSample));
+      /* TODO: Make this whole loop use one big malloc. */
+      bOom = 0;
+      for(i=pIdx->nSample; i<nAlloc; i++){
+        pIdx->aSample[i].anEq  = sqlite3DbMallocZero(db, nSize);
+        if( pIdx->aSample[i].anEq==0 ){ bOom = 1; break; }
+        pIdx->aSample[i].anLt  = sqlite3DbMallocZero(db, nSize);
+        if( pIdx->aSample[i].anLt==0 ){ bOom = 1; break; }
+        pIdx->aSample[i].anDLt = sqlite3DbMallocZero(db, nSize);
+        if( pIdx->aSample[i].anDLt==0 ){ bOom = 1; break; }
+      }
+      if( bOom ){
+        sqlite3_finalize(pStmt);
+        return SQLITE_NOMEM_BKPT;
+      }
+      pIdx->nAlloc = nAlloc;
+    }
+
+    /* Find position for this sample */
+    iPos = pIdx->nSample;
+    if( iPos>0 ){
+      int x, cmp;
+      int iPosGreater = iPos;
+      int iPosSmaller = 0;
+      UnpackedRecord *pUnpacked = sqlite3VdbeAllocUnpackedRecord(pIdx->pKeyInfo);
+      if( pUnpacked==0 ){
+        sqlite3_finalize(pStmt);
+        return SQLITE_NOMEM_BKPT;
+      }
+      sqlite3VdbeRecordUnpack(pIdx->pKeyInfo, sqlite3_column_bytes(pStmt, 4),
+                              sqlite3_column_blob(pStmt, 4), pUnpacked);
+      /* find first less than the new item */
+      {
+        /* Most of the time, last slot is the correct position. */
+        x = iPosGreater - 1;
+        cmp = sqlite3VdbeRecordCompare(pIdx->aSample[x].n,
+                                       pIdx->aSample[x].p, pUnpacked);
+        if( cmp<0 ) iPosSmaller = iPosGreater;
+      }
+      /* TODO: What is special about 32 here? */
+      while( iPosGreater-iPosSmaller>32 ){ /* binary search */
+        x = (iPosSmaller + iPosGreater) / 2;
+        cmp = sqlite3VdbeRecordCompare(pIdx->aSample[x].n,
+                                       pIdx->aSample[x].p, pUnpacked);
+        if( cmp<0 ){
+          iPosSmaller = x;
+        }else{
+          iPosGreater = x;
+        }
+      }
+      for(iPos=iPosGreater; iPos>iPosSmaller; iPos--){
+        x = iPos - 1;
+        cmp = sqlite3VdbeRecordCompare(pIdx->aSample[x].n,
+                                       pIdx->aSample[x].p, pUnpacked);
+        if( cmp<0 ) break;
+      }
+      sqlite3DbFree(db, pUnpacked);
+      /* Make room for sample @pos */
+      if( iPos<pIdx->nSample ){
+        int k;
+        IndexSample tmp = pIdx->aSample[pIdx->nSample];
+        for(k=pIdx->nSample-1; k>=iPos; k--){
+          pIdx->aSample[k+1] = pIdx->aSample[k];
+        }
+        pIdx->aSample[iPos] = tmp;
+      }
+    }
+    assert( iPos>=0 && iPos<pIdx->nAlloc );
+    pSample = &pIdx->aSample[iPos];
+    decodeIntArray((char*)sqlite3_column_text(pStmt,1),nCol,pSample->anEq,0,0);
+    decodeIntArray((char*)sqlite3_column_text(pStmt,2),nCol,pSample->anLt,0,0);
+    decodeIntArray((char*)sqlite3_column_text(pStmt,3),nCol,pSample->anDLt,0,0);
+
+    /* Take a copy of the sample. Add two 0x00 bytes the end of the buffer.
+    ** This is in case the sample record is corrupted. In that case, the
+    ** sqlite3VdbeRecordCompare() may read up to two varints past the
+    ** end of the allocated buffer before it realizes it is dealing with
+    ** a corrupt record. Adding the two 0x00 bytes prevents this from causing
+    ** a buffer overread.  */
+    pSample->n = sqlite3_column_bytes(pStmt, 4);
+    pSample->p = sqlite3DbMallocZero(db, pSample->n + 2);
+    if( pSample->p==0 ){
+      sqlite3_finalize(pStmt);
+      return SQLITE_NOMEM_BKPT;
+    }
+    memcpy(pSample->p, sqlite3_column_blob(pStmt, 4), pSample->n);
+    memset((u8*)pSample->p + pSample->n, 0, 2);
+    pIdx->nSample++;
+    pIdx->pTable->tabFlags |= TF_HasStat4;
+  }
+  return sqlite3_finalize(pStmt);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   int rc = SQLITE_OK;             /* Result codes from subroutines */
 
   assert( db->lookaside.bDisable );
@@ -1843,6 +2408,7 @@
   }
 
   return rc;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 #endif /* SQLITE_ENABLE_STAT3_OR_STAT4 */
 
@@ -1884,19 +2450,49 @@
   }
   for(i=sqliteHashFirst(&pSchema->idxHash); i; i=sqliteHashNext(i)){
     Index *pIdx = sqliteHashData(i);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /* if this is a dynamic attach, don't wipe all the data! */
+    if( iDb <= 1
+     || db->init.busy==0
+     || db->init.zTblName == NULL
+     || pIdx->pTable == NULL
+     || pIdx->pTable->zName == NULL
+     || (strlen(db->init.zTblName) == strlen(pIdx->pTable->zName)
+     && strncasecmp(db->init.zTblName,
+              pIdx->pTable->zName, strlen(db->init.zTblName)) == 0)
+    ){
+       pIdx->hasStat1 = 0;
+       sqlite3DefaultRowEst(pIdx);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     pIdx->hasStat1 = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifdef SQLITE_ENABLE_STAT3_OR_STAT4
     sqlite3DeleteIndexSamples(db, pIdx);
     pIdx->aSample = 0;
 #endif
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    }else{
+      /* We are running InitOne on an attached table, but this one is
+      ** a index for an already attached table */
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
 
   /* Load new statistics out of the sqlite_stat1 table */
   sInfo.db = db;
   sInfo.zDatabase = db->aDb[iDb].zDbSName;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( sqlite3FindTableByAnalysisLoad(db, "sqlite_stat1", sInfo.zDatabase,
+              db->aDb[iDb].local, (db->aDb[iDb].class_override) ? db->aDb[iDb].class : -1)!=0 ){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( sqlite3FindTable(db, "sqlite_stat1", sInfo.zDatabase)!=0 ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     zSql = sqlite3MPrintf(db, 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        "SELECT tbl,idx,stat FROM %Q.sqlite_stat1 WHERE tbl not like 'cdb2.%%.sav'", sInfo.zDatabase);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         "SELECT tbl,idx,stat FROM %Q.sqlite_stat1", sInfo.zDatabase);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( zSql==0 ){
       rc = SQLITE_NOMEM_BKPT;
     }else{
@@ -1921,6 +2517,16 @@
   }
   for(i=sqliteHashFirst(&pSchema->idxHash); i; i=sqliteHashNext(i)){
     Index *pIdx = sqliteHashData(i);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    initAvgEq(pIdx);
+    if( pIdx->pKeyInfo ){
+      int nRef = pIdx->pKeyInfo->nRef;
+      sqlite3KeyInfoUnref(pIdx->pKeyInfo);
+      if( nRef==1 ){ /* was just us -- no one else using */
+        pIdx->pKeyInfo = NULL;
+      }
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3_free(pIdx->aiRowEst);
     pIdx->aiRowEst = 0;
   }
