--- sqlite-3.28.0/src/attach.c	2026-01-09 16:11:01
+++ src/attach.c	2025-11-06 15:48:32
@@ -11,8 +11,15 @@
 *************************************************************************
 ** This file contains code used to implement the ATTACH and DETACH commands.
 */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <string.h>
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #include "sqliteInt.h"
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include "logmsg.h"
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 #ifndef SQLITE_OMIT_ATTACH
 /*
 ** Resolve an expression that was part of an ATTACH or DETACH statement. This
@@ -60,16 +67,33 @@
 ** new database, close the database on db->init.iDb and reopen it as an
 ** empty MemDB.
 */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+int comdb2_dynamic_attach(
+  sqlite3 *db,
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 static void attachFunc(
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   sqlite3_context *context,
   int NotUsed,
   sqlite3_value **argv
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  ,const char *zName,
+  const char *zFile,
+  char **pzErrDyn,
+  int version,
+  int class,
+  int local,
+  int class_override,
+  int proto_version
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 ){
   int i;
   int rc = 0;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   sqlite3 *db = sqlite3_context_db_handle(context);
   const char *zName;
   const char *zFile;
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   char *zPath = 0;
   char *zErr = 0;
   unsigned int flags;
@@ -78,11 +102,17 @@
   char *zErrDyn = 0;
   sqlite3_vfs *pVfs;
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  char *dbName;
+  char *tblName;
+  int  iFndDb;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   UNUSED_PARAMETER(NotUsed);
   zFile = (const char *)sqlite3_value_text(argv[0]);
   zName = (const char *)sqlite3_value_text(argv[1]);
   if( zFile==0 ) zFile = "";
   if( zName==0 ) zName = "";
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 #ifdef SQLITE_ENABLE_DESERIALIZE
 # define REOPEN_AS_MEMDB(db)  (db->init.reopenMemdb)
@@ -95,7 +125,17 @@
     ** from sqlite3_deserialize() to close database db->init.iDb and
     ** reopen it as a MemDB */
     pVfs = sqlite3_vfs_find("memdb");
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( pVfs==0 ){
+      zErrDyn = sqlite3MPrintf(db,
+                               "%s: the \"memdb\" VFS was not found\n",
+                               __func__);
+      rc = SQLITE_ERROR;
+      goto attach_error;
+    }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( pVfs==0 ) return;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     pNew = &db->aDb[db->init.iDb];
     if( pNew->pBt ) sqlite3BtreeClose(pNew->pBt);
     pNew->pBt = 0;
@@ -116,25 +156,83 @@
       );
       goto attach_error;
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /* we store one Btree per foreign db, therefore we need to extract
+    ** the dbname from table */
+    dbName = sqlite3DbStrDup(db, zName);
+    if( dbName==0 ){
+      rc = SQLITE_NOMEM_BKPT;
+      goto attach_error;
+    }else{
+      /* to keep table names unique, zName includes the dbname as well.
+      ** But the Btree will be named after dbname, since they will collect all
+      ** the tables for this db */
+      char *ptr = strchr(dbName, '.');
+      if( ptr ){
+        *ptr = '\0';
+        tblName = (ptr+1);
+      }else{
+        logmsg(LOGMSG_ERROR, "%s: remote dbname URI incomplete \"%s\"?\n",
+               __func__, dbName);
+        zErrDyn = sqlite3MPrintf(db,
+                                 "%s: remote dbname URI incomplete \"%s\"?\n",
+                                 __func__, dbName);
+        sqlite3DbFree(db, dbName);
+
+        rc = SQLITE_ERROR;
+        goto attach_error;
+      }
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     for(i=0; i<db->nDb; i++){
       char *z = db->aDb[i].zDbSName;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      assert( z && dbName);
+      if( sqlite3StrICmp(z, dbName)==0 ){
+        break;
+      }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       assert( z && zName );
       if( sqlite3StrICmp(z, zName)==0 ){
         zErrDyn = sqlite3MPrintf(db, "database %s is already in use", zName);
         goto attach_error;
       }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
   
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    iFndDb = i;
+    if( i!=db->nDb ) {
+      extern int sqlite3BtreeReopen(const char*,Btree*);
+
+      rc = sqlite3BtreeReopen(dbName, db->aDb[iFndDb].pBt);
+
+      db->aDb[iFndDb].zDbSName = dbName;
+      db->aDb[iFndDb].class = class;
+      db->aDb[iFndDb].class_override = class_override;
+      db->aDb[iFndDb].local = local;
+      db->aDb[iFndDb].version = proto_version;
+      goto done_with_open;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     /* Allocate the new entry in the db->aDb[] array and initialize the schema
     ** hash tables.
     */
     if( db->aDb==db->aDbStatic ){
       aNew = sqlite3DbMallocRawNN(db, sizeof(db->aDb[0])*3 );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( aNew==0 ) return -1;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       if( aNew==0 ) return;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       memcpy(aNew, db->aDb, sizeof(db->aDb[0])*2);
     }else{
       aNew = sqlite3DbRealloc(db, db->aDb, sizeof(db->aDb[0])*(db->nDb+1) );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( aNew==0 ) return -1;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       if( aNew==0 ) return;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
     db->aDb = aNew;
     pNew = &db->aDb[db->nDb];
@@ -148,15 +246,30 @@
     rc = sqlite3ParseUri(db->pVfs->zName, zFile, &flags, &pVfs, &zPath, &zErr);
     if( rc!=SQLITE_OK ){
       if( rc==SQLITE_NOMEM ) sqlite3OomFault(db);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( context )
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3_result_error(context, zErr, -1);
       sqlite3_free(zErr);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      return -1;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       return;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
     assert( pVfs );
     flags |= SQLITE_OPEN_MAIN_DB;
     rc = sqlite3BtreeOpen(pVfs, zPath, db, &pNew->pBt, 0, flags);
     db->nDb++;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    pNew->zDbSName = dbName;
+    pNew->class = class;
+    pNew->class_override = class_override;
+    pNew->local = local;
+    pNew->version = proto_version;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     pNew->zDbSName = sqlite3DbStrDup(db, zName);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   db->noSharedCache = 0;
   if( rc==SQLITE_CONSTRAINT ){
@@ -223,6 +336,9 @@
     }
   }
 #endif
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+done_with_open:
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   sqlite3_free( zPath );
 
   /* If the file was opened successfully, read the schema for the new database.
@@ -235,7 +351,40 @@
     db->init.iDb = 0;
     db->mDbFlags &= ~(DBFLAG_SchemaKnownOk);
     if( !REOPEN_AS_MEMDB(db) ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      Table *p;
+      int savedBusy = db->init.busy;
+
+      db->init.busy = 0; /* TODO: prevent assert (?) */
+      char *zTmp = (char*)zName;
+      /* we need to take care of override and local */
+      if (local) {
+        zTmp = sqlite3_mprintf("LOCAL_%s", zName);  
+      } else if (class_override) {
+        extern const char *mach_class_class2tier(int value);
+        zTmp = sqlite3_mprintf("%s_%s", mach_class_class2tier(class), zName);
+      }
+      rc = sqlite3InitTable(db, &zErrDyn, zTmp);
+      db->init.busy = savedBusy;
+      if (zTmp != zName)
+        sqlite3DbFree(db, zTmp);
+
+      /*
+      ** Need to set the version to the table to support per table schema
+      ** refresh
+      */
+      p = sqlite3HashFind(&db->aDb[iFndDb].pSchema->tblHash, tblName);
+      if( !p ){
+        logmsg(LOGMSG_ERROR, "%s: failed to find table \"%s\" after init\n",
+               __func__, tblName);
+        rc = SQLITE_ERROR;
+      }else{
+        p->version = version;
+        p->iDb = iFndDb;
+      }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       rc = sqlite3Init(db, &zErrDyn);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
     sqlite3BtreeLeaveAll(db);
     assert( zErrDyn==0 || rc!=SQLITE_OK );
@@ -271,17 +420,64 @@
     goto attach_error;
   }
   
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  return 0;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   return;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 attach_error:
   /* Return an error if we get here */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  *pzErrDyn = zErrDyn;
+  return rc;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( zErrDyn ){
     sqlite3_result_error(context, zErrDyn, -1);
     sqlite3DbFree(db, zErrDyn);
   }
   if( rc ) sqlite3_result_error_code(context, rc);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+static void attachFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **argv
+){
+  sqlite3 *db = sqlite3_context_db_handle(context);
+  const char *zName;
+  const char *zFile;
+  char *zErrDyn;
+  int rc = 0;
+
+  UNUSED_PARAMETER(NotUsed);
+
+  zFile = (const char *)sqlite3_value_text(argv[0]);
+  zName = (const char *)sqlite3_value_text(argv[1]);
+  if( zFile==0 ) zFile = "";
+  if( zName==0 ) zName = "";
+
+  zErrDyn = NULL;
+  rc = comdb2_dynamic_attach(db, context, 0, argv, zName, zFile, &zErrDyn, 0, 0, 0, 0, 0);
+  if( zErrDyn ){
+    sqlite3_result_error(context, zErrDyn, -1);
+    sqlite3DbFree(db, zErrDyn);
+  }
+  if( rc ) sqlite3_result_error_code(context, rc);
+}
+
+void comdb2_dynamic_detach(sqlite3 *db, int idx)
+{
+  Db *pDb = &db->aDb[idx];
+  sqlite3BtreeClose(pDb->pBt);
+  pDb->pBt = 0;
+  pDb->pSchema = 0;
+  sqlite3CollapseDatabaseArray(db);
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** An SQL user-function registered to do the work of an DETACH statement. The
 ** three arguments to the function come directly from a detach statement:
@@ -323,10 +519,14 @@
     goto detach_error;
   }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  comdb2_dynamic_detach(db, i);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   sqlite3BtreeClose(pDb->pBt);
   pDb->pBt = 0;
   pDb->pSchema = 0;
   sqlite3CollapseDatabaseArray(db);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   return;
 
 detach_error:
