--- sqlite-3.28.0/src/callback.c	2026-01-09 16:11:01
+++ src/callback.c	2025-09-24 15:21:58
@@ -459,10 +459,56 @@
   sqlite3HashClear(&pSchema->fkeyHash);
   pSchema->pSeqTab = 0;
   if( pSchema->schemaFlags & DB_SchemaLoaded ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /*
+    ** we clear the schema when things change and we detect that
+    ** comdb2 uses different mechanisms to protect against that, and
+    ** sqlite3BtreeUpdateMeta is not supported (sqlite3BtreeGetMeta
+    ** returns a constant 0)
+    **/
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     pSchema->iGeneration++;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* TODO: Should the flags be changed for comdb2? */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pSchema->schemaFlags &= ~(DB_SchemaLoaded|DB_ResetWanted);
+}
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/*
+** Follows the idea of sqlite3SchemaClear,but only clears tables and
+** indexes for table "tblname"
+**
+** TODO: review triggers
+** 
+*/
+void sqlite3SchemaClearByName(void *p, const char *tblname){
+  HashElem *pElem;
+  Schema *pSchema = (Schema *)p;
+  int len = sqlite3Strlen30(tblname);
+  int len2;
+
+  for(pElem=sqliteHashFirst(&pSchema->tblHash);
+   pElem;
+   pElem=sqliteHashNext(pElem)){
+
+    Table *pTab = sqliteHashData(pElem);
+    len2 = sqlite3Strlen30(pTab->zName);
+
+    if(len==len2 && strncmp(pTab->zName, tblname, len) == 0){
+
+      /* first we unhash - there is no Delete in a perfect hash */
+      sqlite3HashInsert(&pSchema->tblHash, tblname, 0);
+
+      /* second we blast it away */
+      sqlite3DeleteTable(0, pTab);
+      break;
+    }
+  }
 }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Find and return the schema associated with a BTree.  Create
