--- sqlite-3.28.0/src/func.c	2026-01-09 16:11:01
+++ src/func.c	2025-09-24 15:21:58
@@ -17,7 +17,19 @@
 #include <stdlib.h>
 #include <assert.h>
 #include "vdbeInt.h"
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <unistd.h>
+#include <uuid/uuid.h>
+#include <memcompare.c>
+#include <zlib.h>
+#include "comdb2.h"
+#include "sql.h"
+#include "bdb_int.h"
+#include "md5.h"
+#include "tohex.h"
 
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Return the collating function associated with a function.
 */
@@ -78,7 +90,15 @@
   int NotUsed,
   sqlite3_value **argv
 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  static const char *azType[] = {
+    "integer", "real", "text", "blob", "null",
+    "datetime", "interval_ym", "interval_ds",
+    "datetime", "interval_ds", "decimal"
+  };
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   static const char *azType[] = { "integer", "real", "text", "blob", "null" };
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   int i = sqlite3_value_type(argv[0]) - 1;
   UNUSED_PARAMETER(NotUsed);
   assert( i>=0 && i<ArraySize(azType) );
@@ -87,6 +107,15 @@
   assert( SQLITE_TEXT==3 );
   assert( SQLITE_BLOB==4 );
   assert( SQLITE_NULL==5 );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( SQLITE_DATETIME==6 );
+  assert( SQLITE_INTERVAL_YM==7 );
+  assert( SQLITE_INTERVAL_DS==8 );
+  assert( SQLITE_DATETIMEUS==9 );
+  assert( SQLITE_INTERVAL_DSUS==10 );
+  assert( SQLITE_DECIMAL==11 );
+  assert( SQLITE_NEXTSEQ==(SQLITE_MAX_U32-2) );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   /* EVIDENCE-OF: R-01470-60482 The sqlite3_value_type(V) interface returns
   ** the datatype code for the initial datatype of the sqlite3_value object
   ** V. The returned value is one of SQLITE_INTEGER, SQLITE_FLOAT,
@@ -106,6 +135,15 @@
   assert( argc==1 );
   UNUSED_PARAMETER(argc);
   switch( sqlite3_value_type(argv[0]) ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    case SQLITE_INTERVAL_YM:
+    case SQLITE_INTERVAL_DS:
+    case SQLITE_INTERVAL_DSUS:
+    case SQLITE_DATETIME:
+    case SQLITE_DATETIMEUS:
+    case SQLITE_DECIMAL:
+    case SQLITE_NEXTSEQ:
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     case SQLITE_BLOB:
     case SQLITE_INTEGER:
     case SQLITE_FLOAT: {
@@ -164,6 +202,27 @@
       sqlite3_result_null(context);
       break;
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    case SQLITE_DECIMAL: {
+      decContext ctx;
+      intv_t *val = (intv_t*)sqlite3_value_interval( argv[0], SQLITE_DECIMAL);
+      intv_t res_val;
+
+      dec_ctx_init(&ctx, DEC_INIT_DECQUAD, gbl_decimal_rounding);
+
+      decQuadAbs( &res_val.u.dec, &val->u.dec, &ctx);
+      if( dfp_conv_check_status( &ctx, "quad", "abs(quad)") ){
+          sqlite3_result_error(context, "decimal overflow", -1);
+          break;
+      }
+
+      res_val.type = INTV_DECIMAL_TYPE;
+      res_val.sign = 0;
+
+      sqlite3_result_interval(context, &res_val);
+      break;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     default: {
       /* Because sqlite3_value_double() returns 0.0 if the argument is not
       ** something that can be converted into a number, we have:
@@ -294,6 +353,9 @@
   }
   p0type = sqlite3_value_type(argv[0]);
   p1 = sqlite3_value_int(argv[1]);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if (p1 == 0) p1 = 1;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( p0type==SQLITE_BLOB ){
     len = sqlite3_value_bytes(argv[0]);
     z = sqlite3_value_blob(argv[0]);
@@ -362,9 +424,132 @@
       if( p2<0 ) p2 = 0;
     }
     sqlite3_result_blob64(context, (char*)&z[p1], (u64)p2, SQLITE_TRANSIENT);
+  }
+}
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+extern int comdb2_sql_tick();
+/*
+** Implementation of the sleep() function
+*/
+static void sleepFunc(sqlite3_context *context, int argc, sqlite3_value *argv[]) {
+  int n;
+  if( argc != 1 ){
+    sqlite3_result_int(context, -1);
+    return;
+  }
+  n = sqlite3_value_int(argv[0]);
+  if( n < 0 ){
+    sqlite3_result_int(context, -1);
+    return;
+  }
+  int i;
+  for(i = 0; i < n; i++) {
+    sleep(1);
+    if( comdb2_sql_tick() ){
+      sqlite3_result_error_code(context, SQLITE_ERROR);
+      return;
+    }
+  }
+  sqlite3_result_int(context, i);
+}
+
+static void usleepFunc(sqlite3_context *context, int argc, sqlite3_value *argv[]) {
+  int total, remain, us;
+  if( argc != 1 ){
+    sqlite3_result_int(context, -1);
+    return;
+  }
+  total = remain = sqlite3_value_int(argv[0]);
+  if( total < 0 ){
+    sqlite3_result_int(context, -1);
+    return;
+  }
+  while( remain > 0 ){
+    us = ( remain > 1000000 ) ? 1000000 : remain;
+    remain -= us;
+    usleep(us);
+    if( us == 1000000 && comdb2_sql_tick() ){
+      sqlite3_result_error_code(context, SQLITE_ERROR);
+      return;
+    }
+  }
+  sqlite3_result_int(context, (total - remain));
+}
+
+static void tableNamesFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  sqlite3 *db;
+  int wasPrepareOnly;
+  Parse sParse;
+  const char *zSql;
+  char *zErrMsg;
+  if( sqlite3_value_type(argv[0])!=SQLITE_TEXT ) return;
+  zSql = (const char *)sqlite3_value_text(argv[0]);
+  if( !zSql ) return;
+  db = sqlite3_context_db_handle(context);
+  wasPrepareOnly = (db->flags&SQLITE_PREPARE_ONLY)!=0;
+  db->flags |= SQLITE_PrepareOnly;
+  memset(&sParse, 0, sizeof(Parse));
+  sParse.db = db;
+  sParse.prepFlags = SQLITE_PREPARE_ONLY|SQLITE_PREPARE_SRCLIST_ONLY;
+  zErrMsg = 0;
+  if( sqlite3RunParser(&sParse, zSql, &zErrMsg) ){
+    if( zErrMsg ){
+      sqlite3_result_error(context, zErrMsg, -1);
+      sqlite3DbFree(db, zErrMsg);
+    }
+    goto done;
+  }else if( zErrMsg ){ /* TODO: Is this ever possible? */
+    sqlite3DbFree(db, zErrMsg);
+  }
+  if( sParse.nSrcListOnly>0 ){
+    StrAccum str;
+    int i;
+    sqlite3StrAccumInit(&str, db, 0, 0, db->aLimit[SQLITE_LIMIT_LENGTH]);
+    for(i=0; i<sParse.nSrcListOnly; i++){
+      if( i>0 ) sqlite3_str_append(&str, " ", 1);
+      sqlite3_str_appendall(&str, sParse.azSrcListOnly[i]);
+    }
+    sqlite3_result_text(context, sqlite3StrAccumFinish(&str), -1, SQLITE_DYNAMIC);
   }
+done:
+  if( !wasPrepareOnly ) db->flags &= ~SQLITE_PrepareOnly;
+  if( sParse.pVdbe ){
+    sqlite3VdbeFinalize(sParse.pVdbe);
+    sParse.pVdbe = 0;
+  }
+  sqlite3ParserReset(&sParse);
 }
 
+static void normalizeSqlFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  int rc;
+  sqlite3 *db;
+  const char *zSql;
+  sqlite3_stmt *pStmt = NULL;
+  if( sqlite3_value_type(argv[0])!=SQLITE_TEXT ) return;
+  zSql = (const char *)sqlite3_value_text(argv[0]);
+  if( !zSql ) return;
+  db = sqlite3_context_db_handle(context);
+  rc = sqlite3_prepare_v3(db, zSql, -1, SQLITE_PREPARE_NORMALIZE |
+                          SQLITE_PREPARE_ONLY, &pStmt, NULL);
+  if( rc==SQLITE_OK ){
+    sqlite3_result_text(context, sqlite3_normalized_sql(pStmt),
+                        -1, SQLITE_TRANSIENT);
+    sqlite3_finalize(pStmt);
+  }else{
+    sqlite3_result_error(context, sqlite3_errmsg(db), -1);
+  }
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Implementation of the round() function
 */
@@ -386,10 +571,17 @@
   ** handle the rounding directly,
   ** otherwise use printf.
   */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( n==0 && r>=0 && r<(double)(LARGEST_INT64-1) ){
+    r = (double)((sqlite_int64)(r+0.5));
+  }else if( n==0 && r<0 && (-r)<(double)(LARGEST_INT64-1) ){
+    r = -(double)((sqlite_int64)((-r)+0.5));
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( n==0 && r>=0 && r<LARGEST_INT64-1 ){
     r = (double)((sqlite_int64)(r+0.5));
   }else if( n==0 && r<0 && (-r)<LARGEST_INT64-1 ){
     r = -(double)((sqlite_int64)((-r)+0.5));
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }else{
     zBuf = sqlite3_mprintf("%.*f",n,r);
     if( zBuf==0 ){
@@ -526,9 +718,707 @@
   if( p ){
     sqlite3_randomness(n, p);
     sqlite3_result_blob(context, (char*)p, n, sqlite3_free);
+  }
+}
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/* 36 characters + trailing '\0' - taken from `man uuid_unparse` */
+#define GUID_STR_LENGTH 37
+
+static void guidFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  assert( argc==0 );
+  UNUSED_PARAMETER(argc);
+
+  uuid_t guid;
+  uuid_generate(guid);
+
+  sqlite3_result_blob(context, (char*)guid, sizeof(uuid_t), SQLITE_TRANSIENT);
+}
+
+static void guidStrFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  assert( argc==0 );
+  UNUSED_PARAMETER(argc);
+
+  uuid_t guid;
+  uuid_generate(guid);
+
+  char guid_str[GUID_STR_LENGTH];
+  uuid_unparse(guid, guid_str);
+
+  sqlite3_result_text(context, guid_str, GUID_STR_LENGTH - 1, SQLITE_TRANSIENT);
+}
+
+static void guidFromStrFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  assert( argc==1 );
+  UNUSED_PARAMETER(argc);
+
+  if(sqlite3_value_type(argv[0]) != SQLITE_TEXT) {
+      return;
+  }
+
+  /* be liberal in what we accept - uuid_parse expects a string
+   * in the format of "%08x-%04x-%04x-%04x-%012x", 36 bytes plus the trailing '\0')
+   * but will fail later if we don't have that
+   */
+  if(sqlite3_value_bytes(argv[0]) < (GUID_STR_LENGTH - 1)) {
+      return;
+  }
+
+  const unsigned char * guid_str = sqlite3_value_text(argv[0]);
+
+  uuid_t guid;
+  if(uuid_parse((const char*)guid_str, guid) == 0) {
+      sqlite3_result_blob(context, (char*)guid, sizeof(uuid_t), SQLITE_TRANSIENT);
+  }
+}
+
+static void guidFromByteFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  assert( argc==1 );
+  UNUSED_PARAMETER(argc);
+
+  if(sqlite3_value_type(argv[0]) != SQLITE_BLOB) {
+      return;
+  }
+
+  if(sqlite3_value_bytes(argv[0]) != sizeof(uuid_t)) {
+      return;
+  }
+
+  const void * guid_blob = sqlite3_value_blob(argv[0]);
+
+  char guid_str[GUID_STR_LENGTH];
+  uuid_unparse(guid_blob, guid_str);
+
+  sqlite3_result_text(context, guid_str, GUID_STR_LENGTH - 1, SQLITE_TRANSIENT);
+}
+
+static void comdb2DoubleToBlobFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  u8 aByte[8];
+  assert( argc==1 );
+  if( sqlite3_value_type(argv[0])==SQLITE_INTEGER ){
+    sqlite3Put8byte(aByte, (u64)sqlite3_value_int64(argv[0]));
+  }else if( sqlite3_value_type(argv[0])==SQLITE_FLOAT ){
+    sqlite3Put8byte(aByte, (u64)sqlite3DoubleToInt64(
+                    sqlite3_value_double(argv[0])));
+  }else{
+    sqlite3_result_null(context);
+    return;
+  }
+  sqlite3_result_blob(context, (char*)aByte, sizeof(aByte), SQLITE_TRANSIENT);
+}
+
+static void comdb2BlobToDoubleFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  assert( argc==1 );
+  if( sqlite3_value_type(argv[0])!=SQLITE_BLOB ){
+    sqlite3_result_null(context);
+    return;
+  }
+  if( sqlite3_value_bytes(argv[0])!=8 ){
+    sqlite3_result_null(context);
+    return;
+  }
+  sqlite3_result_double(context, sqlite3Int64ToDouble(
+                        (i64)sqlite3Get8byte(sqlite3_value_blob(argv[0]))));
+}
+
+/*
+** Implementation of the comdb2_sysinfo() SQL function.  The return
+** value depends on the class of system information being requested;
+** however, it is generally pieces of rarely changing (scalar) data
+** related to the overall state of the server -OR- the current SQL
+** query being processed.
+*/
+static void comdb2SysinfoFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  const char *zName;
+  assert( argc==1 );
+  if( sqlite3_value_type(argv[0])!=SQLITE_TEXT ){
+    return;
+  }
+  zName = (const char *)sqlite3_value_text(argv[0]);
+  if( sqlite3_stricmp(zName, "pid")==0 ){
+    sqlite3_result_int64(context, (sqlite3_int64)getpid());
+  }else if( sqlite3_stricmp(zName, "tid")==0 ){
+    sqlite3_result_int64(context, (sqlite3_int64)getarchtid());
+  }else if( sqlite3_stricmp(zName, "master")==0 ){
+    sqlite3_result_text(context, thedb->bdb_env->repinfo->master_host, -1,
+                        SQLITE_TRANSIENT);
+  }else if( sqlite3_stricmp(zName, "host")==0 ){
+    char zHostName[1024];
+    memset(zHostName, 0, sizeof(zHostName));
+    if( gethostname(zHostName, sizeof(zHostName))==0 ){
+      sqlite3_result_text(context, zHostName, -1, SQLITE_TRANSIENT);
+    }else{
+      sqlite3_result_error(context, "unable to obtain host name", -1);
+    }
+  }else if( sqlite3_stricmp(zName, "class")==0 ){
+    sqlite3_result_text(context, get_my_mach_class_str(), -1, SQLITE_TRANSIENT);
+  }else if( sqlite3_stricmp(zName, "version")==0 ){
+    char *zVersion = sqlite3_mprintf("[%s] [%s] [%s] [%s] [%s]", gbl_db_version,
+                                     gbl_db_codename, gbl_db_semver,
+                                     gbl_db_git_version_sha, gbl_db_buildtype);
+    sqlite3_result_text(context, zVersion, -1, SQLITE_TRANSIENT);
+    sqlite3_free(zVersion);
+  }
+}
+
+/*
+** Implementation of the comdb2_ctxinfo() SQL function.  The return
+** value depends on the class of context information being requested;
+** however, it is generally pieces of (scalar) data related to the
+** state of the client connection -OR- the current SQL query being
+** processed.
+*/
+static void comdb2CtxinfoFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  const char *zName;
+  assert( argc==1 );
+  if( sqlite3_value_type(argv[0])!=SQLITE_TEXT ){
+    return;
+  }
+
+  struct sqlclntstate *clnt = get_sql_clnt();
+
+  zName = (const char *)sqlite3_value_text(argv[0]);
+  if( sqlite3_stricmp(zName, "parallel")==0 ){
+    if( clnt ){
+      sqlite3_result_int(context, clnt->conns!=NULL);
+    }
+  }else if( sqlite3_stricmp(zName, "ruleset_result")==0 ){
+    if( clnt ){
+      sqlite3_result_text(context, clnt->work.zRuleRes, -1, SQLITE_STATIC);
+    }
+  }else if( sqlite3_stricmp(zName, "user")==0 ){
+    sqlite3_result_text(context, get_current_user(clnt), -1, SQLITE_STATIC);
+  }
+}
+
+/*
+** Implementation of the comdb2_version() SQL function.  The return
+** value is the same as the stat value for version
+*/
+static void comdb2VersionFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  sqlite3_result_text(context, gbl_db_version, -1, SQLITE_STATIC);
+}
+
+static void comdb2SemVerFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  sqlite3_result_text(context, gbl_db_semver, -1, SQLITE_STATIC);
+}
+
+extern char * comdb2_get_prev_query_cost();
+extern char * comdb2_free_prev_query_cost();
+
+static void comdb2PrevquerycostFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  char * cost = comdb2_get_prev_query_cost();
+  if(cost) {
+      sqlite3_result_text(context, cost, -1, SQLITE_TRANSIENT);
+      comdb2_free_prev_query_cost();
+  }
+}
+
+static void comdb2HostFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  sqlite3_result_text(context, gbl_myhostname, -1, SQLITE_STATIC);
+}
+
+extern int comdb2_get_server_port();
+static void comdb2PortFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  sqlite3_result_int64(context, comdb2_get_server_port());
+}
+
+
+static void comdb2DbnameFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  extern char gbl_dbname[];
+  sqlite3_result_text(context, gbl_dbname, -1, SQLITE_STATIC);
+}
+
+extern unsigned long long comdb2_table_version(const char *tablename);
+/*
+** Implementation of the table_version() SQL function.  This returns
+** the comdb2 table version
+*/
+static void tableVersionFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  const char *tablename;
+  sqlite_int64 version;
+
+  assert(argc==1);
+  if(sqlite3_value_type(argv[0]) != SQLITE_TEXT) {
+    return;
+  }
+
+  tablename = (const char *)sqlite3_value_text(argv[0]);
+
+  version = comdb2_table_version(tablename);
+  if (version<0) {
+    return;
+  }
+
+  sqlite3_result_int64(context, version);
+}
+
+extern char* comdb2_partition_info(const char *partition, const char *option);
+/*
+** Implementation of the table_version() SQL function.  This returns
+** the comdb2 table version
+*/
+static void partitionInfoFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  const char   *partition_name;
+  const char   *option;
+  char         *info;
+
+  assert( argc==2 );
+  if( sqlite3_value_type(argv[0]) != SQLITE_TEXT ){
+    return;
+  }
+  partition_name = (const char *)sqlite3_value_text(argv[0]);
+  
+  if( sqlite3_value_type(argv[1]) != SQLITE_TEXT ){
+    return;
+  }
+  option = (const char *)sqlite3_value_text(argv[1]);
+
+  info = comdb2_partition_info(partition_name, option);
+
+  if( !info ){
+
+    sqlite3_result_null(context);
+  }else{
+    sqlite3_result_text(context, info, -1, SQLITE_TRANSIENT);
+
+    if(info) {
+      free(info);
+    }
+  }
+}
+
+/*
+** Implementation of the comdb2_starttime() SQL function.
+*/
+static void comdb2StartTimeFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  extern int gbl_starttime;
+  dttz_t dt = {gbl_starttime, 0};
+  sqlite3_result_datetime(context, &dt, NULL);
+}
+
+static void md5Func(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+  ){
+    const unsigned char *s;
+    MD5Context ctx = {0};
+    char checksum[16];
+    char out[33];
+    unsigned int len;
+    int type = sqlite3_value_type(argv[0]);
+
+    switch (type) {
+        case SQLITE_TEXT:
+            s = sqlite3_value_text(argv[0]);
+            len = strlen((char*) s);
+            break;
+
+        case SQLITE_BLOB:
+            s = sqlite3_value_blob(argv[0]);
+            len = sqlite3_value_bytes(argv[0]);
+            break;
+
+        default:
+            return;
+    }
+
+    MD5Init(&ctx);
+    MD5Update(&ctx, (unsigned char*) s, len);
+    MD5Final((unsigned char*) checksum, &ctx);
+    util_tohex(out, (char*) checksum, sizeof(checksum));
+    sqlite3_result_text(context, strdup(out), 32, free);
+}
+
+static void comdb2UserFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+
+  sqlite3_result_text(context, get_current_user(get_sql_clnt()), -1,
+                      SQLITE_STATIC);
+}
+
+int gbl_use_current_lsn_for_non_snapshot = 0;
+
+static void comdb2SnapshotLsnFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  struct sqlclntstate *clnt = get_sql_clnt();
+  char lsn[MAX_LSN_STR];
+  int sz = 0;
+  if (clnt != NULL && clnt->dbtran.mode == TRANLEVEL_MODSNAP) {
+    sz = snprintf(lsn, sizeof(lsn), "{%d:%d}",
+             clnt->modsnap_start_lsn_file,
+             clnt->modsnap_start_lsn_offset);
+  } else if (gbl_use_current_lsn_for_non_snapshot) {
+    DB_LSN curlsn;
+    __log_txn_lsn(thedb->bdb_env->dbenv, &curlsn, NULL, NULL);
+    sz = snprintf(lsn, sizeof(lsn), "{%d:%d}", curlsn.file, curlsn.offset);
+  } else {
+    sqlite3_result_null(context);
+    return;
   }
+
+  assert(sz < sizeof(lsn));
+  (void)sz; // suppress unused variable warning
+  sqlite3_result_text(context, lsn, -1, SQLITE_TRANSIENT);
+}
+
+/* The code below is based on the example presented here: https://www.zlib.net/zpipe.c */
+
+#define CHUNK 32768
+static unsigned char *compressFunc_int(sqlite3_context *context,
+                                       unsigned char *uncomprData,
+                                       unsigned int   uncomprLen,
+                                       unsigned int  *comprLen,
+                                       int(*deflateInitFunc)(z_stream*))  {
+    unsigned char buffer[CHUNK];     // Buffer to stage compressed data
+    unsigned char *comprData = NULL; // Final buffer (realloc-ed when needed)
+    int rc;
+    z_stream stream;
+
+    *comprLen = 0;
+
+    stream.zalloc = Z_NULL;
+    stream.zfree = Z_NULL;
+    stream.opaque = Z_NULL;
+
+    rc = deflateInitFunc(&stream);
+    if (rc != Z_OK) {
+        sqlite3_result_error(context, "zlib: failed to initialize", -1);
+        return NULL;
+    }
+
+    stream.avail_in = uncomprLen;
+    stream.next_in = uncomprData;
+
+    do {
+        stream.avail_out = CHUNK;
+        stream.next_out = buffer;
+
+        rc = deflate(&stream, Z_FINISH);
+        if (rc == Z_STREAM_ERROR) {
+            sqlite3_result_error(context, "zlib: stream error", -1);
+            deflateEnd(&stream);
+            return NULL;
+        }
+
+        unsigned int compr_bytes = CHUNK - stream.avail_out;
+        comprData = (comprData == NULL)
+            ? (unsigned char *)contextMalloc(context, compr_bytes)
+            : (unsigned char *)sqlite3_realloc(comprData, *comprLen + compr_bytes);
+        memcpy(comprData + *comprLen, buffer, compr_bytes);
+        *comprLen += compr_bytes;
+    } while (stream.avail_out == 0);
+
+    deflateEnd(&stream);
+    return comprData;
+}
+
+static unsigned char *uncompressFunc_int(sqlite3_context *context,
+                                         unsigned char *comprData,
+                                         unsigned int   comprLen,
+                                         unsigned int  *uncomprLen,
+                                         int(*inflateInitFunc)(z_stream *)) {
+    unsigned char buffer[CHUNK];       // Buffer to stage uncompressed data
+    unsigned char *uncomprData = NULL; // Final buffer (realloc-ed when needed)
+    int rc;
+    z_stream stream;
+
+    *uncomprLen = 0;
+
+    stream.zalloc = Z_NULL;
+    stream.zfree = Z_NULL;
+    stream.opaque = Z_NULL;
+
+    rc = inflateInitFunc(&stream);
+    if (rc != Z_OK) {
+        sqlite3_result_error(context, "zlib: failed to initialize", -1);
+        return NULL;
+    }
+
+    stream.avail_in = comprLen;
+    stream.next_in = comprData;
+
+    do {
+        stream.avail_out = CHUNK;
+        stream.next_out = buffer;
+
+        rc = inflate(&stream, Z_NO_FLUSH);
+        if (rc < 0) {
+            sqlite3_result_error(context, "zlib: internal error", -1);
+            inflateEnd(&stream);
+            return NULL;
+        }
+
+        unsigned int uncompr_bytes = CHUNK - stream.avail_out;
+        uncomprData = (uncomprData == NULL)
+            ? (unsigned char *)contextMalloc(context, uncompr_bytes)
+            : (unsigned char *)sqlite3_realloc(uncomprData, *uncomprLen + uncompr_bytes);
+        memcpy(uncomprData + *uncomprLen, buffer, uncompr_bytes);
+        *uncomprLen += uncompr_bytes;
+    } while (stream.avail_out == 0);
+
+    inflateEnd(&stream);
+    return uncomprData;
+}
+
+static int deflateInit__(z_stream *stream) {
+    return deflateInit(stream, Z_DEFAULT_COMPRESSION);
+}
+
+static int inflateInit__(z_stream *stream) {
+    return inflateInit(stream);
+}
+
+static int deflateInit2__(z_stream *stream) {
+    return deflateInit2(stream, Z_DEFAULT_COMPRESSION,
+                        Z_DEFLATED, 15 | 16 /* Add gzip header & trailer */, 8,
+                        Z_DEFAULT_STRATEGY);
+}
+
+static int inflateInit2__(z_stream *stream) {
+    return inflateInit2(stream, 15 | 16);
+}
+
+static void compressFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+  ){
+    unsigned char *uncomprData;
+    unsigned int uncomprLen;
+
+    int type = sqlite3_value_type(argv[0]);
+
+    switch (type) {
+        case SQLITE_TEXT:
+            uncomprData = (unsigned char *)sqlite3_value_text(argv[0]);
+            uncomprLen = strlen((char*) uncomprData);
+            break;
+
+        case SQLITE_BLOB:
+            uncomprData = (unsigned char *)sqlite3_value_blob(argv[0]);
+            uncomprLen = sqlite3_value_bytes(argv[0]);
+            break;
+
+        case SQLITE_NULL:
+            sqlite3_result_null(context);
+            return;
+
+        default:
+            sqlite3_result_error(context, "zlib: data type not supported", -1);
+            return;
+    }
+
+    unsigned char *comprData = NULL;
+    unsigned int comprLen = 0;
+    comprData = compressFunc_int(context, uncomprData, uncomprLen, &comprLen,
+                                 deflateInit__);
+
+    if (comprData != NULL) {
+        sqlite3_result_blob(context, comprData, comprLen, NULL);
+    }
+    return;
+}
+
+static void uncompressFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+  ){
+    unsigned char *comprData;
+    unsigned int comprLen;
+
+    int type = sqlite3_value_type(argv[0]);
+
+    switch (type) {
+        case SQLITE_BLOB:
+            comprData = (unsigned char *)sqlite3_value_blob(argv[0]);
+            comprLen = sqlite3_value_bytes(argv[0]);
+            break;
+
+        case SQLITE_NULL:
+            sqlite3_result_null(context);
+            return;
+
+        default:
+            sqlite3_result_error(context, "zlib: data type not supported", -1);
+            return;
+    }
+
+    unsigned char *uncomprData = NULL;
+    unsigned int uncomprLen = 0;
+    uncomprData = uncompressFunc_int(context, comprData, comprLen,
+                                     &uncomprLen, inflateInit__);
+
+    if (uncomprData != NULL) {
+        sqlite3_result_blob(context, uncomprData, uncomprLen, NULL);
+    }
+    return;
+}
+
+static void compressGzipFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+  ){
+    unsigned char *uncomprData;
+    unsigned int uncomprLen;
+
+    int type = sqlite3_value_type(argv[0]);
+
+    switch (type) {
+        case SQLITE_TEXT:
+            uncomprData = (unsigned char *)sqlite3_value_text(argv[0]);
+            uncomprLen = strlen((char*) uncomprData);
+            break;
+
+        case SQLITE_BLOB:
+            uncomprData = (unsigned char *)sqlite3_value_blob(argv[0]);
+            uncomprLen = sqlite3_value_bytes(argv[0]);
+            break;
+
+        case SQLITE_NULL:
+            sqlite3_result_null(context);
+            return;
+
+        default:
+            sqlite3_result_error(context, "zlib: data type not supported", -1);
+            return;
+    }
+
+    unsigned char *comprData = NULL;
+    unsigned int comprLen = 0;
+    comprData = compressFunc_int(context, uncomprData, uncomprLen, &comprLen,
+                                 deflateInit2__);
+
+    if (comprData != NULL) {
+        sqlite3_result_blob(context, comprData, comprLen, sqlite3_free);
+    }
+    return;
+}
+
+static void uncompressGzipFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+  ){
+    unsigned char *comprData;
+    unsigned int comprLen;
+
+    int type = sqlite3_value_type(argv[0]);
+
+    switch (type) {
+        case SQLITE_BLOB:
+            comprData = (unsigned char *)sqlite3_value_blob(argv[0]);
+            comprLen = sqlite3_value_bytes(argv[0]);
+            break;
+
+        case SQLITE_NULL:
+            sqlite3_result_null(context);
+            return;
+
+        default:
+            sqlite3_result_error(context, "zlib: data type not supported", -1);
+            return;
+    }
+
+    unsigned char *uncomprData = NULL;
+    unsigned int uncomprLen = 0;
+    uncomprData = uncompressFunc_int(context, comprData, comprLen,
+                                     &uncomprLen, inflateInit2__);
+
+    if (uncomprData != NULL) {
+        sqlite3_result_blob(context, uncomprData, uncomprLen, sqlite3_free);
+    }
+    return;
 }
 
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Implementation of the last_insert_rowid() SQL function.  The return
 ** value is the same as the sqlite3_last_insert_rowid() API function.
@@ -579,7 +1469,12 @@
   sqlite3_result_int(context, sqlite3_total_changes(db));
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 /*
+** Moved to "sqliteInt.h" for use by Comdb2.
+*/
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+/*
 ** A structure defining how to do GLOB-style comparisons.
 */
 struct compareInfo {
@@ -588,6 +1483,7 @@
   u8 matchSet;          /* "[" or 0 */
   u8 noCase;            /* true to ignore case differences */
 };
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** For LIKE and GLOB matching on EBCDIC machines, assume that every
@@ -655,7 +1551,11 @@
 **
 ** This routine is usually quick, but can be N**2 in the worst case.
 */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+int patternCompare(
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 static int patternCompare(
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   const u8 *zPattern,              /* The glob pattern */
   const u8 *zString,               /* The string to compare against the glob */
   const struct compareInfo *pInfo, /* Information about how to do the compare */
@@ -1484,6 +2384,10 @@
   i64 cnt;          /* Number of elements summed */
   u8 overflow;      /* True if integer overflow seen */
   u8 approx;        /* True if non-integer value was input to the sum */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  u8 decs;          /* True if summing decimals */
+  decQuad decSum;   /* decQuad aggregation */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 };
 
 /*
@@ -1511,6 +2415,49 @@
       if( (p->approx|p->overflow)==0 && sqlite3AddInt64(&p->iSum, v) ){
         p->approx = p->overflow = 1;
       }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    }else if( type==SQLITE_DECIMAL ){
+      intv_t v = *(intv_t*)sqlite3_value_interval(argv[0], SQLITE_DECIMAL);
+
+      if( p->decs==0 ){
+        p->decSum = v.u.dec;
+        p->decs = 1;
+
+        if( 0 ){
+          char aaa[128];
+          char bbb[128];
+
+          decQuadToString( &p->decSum, aaa);
+          decQuadToString( &v.u.dec, bbb);
+
+          fprintf(stderr, "%s  = %s\n", aaa, bbb);
+        }
+      }else{
+        decContext ctx;
+        decQuad    res;
+
+        dec_ctx_init( &ctx, DEC_INIT_DECQUAD, gbl_decimal_rounding);
+        decQuadAdd( &res, &p->decSum, &v.u.dec, &ctx);
+
+        if( dfp_conv_check_status(&ctx, "quad", "add(quads)") ){
+          sqlite3_result_error(context, "decimal overflow", -1);
+        }
+
+        if( 0 ){
+          char aaa[128];
+          char bbb[128];
+          char ccc[128];
+
+          decQuadToString( &p->decSum, aaa);
+          decQuadToString( &v.u.dec, bbb);
+          decQuadToString( &res, ccc);
+
+          fprintf(stderr, "%s + %s = %s\n", aaa, bbb, ccc);
+        }
+
+        p->decSum = res;
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }else{
       p->rSum += sqlite3_value_double(argv[0]);
       p->approx = 1;
@@ -1551,6 +2498,14 @@
       sqlite3_result_error(context,"integer overflow",-1);
     }else if( p->approx ){
       sqlite3_result_double(context, p->rSum);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    }else if( p->decs){
+       intv_t res;
+       res.type = INTV_DECIMAL_TYPE;
+       res.sign = 0;
+       res.u.dec = p->decSum;
+       sqlite3_result_interval(context, &res);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }else{
       sqlite3_result_int64(context, p->iSum);
     }
@@ -1560,6 +2515,27 @@
   SumCtx *p;
   p = sqlite3_aggregate_context(context, 0);
   if( p && p->cnt>0 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( p->decs ){
+      decContext ctx;
+      decQuad denom;
+      decQuad res;
+      intv_t  tv;
+
+      dec_ctx_init( &ctx, DEC_INIT_DECQUAD, gbl_decimal_rounding);
+      decQuadFromInt32( &denom, p->cnt);
+      decQuadDivide( &res, &p->decSum, &denom, &ctx);
+      if( dfp_conv_check_status(&ctx, "quad", "divide(quad)") ){
+        sqlite3_result_error(context, "decimal overflow", -1);
+      }
+
+      tv.type = INTV_DECIMAL_TYPE;
+      tv.sign = 0;
+      tv.u.dec = res;
+
+      sqlite3_result_interval( context, &tv);
+    } else
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3_result_double(context, p->rSum/(double)p->cnt);
   }
 }
@@ -1567,6 +2543,16 @@
   SumCtx *p;
   p = sqlite3_aggregate_context(context, 0);
   /* (double)0 In case of SQLITE_OMIT_FLOATING_POINT... */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( p && p->decs ){
+    intv_t res;
+    res.type = INTV_DECIMAL_TYPE;
+    res.sign = 0;
+    res.u.dec = p->decSum;
+    sqlite3_result_interval(context, &res);
+  }
+  else
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   sqlite3_result_double(context, p ? p->rSum : (double)0);
 }
 
@@ -1703,7 +2689,14 @@
   if( pAccum ){
     sqlite3 *db = sqlite3_context_db_handle(context);
     int firstTerm = pAccum->mxAlloc==0;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    struct sqlclntstate *clnt = get_sql_clnt();
+
+    pAccum->mxAlloc = (clnt && clnt->group_concat_mem_limit != 0) ?
+      clnt->group_concat_mem_limit : db->aLimit[SQLITE_LIMIT_LENGTH];
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     pAccum->mxAlloc = db->aLimit[SQLITE_LIMIT_LENGTH];
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( !firstTerm ){
       if( argc==2 ){
         zSep = (char*)sqlite3_value_text(argv[1]);
@@ -1717,6 +2710,14 @@
     zVal = (char*)sqlite3_value_text(argv[0]);
     nVal = sqlite3_value_bytes(argv[0]);
     if( zVal ) sqlite3_str_append(pAccum, zVal, nVal);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    // NC: Attach an error message here as we do not want client to see the
+    // default error message 'transaction too big' emitted for ERR_TRAN_TOO_BIG,
+    // which SQLITE_TOOBIG later gets translated to.
+    if( clnt && pAccum->accError==SQLITE_TOOBIG ){
+      clnt->sqlite_errstr = "string or blob too big";
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
 }
 #ifndef SQLITE_OMIT_WINDOWFUNC
@@ -1885,6 +2886,70 @@
   return 1;
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#if defined(SQLITE_BUILDING_FOR_COMDB2_DBGLOG)
+static void dbglogCookieFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  uint64_t comdb2fastseed(int);
+  sqlite3_result_int64(context, (i64)comdb2fastseed(4));
+}
+
+static void dbglogBeginFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  UNUSED_PARAMETER(argc);
+  int dbglog_begin(const char *);
+  sqlite3_result_int(context, dbglog_begin((const char *)sqlite3_value_text(argv[0])));
+}
+
+static void dbglogEndFunc(
+  sqlite3_context *context,
+  int argc,
+  sqlite3_value **argv
+){
+  unsigned long long cookie;
+  void *mmapped;
+  size_t size;
+  int fd;
+  int rc;
+  int dbglog_mmap_dbglog_file(unsigned long long, void **, size_t *, int *);
+  int dbglog_munmap_dbglog_file(unsigned long long, void *, size_t, int);
+
+  UNUSED_PARAMETER(argc);
+  cookie = (unsigned long long)sqlite3_value_int64(argv[0]);
+
+  rc = dbglog_mmap_dbglog_file(cookie, &mmapped, &size, &fd);
+  if (rc != 0)
+      sqlite3_result_null(context);
+  else {
+      sqlite3_result_blob(context, mmapped, size, SQLITE_TRANSIENT);
+      (void)dbglog_munmap_dbglog_file(cookie, mmapped, size, fd);
+  }
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2_DBGLOG) */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+extern i64 comdb2_last_stmt_cost(void);
+static void comdb2LastCostFunc(
+  sqlite3_context *context,
+  int NotUsed,
+  sqlite3_value **NotUsed2
+){
+  i64 cost;
+  UNUSED_PARAMETER2(NotUsed, NotUsed2);
+  cost = comdb2_last_stmt_cost();
+  sqlite3_result_int64(context, cost);
+}
+
+#endif
+
 /*
 ** All of the FuncDef structures in the aBuiltinFunc[] array above
 ** to the global function hash table.  This occurs at start-time (as
@@ -1921,9 +2986,11 @@
     FUNCTION2(unlikely,          1, 0, 0, noopFunc,  SQLITE_FUNC_UNLIKELY),
     FUNCTION2(likelihood,        2, 0, 0, noopFunc,  SQLITE_FUNC_UNLIKELY),
     FUNCTION2(likely,            1, 0, 0, noopFunc,  SQLITE_FUNC_UNLIKELY),
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
 #ifdef SQLITE_DEBUG
     FUNCTION2(affinity,          1, 0, 0, noopFunc,  SQLITE_FUNC_AFFINITY),
 #endif
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifdef SQLITE_ENABLE_OFFSET_SQL_FUNC
     FUNCTION2(sqlite_offset,     1, 0, 0, noopFunc,  SQLITE_FUNC_OFFSET|
                                                      SQLITE_FUNC_TYPEOF),
@@ -1945,6 +3012,14 @@
     FUNCTION2(typeof,            1, 0, 0, typeofFunc,  SQLITE_FUNC_TYPEOF),
     FUNCTION2(length,            1, 0, 0, lengthFunc,  SQLITE_FUNC_LENGTH),
     FUNCTION(instr,              2, 0, 0, instrFunc        ),
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    VFUNCTION(sleep,             1, 0, 0, sleepFunc        ),
+    VFUNCTION(usleep,            1, 0, 0, usleepFunc       ),
+    VFUNCTION(comdb2_extract_table_names,
+                                 1, 0, 0, tableNamesFunc   ),
+    VFUNCTION(comdb2_normalize_sql,
+                                 1, 0, 0, normalizeSqlFunc ),
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     FUNCTION(printf,            -1, 0, 0, printfFunc       ),
     FUNCTION(unicode,            1, 0, 0, unicodeFunc      ),
     FUNCTION(char,              -1, 0, 0, charFunc         ),
@@ -1959,14 +3034,22 @@
     FUNCTION2(ifnull,            2, 0, 0, noopFunc,  SQLITE_FUNC_COALESCE),
     VFUNCTION(random,            0, 0, 0, randomFunc       ),
     VFUNCTION(randomblob,        1, 0, 0, randomBlob       ),
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    VFUNCTION(guid,              0, 0, 0, guidFunc         ),
+    VFUNCTION(guid_str,          0, 0, 0, guidStrFunc      ),
+    FUNCTION(guid,               1, 0, 0, guidFromStrFunc  ),
+    FUNCTION(guid_str,           1, 0, 0, guidFromByteFunc ),
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     FUNCTION(nullif,             2, 0, 1, nullifFunc       ),
     DFUNCTION(sqlite_version,    0, 0, 0, versionFunc      ),
     DFUNCTION(sqlite_source_id,  0, 0, 0, sourceidFunc     ),
     FUNCTION(sqlite_log,         2, 0, 0, errlogFunc       ),
     FUNCTION(quote,              1, 0, 0, quoteFunc        ),
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
     VFUNCTION(last_insert_rowid, 0, 0, 0, last_insert_rowid),
     VFUNCTION(changes,           0, 0, 0, changes          ),
     VFUNCTION(total_changes,     0, 0, 0, total_changes    ),
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
     FUNCTION(replace,            3, 0, 0, replaceFunc      ),
     FUNCTION(zeroblob,           1, 0, 0, zeroblobFunc     ),
     FUNCTION(substr,             2, 0, 0, substrFunc       ),
@@ -1991,6 +3074,37 @@
     LIKEFUNC(like, 2, &likeInfoNorm, SQLITE_FUNC_LIKE),
     LIKEFUNC(like, 3, &likeInfoNorm, SQLITE_FUNC_LIKE),
 #endif
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    FUNCTION(comdb2_double_to_blob, 1, 0, 0, comdb2DoubleToBlobFunc),
+    FUNCTION(comdb2_blob_to_double, 1, 0, 0, comdb2BlobToDoubleFunc),
+    FUNCTION(comdb2_sysinfo,        1, 0, 0, comdb2SysinfoFunc),
+    FUNCTION(comdb2_ctxinfo,        1, 0, 0, comdb2CtxinfoFunc),
+    FUNCTION(comdb2_version,        0, 0, 0, comdb2VersionFunc),
+    FUNCTION(comdb2_semver,         0, 0, 0, comdb2SemVerFunc),
+    FUNCTION(table_version,         1, 0, 0, tableVersionFunc),
+    FUNCTION(partition_info,        2, 0, 0, partitionInfoFunc),
+    FUNCTION(comdb2_host,           0, 0, 0, comdb2HostFunc),
+    FUNCTION(comdb2_node,           0, 0, 0, comdb2HostFunc),
+    FUNCTION(comdb2_port,           0, 0, 0, comdb2PortFunc),
+    FUNCTION(comdb2_dbname,         0, 0, 0, comdb2DbnameFunc),
+    FUNCTION(comdb2_prevquerycost,  0, 0, 0, comdb2PrevquerycostFunc),
+    FUNCTION(comdb2_starttime,      0, 0, 0, comdb2StartTimeFunc),
+    FUNCTION(comdb2_user,           0, 0, 0, comdb2UserFunc),
+    FUNCTION(comdb2_last_cost,      0, 0, 0, comdb2LastCostFunc),
+    FUNCTION(comdb2_snapshot_lsn,   0, 0, 0, comdb2SnapshotLsnFunc),
+    FUNCTION(checksum_md5,          1, 0, 0, md5Func),
+    FUNCTION(compress,              1, 0, 0, compressFunc),
+    FUNCTION(uncompress,            1, 0, 0, uncompressFunc),
+    FUNCTION(compress_zlib,         1, 0, 0, compressFunc),
+    FUNCTION(uncompress_zlib,       1, 0, 0, uncompressFunc),
+    FUNCTION(compress_gzip,         1, 0, 0, compressGzipFunc),
+    FUNCTION(uncompress_gzip,       1, 0, 0, uncompressGzipFunc),
+#if defined(SQLITE_BUILDING_FOR_COMDB2_DBGLOG)
+    FUNCTION(dbglog_cookie,         0, 0, 0, dbglogCookieFunc),
+    FUNCTION(dbglog_begin,          1, 0, 0, dbglogBeginFunc),
+    FUNCTION(dbglog_end,            1, 0, 0, dbglogEndFunc),
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2_DBGLOG) */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifdef SQLITE_ENABLE_UNKNOWN_SQL_FUNCTION
     FUNCTION(unknown,           -1, 0, 0, unknownFunc      ),
 #endif
@@ -1999,7 +3113,9 @@
     FUNCTION2(coalesce,         -1, 0, 0, noopFunc,  SQLITE_FUNC_COALESCE),
   };
 #ifndef SQLITE_OMIT_ALTERTABLE
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   sqlite3AlterFunctions();
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
   sqlite3WindowFunctions();
 #if defined(SQLITE_ENABLE_STAT3) || defined(SQLITE_ENABLE_STAT4)
@@ -2024,3 +3140,41 @@
   }
 #endif
 }
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+static int strptrcmp(const void *p1, const void *p2) {
+      return strcmp(*(char *const *)p1, *(char *const *)p2);
+}
+
+void sqlite3GetAllBuiltinFunctions(void **data, int *tot)
+{
+    int cnt = 100;
+    int num = 0;
+    char **arr = malloc((cnt * sizeof(char *)));
+    int i;
+    FuncDef *p;
+    for(i=0; i<SQLITE_FUNC_HASH_SZ; i++){
+        for(p=sqlite3BuiltinFunctions.a[i]; p; p=p->u.pHash){
+            if (num >= cnt) {
+                cnt *= 2;
+                arr = realloc(arr, cnt * sizeof(char*));
+            }
+            arr[num++] = sqlite3_mprintf("%s()", p->zName);
+        }
+    }
+    arr[num++] = sqlite3_mprintf("sys.cmd.send()");
+    arr[num++] = sqlite3_mprintf("sys.cmd.verify()");
+    qsort(arr, num, sizeof(char *), strptrcmp);
+    *tot = num;
+    *data = arr;
+}
+
+void sqlite3FreeAllBuiltinFunctions(void **data, int tot)
+{
+    char **arr = (char **)data;
+    for (int i = 0; i < tot; i++) {
+        sqlite3_free(arr[i]);
+    }
+    free(arr);
+}
+#endif
