--- sqlite-3.28.0/src/global.c	2026-01-09 16:11:01
+++ src/global.c	2025-09-24 15:21:58
@@ -221,7 +221,11 @@
    SQLITE_MAX_MMAP_SIZE,      /* mxMmap */
    (void*)0,                  /* pPage */
    0,                         /* szPage */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+   0,                         /* nPage */
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
    SQLITE_DEFAULT_PCACHE_INITSZ, /* nPage */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
    0,                         /* mxParserStack */
    0,                         /* sharedCacheEnabled */
    SQLITE_SORTER_PMASZ,       /* szPma */
