--- sqlite-3.28.0/src/loadext.c	2026-01-09 16:11:01
+++ src/loadext.c	2025-09-24 15:21:58
@@ -204,7 +204,11 @@
   sqlite3_get_table,
   0,     /* Was sqlite3_global_recover(), but that function is deprecated */
   sqlite3_interrupt,
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  0,
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   sqlite3_last_insert_rowid,
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   sqlite3_libversion,
   sqlite3_libversion_number,
   sqlite3_malloc,
@@ -342,11 +346,19 @@
   /*
   ** Added for 3.7.4
   */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  0,
+  0,
+  0,
+  0,
+  0,
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   sqlite3_backup_finish,
   sqlite3_backup_init,
   sqlite3_backup_pagecount,
   sqlite3_backup_remaining,
   sqlite3_backup_step,
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifndef SQLITE_OMIT_COMPILEOPTION_DIAGS
   sqlite3_compileoption_get,
   sqlite3_compileoption_used,
