--- sqlite-3.28.0/src/main.c	2026-01-09 16:11:01
+++ src/main.c	2025-09-24 15:21:58
@@ -15,6 +15,9 @@
 ** accessed by users of the library.
 */
 #include "sqliteInt.h"
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <sqlglue.h>
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 #ifdef SQLITE_ENABLE_FTS3
 # include "fts3.h"
@@ -25,6 +28,14 @@
 #if defined(SQLITE_ENABLE_ICU) || defined(SQLITE_ENABLE_ICU_COLLATIONS)
 # include "sqliteicu.h"
 #endif
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+# include "comdb2systbl.h"
+# ifdef SQLITE_ENABLE_SERIES
+#  include "series.h"
+# endif
+# include <logmsg.h>
+# include <stddef.h>
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifdef SQLITE_ENABLE_JSON1
 int sqlite3Json1Init(sqlite3*);
 #endif
@@ -34,6 +45,9 @@
 #ifdef SQLITE_ENABLE_FTS5
 int sqlite3Fts5Init(sqlite3*);
 #endif
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+int sqlite3RegexpInit(sqlite3*);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 #ifndef SQLITE_AMALGAMATION
 /* IMPLEMENTATION-OF: R-46656-45156 The sqlite3_version[] string constant
@@ -105,6 +119,16 @@
 ** See also the "PRAGMA data_store_directory" SQL command.
 */
 char *sqlite3_data_directory = 0;
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+void sqlite3PCacheBufferSetup(void *p, int sz, int n){ return; }
+int sqlite3PcacheInitialize(void){ return 0; }
+void sqlite3PcacheShutdown(void){ return; }
+void sqlite3PCacheSetDefault(void){ return; }
+int sqlite3HeaderSizeBtree(void){ return 0; }
+int sqlite3HeaderSizePcache(void){ return 0; }
+int sqlite3HeaderSizePcache1(void){ return 0; }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Initialize SQLite.  
@@ -952,7 +976,20 @@
   return r;
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 /*
+** Dummy collating sequence used by datacopy index
+*/
+static int datacopyCollatingFunc(
+  void *NotUsed,
+  int nKey1, const void *pKey1,
+  int nKey2, const void *pKey2
+){
+  return 0;
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
+/*
 ** Return the ROWID of the most recent insert
 */
 sqlite_int64 sqlite3_last_insert_rowid(sqlite3 *db){
@@ -1118,8 +1155,21 @@
   ** SQLITE_BUSY if the connection can not be closed immediately.
   */
   if( !forceZombie && connectionIsBusy(db) ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_ERROR, "%s:%d SQLITE_BUSY\n", __FILE__, __LINE__);
+    cheap_stack_trace();
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3ErrorWithMsg(db, SQLITE_BUSY, "unable to close due to unfinalized "
        "statements or unfinished backups");
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    {
+      sqlite3_stmt *pStmt = 0;
+      while( (pStmt = sqlite3_next_stmt(db,pStmt))!=0 ){
+        logmsg(LOGMSG_DEBUG, "%s:%d NOT FINALIZED: %p ==> %p {%s}\n",
+               __FILE__, __LINE__, db, pStmt, sqlite3_sql(pStmt));
+      }
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3_mutex_leave(db->mutex);
     return SQLITE_BUSY;
   }
@@ -1478,6 +1528,29 @@
     /* SQLITE_NOTICE      */ "notification message",
     /* SQLITE_WARNING     */ "warning message",
   };
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  static const char* const aComdb2Msg[] = {
+    /* SQLITE_DEADLOCK          */ "transaction aborted due to deadlock",
+    /* SQLITE_ACCESS            */ "access denied",
+    /* SQLITE_LIMIT_DEPRECATED  */ 0,
+    /* 203, unused              */ 0,
+    /* SQLITE_TRANTOOCOMPLEX    */ "Transaction rollback too large",
+    /* SQLITE_TRAN_CANCELLED    */ "Unable to maintain snapshot, too many resources blocked",
+    /* SQLITE_TRAN_NOLOG        */ "Unable to maintain snapshot, too many log files",
+    /* SQLITE_TRAN_NOUNDO       */ "Database changed due to sc or fastinit; snapshot failure",
+    /* SQLITE_CONV_ERROR        */ "type conversion failure",
+    /* SQLITE_COMDB2SCHEMA      */ "table schema changed",
+    /* SQLITE_CLIENT_CHANGENODE */ "Non-durable write",
+    /* SQLITE_DDL_MISUSE        */ "overlapping tables detected in transactional DDL",
+    /* SQLITE_TIMEDOUT          */ "query timed out",
+    /* SQLITE_COST_TOO_HIGH     */ "query cost too high",
+    /* SQLITE_NO_TEMPTABLES     */ "temporary tables disallowed",
+    /* SQLITE_NO_TABLESCANS     */ "table scans disallowed",
+  };
+  static const int comdb2MinErr = SQLITE_DEADLOCK;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
   const char *zErr = "unknown error";
   switch( rc ){
     case SQLITE_ABORT_ROLLBACK: {
@@ -1497,6 +1570,14 @@
       if( ALWAYS(rc>=0) && rc<ArraySize(aMsg) && aMsg[rc]!=0 ){
         zErr = aMsg[rc];
       }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      else if( rc>=comdb2MinErr ){
+        rc -= comdb2MinErr;
+        if( rc<ArraySize(aComdb2Msg) && aComdb2Msg[rc]!=0 ){
+          zErr = aComdb2Msg[rc];
+        }
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       break;
     }
   }
@@ -1762,6 +1843,10 @@
   p = sqlite3FindFunction(db, zFunctionName, nArg, (u8)enc, 0);
   if( p && (p->funcFlags & SQLITE_FUNC_ENCMASK)==(u32)enc && p->nArg==nArg ){
     if( db->nVdbeActive ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      logmsg(LOGMSG_ERROR, "%s:%d SQLITE_BUSY\n", __FILE__, __LINE__);
+      cheap_stack_trace();
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3ErrorWithMsg(db, SQLITE_BUSY, 
         "unable to delete/modify user-function due to active statements");
       assert( !db->mallocFailed );
@@ -2540,6 +2625,10 @@
   pColl = sqlite3FindCollSeq(db, (u8)enc2, zName, 0);
   if( pColl && pColl->xCmp ){
     if( db->nVdbeActive ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      logmsg(LOGMSG_ERROR, "%s:%d SQLITE_BUSY\n", __FILE__, __LINE__);
+      cheap_stack_trace();
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3ErrorWithMsg(db, SQLITE_BUSY, 
         "unable to delete/modify collation sequence due to active statements");
       return SQLITE_BUSY;
@@ -2942,8 +3031,39 @@
   *pFlags = flags;
   *pzFile = zFile;
   return rc;
+}
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+static void register_lua_sfuncs(sqlite3 *db, struct sqlthdstate *thd)
+{
+    listc_t sfuncs;
+    get_sfuncs(&sfuncs);
+    struct lua_func_t *sfunc;
+    LISTC_FOR_EACH(&sfuncs, sfunc, lnk)
+    {
+        lua_func_arg_t *arg = malloc(sizeof(lua_func_arg_t));
+        arg->thd = thd;
+        arg->name = sfunc->name;
+        sqlite3_create_function_v2(db, sfunc->name, -1, SQLITE_UTF8 | sfunc->flags, arg, lua_func, NULL, NULL, free);
+    }
 }
 
+static void register_lua_afuncs(sqlite3 *db, struct sqlthdstate *thd)
+{
+    listc_t afuncs;
+    get_afuncs(&afuncs);
+    struct lua_func_t *afunc;
+    LISTC_FOR_EACH(&afuncs, afunc, lnk)
+    {
+        lua_func_arg_t *arg = malloc(sizeof(lua_func_arg_t));
+        arg->thd = thd;
+        arg->name = afunc->name;
+        // afunc->flags are purposefully ignored here.
+        sqlite3_create_function_v2(db, afunc->name, -1, SQLITE_UTF8, arg, NULL, lua_step, lua_final, free);
+    }
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 #if defined(SQLITE_HAS_CODEC)
 /*
 ** Process URI filename query parameters relevant to the SQLite Encryption
@@ -2989,6 +3109,9 @@
   sqlite3 **ppDb,        /* OUT: Returned database handle */
   unsigned int flags,    /* Operational flags */
   const char *zVfs       /* Name of the VFS to use */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  ,struct sqlthdstate *thd
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 ){
   sqlite3 *db;                    /* Store allocated handle here */
   int rc;                         /* Return code */
@@ -3063,6 +3186,10 @@
     }
   }
   sqlite3_mutex_enter(db->mutex);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  db->isPreparer = 0;
+  assert(offsetof(sqlite3, isPreparer) == 0);
+#endif
   db->errMask = 0xff;
   db->nDb = 2;
   db->magic = SQLITE_MAGIC_BUSY;
@@ -3138,6 +3265,9 @@
   db->pDfltColl = sqlite3FindCollSeq(db, SQLITE_UTF8, sqlite3StrBINARY, 0);
   assert( db->pDfltColl!=0 );
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  createCollation(db, "DATACOPY", SQLITE_UTF8, 0, datacopyCollatingFunc, 0);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   /* Parse the filename/URI argument
   **
   ** Only allow sensible combinations of bits in the flags argument.  
@@ -3270,6 +3400,19 @@
   }
 #endif
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( !db->mallocFailed && rc==SQLITE_OK ){
+    rc = comdb2SystblInit(db);
+  }
+  if( !db->mallocFailed && rc==SQLITE_OK){
+    rc = sqlite3RegexpInit(db);
+  }
+#ifdef SQLITE_ENABLE_SERIES
+  if( !db->mallocFailed && rc==SQLITE_OK ){
+    rc = sqlite3SeriesInit(db);
+  }
+#endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifdef SQLITE_ENABLE_JSON1
   if( !db->mallocFailed && rc==SQLITE_OK){
     rc = sqlite3Json1Init(db);
@@ -3322,6 +3465,18 @@
     sqlite3GlobalConfig.xSqllog(pArg, db, zFilename, 0);
   }
 #endif
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+  pthread_mutex_lock(&mutex);
+  /* these modify global structures */
+  if( thd!=NULL ){
+    register_lua_sfuncs(db, thd);
+    register_lua_afuncs(db, thd);
+  }
+  register_date_functions(db);
+  pthread_mutex_unlock(&mutex);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #if defined(SQLITE_HAS_CODEC)
   if( rc==SQLITE_OK ) sqlite3CodecQueryParameters(db, 0, zOpen);
 #endif
@@ -3336,9 +3491,17 @@
 int sqlite3_open(
   const char *zFilename, 
   sqlite3 **ppDb 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  , struct sqlthdstate *thd
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
   return openDatabase(zFilename, ppDb,
+                    SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, 0, thd);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+  return openDatabase(zFilename, ppDb,
                       SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, 0);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 int sqlite3_open_v2(
   const char *filename,   /* Database filename (UTF-8) */
@@ -3346,7 +3509,11 @@
   int flags,              /* Flags */
   const char *zVfs        /* Name of VFS module to use */
 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  return openDatabase(filename, ppDb, (unsigned int)flags, zVfs, NULL);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   return openDatabase(filename, ppDb, (unsigned int)flags, zVfs);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
 #ifndef SQLITE_OMIT_UTF16
@@ -3374,8 +3541,13 @@
   sqlite3ValueSetStr(pVal, -1, zFilename, SQLITE_UTF16NATIVE, SQLITE_STATIC);
   zFilename8 = sqlite3ValueText(pVal, SQLITE_UTF8);
   if( zFilename8 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    rc = openDatabase(zFilename8, ppDb,
+                      SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, 0, 0);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     rc = openDatabase(zFilename8, ppDb,
                       SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE, 0);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( *ppDb || rc==SQLITE_NOMEM );
     if( rc==SQLITE_OK && !DbHasProperty(*ppDb, 0, DB_SchemaLoaded) ){
       SCHEMA_ENC(*ppDb) = ENC(*ppDb) = SQLITE_UTF16NATIVE;
@@ -3700,6 +3872,99 @@
   return rc;
 }
 
+#ifdef SQLITE_BUILDING_FOR_COMDB2
+/**
+ * Returns the information about application defined functions used in indexes
+ * by the specified table.
+ */
+// TODO: Maybe better if we create an Index -> uFuncs (used functions by the index)
+//      in sqlite3CreateIndex so that we don't have to do this dance again?
+int sqlite3_table_index_funcs(sqlite3 *db,            /* Connection handle */
+                              const char *zDbName,    /* Database name or NULL */
+                              const char *zTableName, /* Table name */
+                              char ***pzFuncs,        /* OUTPUT: The application defined functions */
+                              int *nFuncs)
+{                                  /* OUTPUT: The number of application defined functions used */
+  Table *pTab = 0;                 /* Table corresponding to zTableName*/
+  char **aNew = NULL;              /* Array of functions */
+  int nNew = 0;                    /* Number of a functions discovered*/
+  int dNew = 10;                   /* Array increment for aNew*/
+  int rc = 0;                      /* Return code */
+  int add = 0;                     /* If new function discovered */
+  const char *zFunc = 0;           /* Function used by the index */
+  char *zErrMsg = 0;               /* Error message */
+  struct ExprList_item *pListItem; /* For looping over pList */
+
+  sqlite3_mutex_enter(db->mutex);
+  sqlite3BtreeEnterAll(db);
+  rc = sqlite3Init(db, &zErrMsg);
+  if (SQLITE_OK != rc) {
+    goto error_out;
+  }
+
+  pTab = sqlite3FindTable(db, zTableName, zDbName);
+  if (!pTab || pTab->pSelect) {
+    pTab = 0;
+    goto error_out;
+  }
+
+  /*
+  ** Go over all the indexes in the table. Find the non-column expression
+  ** in aiColumn. Check if the expression corresponds to a function name.
+  ** Add the function name to aNew if we haven't already.
+  **
+  ** TODO: This isn't the most efficient. Can add something like pTab->hasFuncIdx
+  **       and skip over the tables which won't have functions in indices.
+  */
+  Index *pIndex = pTab->hasExprIdx ? pTab->pIndex : 0;
+  while (pIndex) {
+    pListItem = pIndex->aColExpr->a;
+    for (int j = 0; j < pIndex->nKeyCol; ++j, pListItem++) {
+      if (pIndex->aiColumn[j] == XN_EXPR) {
+        Expr *pExpr = sqlite3ExprSkipCollate(pListItem->pExpr);
+        if (pExpr && (pExpr->op == TK_FUNCTION)) {
+          add = 1;
+          zFunc = pExpr->u.zToken;
+          for (int i = 0; i < nNew; ++i) {
+            if (strcmp(aNew[i], zFunc) == 0) {
+              add = 0;
+            }
+          }
+          if (add) {
+            if (nNew % dNew == 0) {
+              aNew = realloc(aNew, (255) * (nNew + dNew));
+              if (!aNew) {
+                rc = SQLITE_NOMEM_BKPT;
+                goto error_out;
+              }
+            }
+            aNew[nNew++] = strndup(pExpr->u.zToken, 255);
+          }
+        }
+      }
+    }
+    pIndex = pIndex->pNext;
+  }
+
+error_out:
+  sqlite3BtreeLeaveAll(db);
+
+  if ( pzFuncs ) *pzFuncs = aNew;
+  if ( nFuncs ) *nFuncs = nNew;
+
+  if ( SQLITE_OK == rc && !pTab ) {
+    sqlite3DbFree(db, zErrMsg);
+    zErrMsg = sqlite3MPrintf(db, "no such table: %s", zTableName);
+    rc = SQLITE_ERROR;
+  }
+  sqlite3ErrorWithMsg(db, rc, (zErrMsg ? "%s" : 0), zErrMsg);
+  sqlite3DbFree(db, zErrMsg);
+  rc = sqlite3ApiExit(db, rc);
+  sqlite3_mutex_leave(db->mutex);
+  return rc;
+}
+#endif
+
 /*
 ** Sleep for a little while.  Return the amount of time slept.
 */
@@ -3746,6 +4011,9 @@
     sqlite3_file *fd;
     sqlite3BtreeEnter(pBtree);
     pPager = sqlite3BtreePager(pBtree);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( pPager ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( pPager!=0 );
     fd = sqlite3PagerFile(pPager);
     assert( fd!=0 );
@@ -3763,7 +4031,12 @@
       rc = SQLITE_OK;
     }else{
       rc = sqlite3OsFileControl(fd, op, pArg);
+    }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    } else {
+      rc = SQLITE_NOTFOUND;
     }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3BtreeLeave(pBtree);
   }
   sqlite3_mutex_leave(db->mutex);
@@ -4382,3 +4655,9 @@
   return 0;
 }
 #endif /* SQLITE_OMIT_COMPILEOPTION_DIAGS */
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+int sqlite3_is_preparer(sqlite3 *db) {
+    return db->isPreparer;
+}
+#endif
