--- sqlite-3.28.0/src/os_unix.c	2026-01-09 16:11:01
+++ src/os_unix.c	2025-09-24 15:21:58
@@ -46,6 +46,12 @@
 #include "sqliteInt.h"
 #if SQLITE_OS_UNIX              /* This file is used on unix only */
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#ifdef SQLITE_UNIX_THREADS
+#include <pthread.h>
+#endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** There are various methods for file locking used for concurrency
 ** control:
@@ -141,6 +147,9 @@
 #if SQLITE_THREADSAFE
 # include <pthread.h>
 #endif
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <logmsg.h>
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Default permissions when creating a new file
@@ -1715,6 +1724,10 @@
   if( (pFile->eFileLock!=pInode->eFileLock && 
           (pInode->eFileLock>=PENDING_LOCK || eFileLock>SHARED_LOCK))
   ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_ERROR, "%s:%d SQLITE_BUSY\n", __FILE__, __LINE__);
+    cheap_stack_trace();
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     rc = SQLITE_BUSY;
     goto end_lock;
   }
@@ -1752,6 +1765,10 @@
       if( rc!=SQLITE_BUSY ){
         storeLastErrno(pFile, tErrno);
       }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      logmsg(LOGMSG_ERROR, "%s:%d SQLITE_BUSY\n", __FILE__, __LINE__);
+      cheap_stack_trace();
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       goto end_lock;
     }
   }
@@ -1781,6 +1798,10 @@
       /* This could happen with a network mount */
       tErrno = errno;
       rc = SQLITE_IOERR_UNLOCK; 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      logmsg(LOGMSG_ERROR, "%s:%d SQLITE_BUSY\n", __FILE__, __LINE__);
+      cheap_stack_trace();
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
 
     if( rc ){
@@ -1796,6 +1817,10 @@
   }else if( eFileLock==EXCLUSIVE_LOCK && pInode->nShared>1 ){
     /* We are trying for an exclusive lock but another thread in this
     ** same process is still holding a shared lock. */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_ERROR, "%s:%d SQLITE_BUSY\n", __FILE__, __LINE__);
+    cheap_stack_trace();
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     rc = SQLITE_BUSY;
   }else{
     /* The request was for a RESERVED or EXCLUSIVE lock.  It is
@@ -1820,6 +1845,12 @@
       if( rc!=SQLITE_BUSY ){
         storeLastErrno(pFile, tErrno);
       }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( rc==SQLITE_BUSY ){
+        logmsg(LOGMSG_ERROR, "%s:%d SQLITE_BUSY\n", __FILE__, __LINE__);
+        cheap_stack_trace();
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
   }
   
@@ -4180,13 +4211,25 @@
 ** Instead, it should be called via macro osGetpagesize().
 */
 static int unixGetpagesize(void){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 #if OS_VXWORKS
   return 1024;
 #elif defined(_BSD_SOURCE)
   return getpagesize();
+#elif defined(_DEFAULT_SOURCE)
+  return getpagesize();
 #else
   return (int)sysconf(_SC_PAGESIZE);
 #endif
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+#if OS_VXWORKS
+  return 1024;
+#elif defined(_BSD_SOURCE)
+  return getpagesize();
+#else
+  return (int)sysconf(_SC_PAGESIZE);
+#endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
 #endif /* !defined(SQLITE_OMIT_WAL) || SQLITE_MAX_MMAP_SIZE>0 */
@@ -5647,12 +5690,19 @@
   }
   return rc;
 }
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+char *comdb2_get_tmp_dir(void);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Return the name of a directory in which to put temporary files.
 ** If no suitable temporary file directory can be found, return NULL.
 */
 static const char *unixTempFileDir(void){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+   return comdb2_get_tmp_dir();
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   static const char *azDirs[] = {
      0,
      0,
@@ -5679,6 +5729,7 @@
     zDir = azDirs[i++];
   }
   return 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
 /*
