--- sqlite-3.28.0/src/printf.c	2026-01-09 16:11:01
+++ src/printf.c	2025-09-24 15:21:58
@@ -897,6 +897,24 @@
       /* Force exponential buffer size growth as long as it does not overflow,
       ** to avoid having to call this routine too often */
       szNew += p->nChar;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    } else {
+      /* It is now impossible to grow the buffer further exponentially.
+      ** However, we could still have some room left to grow during this last
+      ** leg of expansion. We do that by setting new size to maximum allowed
+      ** size. One more thing to notice is that in the previous iteration, the
+      ** sqlite memory allocator could have allocated more that the requested
+      ** size (due to roundup) and the current buffer size (p->nAlloc) could
+      ** have already exceeded the maximum allowed limit (p->mxAlloc). Let us
+      ** bail out if we are already past the limit.
+      **/
+      if( p->nAlloc >= p->mxAlloc ){
+        sqlite3_str_reset(p);
+        setStrAccumError(p, SQLITE_TOOBIG);
+        return 0;
+      }
+      szNew = p->mxAlloc;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
     if( szNew > p->mxAlloc ){
       sqlite3_str_reset(p);
@@ -1241,6 +1259,7 @@
   }
 }
 
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
 #if defined(SQLITE_DEBUG) || defined(SQLITE_HAVE_OS_TRACE)
 /*
 ** A version of printf() that understands %lld.  Used for debugging.
@@ -1267,6 +1286,7 @@
 #endif
 }
 #endif
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 
 /*
