--- sqlite-3.28.0/src/random.c	2026-01-09 16:11:01
+++ src/random.c	2025-09-24 15:21:58
@@ -21,7 +21,11 @@
 /* All threads share a single random number generator.
 ** This structure is the current state of the generator.
 */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+static __thread SQLITE_WSD struct sqlite3PrngType {
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 static SQLITE_WSD struct sqlite3PrngType {
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   unsigned char isInit;          /* True if initialized */
   unsigned char i, j;            /* State variables */
   unsigned char s[256];          /* State variables */
