--- sqlite-3.28.0/src/resolve.c	2026-01-09 16:11:01
+++ src/resolve.c	2025-09-24 15:21:58
@@ -15,6 +15,13 @@
 ** table and column.
 */
 #include "sqliteInt.h"
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+extern int gbl_strict_dbl_quotes;
+int sqlite3IsComdb2Rowid(Table *pTab, const char *);
+int sqlite3IsComdb2RowTimestamp(Table *pTab, const char *);
+int is_comdb2_index_blob(const char *dbname, int icol);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Walk the expression tree pExpr and increase the aggregate function
@@ -262,9 +269,11 @@
           if( sqlite3StrICmp(zTabName, zTab)!=0 ){
             continue;
           }
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
           if( IN_RENAME_OBJECT && pItem->zAlias ){
             sqlite3RenameTokenRemap(pParse, 0, (void*)&pExpr->y.pTab);
           }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
         }
         if( 0==(cntTab++) ){
           pMatch = pItem;
@@ -398,7 +407,34 @@
       pExpr->iColumn = -1;
       pExpr->affinity = SQLITE_AFF_INTEGER;
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /*
+    ** if it's a "comdb2_rowid" set a magic -2 for the column (instead of
+    ** the usual magic -1 for a "rowid."  inside of sqlite3exprcode we sniff
+    ** this out, and when we see a -2, we add a properly formulated
+    ** instruction (OP_Rowid with P3==1 (instead of the usual 0).
+    ** the vm sniffs this out when it runs OP_Rowid and executes the comdb2
+    ** backend call to get the rrn+genid.
+    */
+    else if( cnt==0 && cntTab==1 && pMatch && sqlite3IsComdb2Rowid(pMatch->pTab, zCol) ){
+       cnt = 1;
+       pExpr->iColumn = -2;
+       pExpr->affinity = SQLITE_AFF_TEXT;
+    }else if( cnt==0 && cntTab==1 && pMatch && sqlite3IsComdb2RowTimestamp(pMatch->pTab, zCol) ){
+       cnt = 1;
+       pExpr->iColumn = -3;
+       pExpr->affinity = SQLITE_AFF_TEXT;
+    }
 
+    /* Check if a partial index or an expression index contains blob fields. */
+    if( (pNC->ncFlags & (NC_PartIdx|NC_IdxExpr))!=0
+     && pExpr->y.pTab
+     && pExpr->iColumn>=0
+    ){
+      is_comdb2_index_blob(pExpr->y.pTab->zName, pExpr->iColumn);
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
     /*
     ** If the input is of the form Z (not Y.Z or X.Y.Z) then the name Z
     ** might refer to an result-set alias.  This happens, for example, when
@@ -447,9 +483,11 @@
           cnt = 1;
           pMatch = 0;
           assert( zTab==0 && zDb==0 );
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
           if( IN_RENAME_OBJECT ){
             sqlite3RenameTokenRemap(pParse, 0, (void*)pExpr);
           }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
           goto lookupname_end;
         }
       } 
@@ -477,6 +515,12 @@
   if( cnt==0 && zTab==0 ){
     assert( pExpr->op==TK_ID );
     if( ExprHasProperty(pExpr,EP_DblQuoted) ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( gbl_strict_dbl_quotes ){
+        sqlite3ErrorMsg(pParse, "double-quoted string literal: \"%w\"", zCol);
+        return WRC_Abort;
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       /* If a double-quoted identifier does not match any known column name,
       ** then treat it as a string.
       **
@@ -666,6 +710,9 @@
       assert( pSrcList && pSrcList->nSrc==1 );
       pItem = pSrcList->a;
       assert( HasRowid(pItem->pTab) && pItem->pTab->pSelect==0 );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      assert( HasRowid(pItem->pTab) && pItem->pTab->pSelect==0);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       pExpr->op = TK_COLUMN;
       pExpr->y.pTab = pItem->pTab;
       pExpr->iTable = pItem->iCursor;
@@ -709,10 +756,12 @@
         }
         zTable = pLeft->u.zToken;
         zColumn = pRight->u.zToken;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
         if( IN_RENAME_OBJECT ){
           sqlite3RenameTokenRemap(pParse, (void*)pExpr, (void*)pRight);
           sqlite3RenameTokenRemap(pParse, (void*)&pExpr->y.pTab, (void*)pLeft);
         }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
       }
       return lookupName(pParse, zDb, zTable, zColumn, pNC, pExpr);
     }
@@ -845,6 +894,11 @@
         ){
           sqlite3ErrorMsg(pParse, "no such function: %.*s", nId, zId);
           pNC->nErr++;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        }else if( no_such_func && (pNC->ncFlags & (NC_IdxExpr|NC_PartIdx)) ){
+          sqlite3ErrorMsg(pParse, "no such function: %.*s", nId, zId);
+          pNC->nErr++;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         }else if( wrong_num_args ){
           sqlite3ErrorMsg(pParse,"wrong number of arguments to function %.*s()",
                nId, zId);
@@ -866,15 +920,22 @@
 #ifndef SQLITE_OMIT_WINDOWFUNC
         if( pExpr->y.pWin ){
           Select *pSel = pNC->pWinSelect;
-          sqlite3WindowUpdate(pParse, pSel->pWinDefn, pExpr->y.pWin, pDef);
-          sqlite3WalkExprList(pWalker, pExpr->y.pWin->pPartition);
-          sqlite3WalkExprList(pWalker, pExpr->y.pWin->pOrderBy);
-          sqlite3WalkExpr(pWalker, pExpr->y.pWin->pFilter);
+          Window *pWin = pExpr->y.pWin;
+          if( IN_RENAME_OBJECT==0 ){
+            sqlite3WindowUpdate(pParse, pSel->pWinDefn, pWin, pDef);
+          }
+          sqlite3WalkExprList(pWalker, pWin->pPartition);
+          sqlite3WalkExprList(pWalker, pWin->pOrderBy);
+          sqlite3WalkExpr(pWalker, pWin->pFilter);
           if( 0==pSel->pWin 
            || 0==sqlite3WindowCompare(pParse, pSel->pWin, pExpr->y.pWin) 
           ){
-            pExpr->y.pWin->pNextWin = pSel->pWin;
-            pSel->pWin = pExpr->y.pWin;
+            pWin->pNextWin = pSel->pWin;
+            if( pSel->pWin ){
+              pSel->pWin->ppThis = &pWin->pNextWin;
+            }
+            pSel->pWin = pWin;
+            pWin->ppThis = &pSel->pWin;
           }
           pNC->ncFlags |= NC_HasWin;
         }else
@@ -1042,7 +1103,7 @@
   int rc;            /* Return code from subprocedures */
   u8 savedSuppErr;   /* Saved value of db->suppressErr */
 
-  assert( sqlite3ExprIsInteger(pE, &i)==0 );
+  assert( sqlite3ExprIsInteger(pE, &i, 0)==0 );
   pEList = pSelect->pEList;
 
   /* Resolve all names in the ORDER BY term expression
@@ -1138,7 +1199,7 @@
       Expr *pE, *pDup;
       if( pItem->done ) continue;
       pE = sqlite3ExprSkipCollate(pItem->pExpr);
-      if( sqlite3ExprIsInteger(pE, &iCol) ){
+      if( sqlite3ExprIsInteger(pE, &iCol, 0) ){
         if( iCol<=0 || iCol>pEList->nExpr ){
           resolveOutOfRangeError(pParse, "ORDER", i+1, pEList->nExpr);
           return 1;
@@ -1253,16 +1314,19 @@
 
 #ifndef SQLITE_OMIT_WINDOWFUNC
 /*
-** Walker callback for resolveRemoveWindows().
+** Walker callback for sqlite3WindowRemoveExprFromSelect() and
+** sqlite3WindowRemoveExprListFromSelect()
 */
 static int resolveRemoveWindowsCb(Walker *pWalker, Expr *pExpr){
   if( ExprHasProperty(pExpr, EP_WinFunc) ){
-    Window **pp;
-    for(pp=&pWalker->u.pSelect->pWin; *pp; pp=&(*pp)->pNextWin){
-      if( *pp==pExpr->y.pWin ){
-        *pp = (*pp)->pNextWin;
-        break;
-      }    
+    Window *pWin = pExpr->y.pWin;
+
+    if( pWin->ppThis ){
+      *pWin->ppThis = pWin->pNextWin;
+
+
+      if( pWin->pNextWin ) pWin->pNextWin->ppThis = pWin->ppThis;
+      pWin->ppThis = 0;
     }
   }
   return WRC_Continue;
@@ -1272,16 +1336,33 @@
 ** Remove any Window objects owned by the expression pExpr from the
 ** Select.pWin list of Select object pSelect.
 */
-static void resolveRemoveWindows(Select *pSelect, Expr *pExpr){
-  Walker sWalker;
-  memset(&sWalker, 0, sizeof(Walker));
-  sWalker.xExprCallback = resolveRemoveWindowsCb;
-  sWalker.u.pSelect = pSelect;
-  sqlite3WalkExpr(&sWalker, pExpr);
+void sqlite3WindowRemoveExprFromSelect(Select *pSelect, Expr *pExpr){
+  if( pSelect->pWin ){
+    Walker sWalker;
+    memset(&sWalker, 0, sizeof(Walker));
+    sWalker.xExprCallback = resolveRemoveWindowsCb;
+    sWalker.u.pSelect = pSelect;
+    sqlite3WalkExpr(&sWalker, pExpr);
+  }
 }
-#else
-# define resolveRemoveWindows(x,y)
-#endif
+
+/*
+** Remove any Window objects owned by the expression list from the
+** Select.pWin list of Select object pSelect.
+*/
+void sqlite3WindowRemoveExprListFromSelect(Select *pSelect, ExprList *pList){
+  if( pList && pSelect->pWin ){
+    int i;
+    Walker sWalker;
+    memset(&sWalker, 0, sizeof(Walker));
+    sWalker.xExprCallback = resolveRemoveWindowsCb;
+    sWalker.u.pSelect = pSelect;
+    for(i=0; i<pList->nExpr; i++){
+      sqlite3WalkExpr(&sWalker, pList->a[i].pExpr);
+    }
+  }
+}
+#endif /* SQLITE_OMIT_WINDOWFUNC */
 
 /*
 ** pOrderBy is an ORDER BY or GROUP BY clause in SELECT statement pSelect.
@@ -1330,7 +1411,7 @@
         continue;
       }
     }
-    if( sqlite3ExprIsInteger(pE2, &iCol) ){
+    if( sqlite3ExprIsInteger(pE2, &iCol, 0) ){
       /* The ORDER BY term is an integer constant.  Again, set the column
       ** number so that sqlite3ResolveOrderGroupBy() will convert the
       ** order-by term to a copy of the result-set expression */
@@ -1352,7 +1433,7 @@
         /* Since this expresion is being changed into a reference
         ** to an identical expression in the result set, remove all Window
         ** objects belonging to the expression from the Select.pWin list. */
-        resolveRemoveWindows(pSelect, pE);
+        sqlite3WindowRemoveExprFromSelect(pSelect, pE);
         pItem->u.x.iOrderByCol = j+1;
       }
     }
