--- sqlite-3.28.0/src/select.c	2026-01-09 16:11:01
+++ src/select.c	2026-01-09 15:00:03
@@ -82,6 +82,24 @@
 };
 #define SORTFLAG_UseSorter  0x01   /* Use SorterOpen instead of OpenEphemeral */
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+extern int comdb2_register_limit(int, int);
+extern void comdb2_register_offset(int, int, int);
+extern const char *comdb2_get_dbname(void);
+extern void comdb2_set_verify_remote_schemas(void);
+
+static void _set_src_recording(
+  Parse *pParse,
+  Select *pSub
+){
+  int tbl;
+  pSub->recording = 1;
+  for(tbl=0; tbl<pSub->pSrc->nSrc; tbl++){
+    SET_CURSOR_RECORDING(pParse, pSub->pSrc->a[tbl].iCursor);
+  }
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Delete all the content of a Select structure.  Deallocate the structure
 ** itself only if bFree is true.
@@ -102,6 +120,7 @@
     }
 #endif
     if( OK_IF_ALWAYS_TRUE(p->pWith) ) sqlite3WithDelete(db, p->pWith);
+    assert( p->pWin==0 );
     if( bFree ) sqlite3DbFreeNN(db, p);
     p = pPrior;
     bFree = 1;
@@ -135,9 +154,9 @@
   u32 selFlags,         /* Flag parameters, such as SF_Distinct */
   Expr *pLimit          /* LIMIT value.  NULL means not used */
 ){
-  Select *pNew;
+  Select *pNew, *pAllocated;
   Select standin;
-  pNew = sqlite3DbMallocRawNN(pParse->db, sizeof(*pNew) );
+  pAllocated = pNew = sqlite3DbMallocRawNN(pParse->db, sizeof(*pNew) );
   if( pNew==0 ){
     assert( pParse->db->mallocFailed );
     pNew = &standin;
@@ -164,6 +183,9 @@
   pNew->pPrior = 0;
   pNew->pNext = 0;
   pNew->pLimit = pLimit;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  pNew->recording = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pNew->pWith = 0;
 #ifndef SQLITE_OMIT_WINDOWFUNC
   pNew->pWin = 0;
@@ -171,12 +193,11 @@
 #endif
   if( pParse->db->mallocFailed ) {
     clearSelect(pParse->db, pNew, pNew!=&standin);
-    pNew = 0;
+    pAllocated = 0;
   }else{
     assert( pNew->pSrc!=0 || pParse->nErr>0 );
   }
-  assert( pNew!=&standin );
-  return pNew;
+  return pAllocated;
 }
 
 
@@ -384,7 +405,7 @@
 ** after the t1 loop and rows with t1.x!=5 will never appear in
 ** the output, which is incorrect.
 */
-static void setJoinExpr(Expr *p, int iTable){
+void sqlite3SetJoinExpr(Expr *p, int iTable){
   while( p ){
     ExprSetProperty(p, EP_FromJoin);
     assert( !ExprHasProperty(p, EP_TokenOnly|EP_Reduced) );
@@ -393,15 +414,15 @@
     if( p->op==TK_FUNCTION && p->x.pList ){
       int i;
       for(i=0; i<p->x.pList->nExpr; i++){
-        setJoinExpr(p->x.pList->a[i].pExpr, iTable);
+        sqlite3SetJoinExpr(p->x.pList->a[i].pExpr, iTable);
       }
     }
-    setJoinExpr(p->pLeft, iTable);
+    sqlite3SetJoinExpr(p->pLeft, iTable);
     p = p->pRight;
   } 
 }
 
-/* Undo the work of setJoinExpr().  In the expression tree p, convert every
+/* Undo the work of sqlite3SetJoinExpr(). In the expression p, convert every
 ** term that is marked with EP_FromJoin and iRightJoinTable==iTable into
 ** an ordinary term that omits the EP_FromJoin mark.
 **
@@ -488,7 +509,7 @@
     ** an AND operator.
     */
     if( pRight->pOn ){
-      if( isOuter ) setJoinExpr(pRight->pOn, pRight->iCursor);
+      if( isOuter ) sqlite3SetJoinExpr(pRight->pOn, pRight->iCursor);
       p->pWhere = sqlite3ExprAnd(pParse->db, p->pWhere, pRight->pOn);
       pRight->pOn = 0;
     }
@@ -1551,6 +1572,11 @@
 #ifndef SQLITE_OMIT_SUBQUERY
     case SRT_Set: {
       assert( nColumn==sqlite3Strlen30(pDest->zAffSdst) );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      /* FIXME TODO XXX 
+       * This might be incorrect. prod has some hack to make the following
+       * use numeric type. The opcodes here have changed. Need to verify */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3VdbeAddOp4(v, OP_MakeRecord, regRow, nColumn, regRowid,
                         pDest->zAffSdst, nColumn);
       sqlite3VdbeAddOp4Int(v, OP_IdxInsert, iParm, regRowid, regRow, nColumn);
@@ -1641,10 +1667,13 @@
   char const *zOrigCol = 0;
 #endif
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* I need this "bug" to return "wrong" types for comdb2sql */
+  if( NEVER(pExpr==0) || pNC->pSrcList==0 ) return 0;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( pExpr!=0 );
   assert( pNC->pSrcList!=0 );
-  assert( pExpr->op!=TK_AGG_COLUMN );  /* This routine runes before aggregates
-                                       ** are processed */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   switch( pExpr->op ){
     case TK_COLUMN: {
       /* The expression is a column. Locate the table the column is being
@@ -1711,8 +1740,21 @@
         if( iCol<0 ) iCol = pTab->iPKey;
         assert( iCol==XN_ROWID || (iCol>=0 && iCol<pTab->nCol) );
         if( iCol<0 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          switch( iCol ){
+            default:
+              zType = "INTEGER";
+              zOrigCol = "rowid";
+              break;
+            case -3:
+              zType = "DATETIME";
+              zOrigCol = "comdb2_rowtimestamp";
+              break;
+          }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           zType = "INTEGER";
           zOrigCol = "rowid";
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         }else{
           zOrigCol = pTab->aCol[iCol].zName;
           zType = sqlite3ColumnType(&pTab->aCol[iCol],0);
@@ -1861,7 +1903,9 @@
   pTabList = pSelect->pSrc;
   pEList = pSelect->pEList;
   assert( v!=0 );
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   assert( pTabList!=0 );
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   pParse->colNamesSet = 1;
   fullName = (db->flags & SQLITE_FullColNames)!=0;
   srcName = (db->flags & SQLITE_ShortColNames)!=0 || fullName;
@@ -1884,7 +1928,18 @@
       if( iCol<0 ) iCol = pTab->iPKey;
       assert( iCol==-1 || (iCol>=0 && iCol<pTab->nCol) );
       if( iCol<0 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        switch( iCol ){
+          default:
+            zCol = "rowid";
+            break;
+          case -3:
+            zCol = "comdb2_rowtimestamp";
+            break;
+        }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         zCol = "rowid";
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       }else{
         zCol = pTab->aCol[iCol].zName;
       }
@@ -1966,7 +2021,6 @@
         pColExpr = pColExpr->pRight;
         assert( pColExpr!=0 );
       }
-      assert( pColExpr->op!=TK_AGG_COLUMN );
       if( pColExpr->op==TK_COLUMN ){
         /* For columns use the column name name */
         int iCol = pColExpr->iColumn;
@@ -2173,7 +2227,7 @@
     p->iLimit = iLimit = ++pParse->nMem;
     v = sqlite3GetVdbe(pParse);
     assert( v!=0 );
-    if( sqlite3ExprIsInteger(pLimit->pLeft, &n) ){
+    if( sqlite3ExprIsInteger(pLimit->pLeft, &n, pParse) ){
       sqlite3VdbeAddOp2(v, OP_Integer, n, iLimit);
       VdbeComment((v, "LIMIT counter"));
       if( n==0 ){
@@ -2187,7 +2241,13 @@
       sqlite3VdbeAddOp1(v, OP_MustBeInt, iLimit); VdbeCoverage(v);
       VdbeComment((v, "LIMIT counter"));
       sqlite3VdbeAddOp2(v, OP_IfNot, iLimit, iBreak); VdbeCoverage(v);
+    }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    int is_parallel;
+    if( (is_parallel = comdb2_register_limit(iLimit, ++pParse->nMem))!=0 ){
+      sqlite3VdbeAddOp2(v, OP_IntCopy, iLimit, pParse->nMem);
     }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( pLimit->pRight ){
       p->iOffset = iOffset = ++pParse->nMem;
       pParse->nMem++;   /* Allocate an extra register for limit+offset */
@@ -2196,6 +2256,12 @@
       VdbeComment((v, "OFFSET counter"));
       sqlite3VdbeAddOp3(v, OP_OffsetLimit, iLimit, iOffset+1, iOffset);
       VdbeComment((v, "LIMIT+OFFSET"));
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( is_parallel ){
+        comdb2_register_offset(iOffset, iOffset+1, ++pParse->nMem);
+        sqlite3VdbeAddOp2(v, OP_IntCopy, iOffset, pParse->nMem);
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
   }
 }
@@ -2604,7 +2670,7 @@
         pPrior->iOffset = p->iOffset;
         pPrior->pLimit = p->pLimit;
         rc = sqlite3Select(pParse, pPrior, &dest);
-        p->pLimit = 0;
+        pPrior->pLimit = 0;
         if( rc ){
           goto multi_select_end;
         }
@@ -2625,8 +2691,8 @@
         pDelete = p->pPrior;
         p->pPrior = pPrior;
         p->nSelectRow = sqlite3LogEstAdd(p->nSelectRow, pPrior->nSelectRow);
-        if( pPrior->pLimit
-         && sqlite3ExprIsInteger(pPrior->pLimit->pLeft, &nLimit)
+        if( p->pLimit
+         && sqlite3ExprIsInteger(p->pLimit->pLeft, &nLimit, 0)
          && nLimit>0 && p->nSelectRow > sqlite3LogEst((u64)nLimit) 
         ){
           p->nSelectRow = sqlite3LogEst((u64)nLimit);
@@ -2692,9 +2758,7 @@
                           selectOpName(p->op)));
         rc = sqlite3Select(pParse, p, &uniondest);
         testcase( rc!=SQLITE_OK );
-        /* Query flattening in sqlite3Select() might refill p->pOrderBy.
-        ** Be sure to delete p->pOrderBy, therefore, to avoid a memory leak. */
-        sqlite3ExprListDelete(db, p->pOrderBy);
+        assert( p->pOrderBy==0 );
         pDelete = p->pPrior;
         p->pPrior = pPrior;
         p->pOrderBy = 0;
@@ -2744,6 +2808,9 @@
         assert( p->pOrderBy==0 );
   
         addr = sqlite3VdbeAddOp2(v, OP_OpenEphemeral, tab1, 0);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        sqlite3VdbeChangeP5(v, BTREE_UNORDERED); 
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         assert( p->addrOpenEphm[0] == -1 );
         p->addrOpenEphm[0] = addr;
         findRightmost(p)->selFlags |= SF_UsesEphemeral;
@@ -3771,6 +3838,12 @@
   ** See also tickets #306, #350, and #3300.
   */
   if( (pSubitem->fg.jointype & JT_OUTER)!=0 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    extern int gbl_enable_sq_flattening_optimization;
+    if (!gbl_enable_sq_flattening_optimization) {
+      return 0;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     isLeftJoin = 1;
     if( pSubSrc->nSrc>1 || isAgg || IsVirtual(pSubSrc->a[0].pTab) ){
       /*  (3a)             (3c)     (3b) */
@@ -3884,6 +3957,9 @@
     p->pLimit = 0;
     pNew = sqlite3SelectDup(db, p, 0);
     p->pLimit = pLimit;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( p->recording ) _set_src_recording(pParse, pSub);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     p->pOrderBy = pOrderBy;
     p->pSrc = pSrc;
     p->op = TK_ALL;
@@ -3905,6 +3981,9 @@
   */
   pSub = pSub1 = pSubitem->pSelect;
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( p->recording ) _set_src_recording(pParse, pSub);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   /* Delete the transient table structure associated with the
   ** subquery
   */
@@ -4010,7 +4089,7 @@
     ** We look at every expression in the outer query and every place we see
     ** "a" we substitute "x*3" and every place we see "b" we substitute "y+10".
     */
-    if( pSub->pOrderBy ){
+    if( pSub->pOrderBy && (pParent->selFlags & SF_NoopOrderBy)==0 ){
       /* At this point, any non-zero iOrderByCol values indicate that the
       ** ORDER BY column expression is identical to the iOrderByCol'th
       ** expression returned by SELECT statement pSub. Since these values
@@ -4032,7 +4111,7 @@
     pWhere = pSub->pWhere;
     pSub->pWhere = 0;
     if( isLeftJoin>0 ){
-      setJoinExpr(pWhere, iNewParent);
+      sqlite3SetJoinExpr(pWhere, iNewParent);
     }
     pParent->pWhere = sqlite3ExprAnd(db, pWhere, pParent->pWhere);
     if( db->mallocFailed==0 ){
@@ -4379,6 +4458,57 @@
   return nChng;
 }
 #endif /* !defined(SQLITE_OMIT_SUBQUERY) || !defined(SQLITE_OMIT_VIEW) */
+
+
+/*
+** Check to see if a subquery contains result-set columns that are
+** never used.  If it does, change the value of those result-set columns
+** to NULL so that they do not cause unnecessary work to compute.
+**
+** Return the number of column that were changed to NULL.
+*/
+static int disableUnusedSubqueryResultColumns(struct SrcList_item *pItem){
+  int nCol;
+  Select *pSub;      /* The subquery to be simplified */
+  Select *pX;        /* For looping over compound elements of pSub */
+  Table *pTab;       /* The table that describes the subquery */
+  int j;             /* Column number */
+  int nChng = 0;     /* Number of columns converted to NULL */
+
+  assert( pItem!=0 );
+  assert( pItem->pTab!=0 );
+  pTab = pItem->pTab;
+  if( pTab->tabFlags & TF_Ephemeral ) return 0;
+  assert( pItem->pSelect!=0 );
+  pSub = pItem->pSelect;
+  assert( pSub->pEList->nExpr==pTab->nCol );
+  for(pX=pSub; pX; pX=pX->pPrior){
+    if( pX->pPrior && pX->op!=TK_ALL ){
+      /* This optimization does not work for compound subqueries that
+      ** use UNION, INTERSECT, or EXCEPT.  Only UNION ALL is allowed. */
+      return 0;
+    }
+    if( pX->pWin ){
+      /* This optimization does not work for subqueries that use window
+      ** functions. */
+      return 0;
+    }
+  }
+  nCol = pTab->nCol;
+  for(j=0; j<nCol; j++){
+    Select *pX;
+    Bitmask m = j<BMS-1 ? MASKBIT(j) : (((Bitmask)1)<<(BMS-1));
+    if( (m & pItem->colUsed)!=0 ) continue;
+    for(pX=pSub; pX; pX=pX->pPrior) {
+      Expr *pY = pX->pEList->a[j].pExpr;
+      if( pY->op==TK_NULL ) continue;
+      pY->op = TK_NULL;
+      /* pX->selFlags |= SF_PushDown; */
+      nChng++;
+    }
+  }
+  return nChng;
+}
 
 /*
 ** The pFunc is the only aggregate function in the query.  Check to see
@@ -4850,7 +4980,11 @@
   /* Make sure cursor numbers have been assigned to all entries in
   ** the FROM clause of the SELECT statement.
   */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  sqlite3SrcListAssignCursors(pParse, pTabList, p->recording);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   sqlite3SrcListAssignCursors(pParse, pTabList);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   /* Look up every table named in the FROM clause of the select.  If
   ** an entry of the FROM clause is a subquery instead of a table or view,
@@ -5398,7 +5532,30 @@
 # define explainSimpleCount(a,b,c)
 #endif
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 /*
+**
+** Check list of identified source tables and check to see any of them is remote
+**
+*/
+static int sql_has_remotes(
+  Parse *pParse,       /* The parsing context */
+  SrcList *pList        /* The source list */
+) {
+  int i;
+
+  for(i=0;i<pList->nSrc; i++){
+    char *dbname = pList->a[i].zDatabase;
+    if(dbname && strcasecmp(dbname,"main") && strcasecmp(dbname, "temp") &&
+            strcasecmp(dbname, comdb2_get_dbname())) {
+      return 1;
+    }
+  }
+  return 0;
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
+/*
 ** sqlite3WalkExpr() callback used by havingToWhere().
 **
 ** If the node passed to the callback is a TK_AND node, return 
@@ -5481,7 +5638,9 @@
       ** names in the same FROM clause. */
       continue;
     }
-    if( sqlite3ExprCompare(0, pThis->pSelect->pWhere, pS1->pWhere, -1) ){
+    if( sqlite3ExprCompare(0, pThis->pSelect->pWhere, pS1->pWhere, -1)
+     || sqlite3ExprCompare(0, pThis->pSelect->pHaving, pS1->pHaving, -1) 
+    ){
       /* The view was modified by some other optimization such as
       ** pushDownWhereTerms() */
       continue;
@@ -5506,7 +5665,8 @@
 **   *  The subquery is a UNION ALL of two or more terms
 **   *  The subquery does not have a LIMIT clause
 **   *  There is no WHERE or GROUP BY or HAVING clauses on the subqueries
-**   *  The outer query is a simple count(*)
+**   *  The outer query is a simple count(*) with no WHERE clause or other
+**      extraneous syntax.
 **
 ** Return TRUE if the optimization is undertaken.
 */
@@ -5517,6 +5677,8 @@
   sqlite3 *db;
   if( (p->selFlags & SF_Aggregate)==0 ) return 0;   /* This is an aggregate */
   if( p->pEList->nExpr!=1 ) return 0;               /* Single result column */
+  if( p->pWhere ) return 0;
+  if( p->pGroupBy ) return 0;
   pExpr = p->pEList->a[0].pExpr;
   if( pExpr->op!=TK_AGG_FUNCTION ) return 0;        /* Result is an aggregate */
   if( sqlite3_stricmp(pExpr->u.zToken,"count") ) return 0;  /* Is count() */
@@ -5636,12 +5798,44 @@
            pDest->eDest==SRT_DistQueue || pDest->eDest==SRT_Fifo);
     /* If ORDER BY makes no difference in the output then neither does
     ** DISTINCT so it can be removed too. */
+    sqlite3WindowRemoveExprListFromSelect(p, p->pOrderBy);
     sqlite3ExprListDelete(db, p->pOrderBy);
     p->pOrderBy = 0;
     p->selFlags &= ~SF_Distinct;
+    p->selFlags |= SF_NoopOrderBy;
   }
   sqlite3SelectPrep(pParse, p, 0);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( !db->mallocFailed && pParse->prepFlags&SQLITE_PREPARE_SRCLIST_ONLY ){
+    SrcList *pSrc = p->pSrc;
+    int iSrc;
+    if( !pSrc ) goto select_end;
+    if( pSrc->nSrc==0 ) goto select_end;
+    pParse->azSrcListOnly = sqlite3DbMallocZero(db, pSrc->nSrc * sizeof(char*));
+    if( db->mallocFailed ) goto select_end;
+    for(iSrc=0; iSrc<pSrc->nSrc; iSrc++){
+      const char *zSrcDatabase = pSrc->a[iSrc].zDatabase;
+      if( !zSrcDatabase ) zSrcDatabase = "main";
+      pParse->azSrcListOnly[iSrc] = sqlite3MPrintf(db,
+          "%s.%s", zSrcDatabase, pSrc->a[iSrc].zName
+      );
+      if( db->mallocFailed ) goto select_end;
+    }
+    pParse->nSrcListOnly = pSrc->nSrc;
+    rc = 0;
+    goto select_end;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( pParse->nErr || db->mallocFailed ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( pParse->checkSchema == 1 /* parsing error */ && 
+            pParse->zErrMsg && strncasecmp(pParse->zErrMsg, "no such column", 
+                strlen("no such column")) == 0){
+      if (sql_has_remotes(pParse, p->pSrc)) {
+        comdb2_set_verify_remote_schemas();
+      }
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     goto select_end;
   }
   assert( p->pEList!=0 );
@@ -5755,6 +5949,10 @@
   }
 #endif
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  ast_t *ast = ast_init(pParse, __func__);
+  if( ast ) ast_push(ast, AST_TYPE_SELECT, v, p);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifndef SQLITE_OMIT_COMPOUND_SELECT
   /* Handle compound SELECT statements using the separate multiSelect()
   ** procedure.
@@ -5871,6 +6069,20 @@
 #endif
     }else{
       SELECTTRACE(0x100,pParse,p,("Push-down not possible\n"));
+    }
+
+    /* Convert unused result columns of the subquery into simple NULL
+    ** expressions, to avoid unneeded searching and computation.
+    */
+    if( disableUnusedSubqueryResultColumns(pItem) ){
+#if TREETRACE_ENABLED
+      if( sqlite3TreeTrace & 0x4000 ){
+        TREETRACE(0x4000,pParse,p,
+            ("Change unused result columns to NULL for subquery %d:\n",
+             pSub->selId));
+        sqlite3TreeViewSelect(0, p, 0);
+      }
+#endif
     }
 
     zSavedAuthContext = pParse->zAuthContext;
@@ -5969,6 +6181,17 @@
   }
 #endif
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#ifdef SQLITE_COUNTOFVIEW_OPTIMIZATION
+  if( OptimizationEnabled(db, SQLITE_QueryFlattener|SQLITE_CountOfView)
+   && countOfViewOptimization(pParse, p)
+  ){
+    if( db->mallocFailed ) goto select_end;
+    pEList = p->pEList;
+    pTabList = p->pSrc;
+  }
+#endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   /* If the query is DISTINCT with an ORDER BY but is not an aggregate, and 
   ** if the select-list is the same as the ORDER BY list, then this query
   ** can be rewritten as a GROUP BY. In other words, this:
@@ -6484,6 +6707,10 @@
         int iRoot = pTab->tnum;              /* Root page of scanned b-tree */
 
         sqlite3CodeVerifySchema(pParse, iDb);
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        sqlite3VdbeAddTable(v, pTab);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         sqlite3TableLock(pParse, iDb, pTab->tnum, 0, pTab->zName);
 
         /* Search for the index that has the lowest scan cost.
@@ -6505,12 +6732,31 @@
             pBest = pIdx;
           }
         }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        /* prefer data-btree if parallel count is enabled */
+        extern int gbl_direct_count;
+        extern int gbl_parallel_count;
+        int use_data = gbl_direct_count && gbl_parallel_count;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         if( pBest ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          if( use_data==0 ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           iRoot = pBest->tnum;
           pKeyInfo = sqlite3KeyInfoOfIndex(pParse, pBest);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         }
 
         /* Open a read-only cursor, execute the OP_Count, close the cursor. */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        /* TODO: use op4 ? */
+        if( p->recording ){
+            /* sqlite3VdbeAddOp3(v, OP_OpenRead_Record, iCsr, iRoot, iDb); */
+            sqlite3VdbeAddOp4Int(v, OP_OpenRead_Record, iCsr, iRoot, iDb, 1);
+        }else
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         sqlite3VdbeAddOp4Int(v, OP_OpenRead, iCsr, iRoot, iDb, 1);
         if( pKeyInfo ){
           sqlite3VdbeChangeP4(v, -1, (char *)pKeyInfo, P4_KEYINFO);
