--- sqlite-3.28.0/src/sqlite.h.in	2026-01-09 16:11:01
+++ src/sqlite.h.in	2025-09-24 15:21:58
@@ -126,6 +126,10 @@
 #define SQLITE_VERSION        "--VERS--"
 #define SQLITE_VERSION_NUMBER --VERSION-NUMBER--
 #define SQLITE_SOURCE_ID      "--SOURCE-ID--"
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <types.h>
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** CAPI3REF: Run-Time Library Version Numbers
@@ -454,6 +458,37 @@
 #define SQLITE_WARNING     28   /* Warnings from sqlite3_log() */
 #define SQLITE_ROW         100  /* sqlite3_step() has another row ready */
 #define SQLITE_DONE        101  /* sqlite3_step() has finished executing */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#define SQLITE_SCHEMA_REMOTE      50 /* distributed schema has inconsistent 
+                                      * local cache */
+#define SQLITE_SCHEMA_DOHSQL      51 /* see dohsql */
+#define SQLITE_EARLYSTOP_DOHSQL   52 /* see dohsql */
+#define SQLITE_SCHEMA_PUSH_REMOTE 53 /* push query to remote node */
+#define SQLITE_SCHEMA_PUSH_REMOTE_WRITE 54 /* push write query to remote node */
+#define SQLITE_DEADLOCK          200 /* deadlock happened, transaction aborted */
+#define SQLITE_ACCESS            201 /* failed permissions */
+#define SQLITE_LIMIT_DEPRECATED  202 /* SQLITE_LIMIT -- no longer used */
+#define SQLITE_TRANTOOCOMPLEX    204 /* I refuse to use 203 for this */
+#define SQLITE_TRAN_CANCELLED    205 /* resource had to be freed, transaction
+                                      * was aborted */
+#define SQLITE_TRAN_NOLOG        206 /* log has been freed, unable to maintain
+                                      * snapshot */
+#define SQLITE_TRAN_NOUNDO       207 /* the database suffered an NON-undoable
+                                      * operation: fastinit, sc */
+#define SQLITE_CONV_ERROR        208 /* error converting types, sqlite_column_*
+                                      * routines */
+#define SQLITE_COMDB2SCHEMA      209 /* The current Vdbe has been changed under
+                                      * the hood (the compiled statement
+                                      * executed comdb2 custom code) */
+#define SQLITE_CLIENT_CHANGENODE 210 /* Ask the client api to retry against
+                                      * another node */
+#define SQLITE_DDL_MISUSE        211 /* overlapping tables detected in
+                                      * transactional DDLs */
+#define SQLITE_TIMEDOUT          212 /* query timed out */
+#define SQLITE_COST_TOO_HIGH     213 /* query cost too high */
+#define SQLITE_NO_TEMPTABLES     214 /* temporary tables disallowed */
+#define SQLITE_NO_TABLESCANS     215 /* table scans disallowed */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 /* end-of-error-codes */
 
 /*
@@ -2999,6 +3034,27 @@
 #define SQLITE_SAVEPOINT            32   /* Operation       Savepoint Name  */
 #define SQLITE_COPY                  0   /* No longer used */
 #define SQLITE_RECURSIVE            33   /* NULL            NULL            */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#define SQLITE_REBUILD_TABLE       160   /* NULL            NULL            */
+#define SQLITE_REBUILD_INDEX       161   /* NULL            NULL            */
+#define SQLITE_REBUILD_DATA        162   /* NULL            NULL            */
+#define SQLITE_REBUILD_DATABLOB    163   /* NULL            NULL            */
+#define SQLITE_TRUNCATE_TABLE      164   /* NULL            NULL            */
+#define SQLITE_CREATE_PROC         166   /* NULL            NULL            */
+#define SQLITE_DROP_PROC           168   /* NULL            NULL            */
+#define SQLITE_CREATE_PART         169   /* NULL            NULL            */
+#define SQLITE_DROP_PART           170   /* NULL            NULL            */
+#define SQLITE_GET_TUNABLE         171   /* NULL            NULL            */
+#define SQLITE_PUT_TUNABLE         172   /* NULL            NULL            */
+#define SQLITE_GRANT               173   /* NULL            NULL            */
+#define SQLITE_REVOKE              174   /* NULL            NULL            */
+#define SQLITE_CREATE_LUA_FUNCTION 175   /* NULL            NULL            */
+#define SQLITE_DROP_LUA_FUNCTION   176   /* NULL            NULL            */
+#define SQLITE_CREATE_LUA_TRIGGER  177   /* NULL            NULL            */
+#define SQLITE_DROP_LUA_TRIGGER    178   /* NULL            NULL            */
+#define SQLITE_CREATE_LUA_CONSUMER 179   /* NULL            NULL            */
+#define SQLITE_DROP_LUA_CONSUMER   180   /* NULL            NULL            */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** CAPI3REF: Tracing And Profiling Functions
@@ -3391,9 +3447,15 @@
 **
 ** See also: [sqlite3_temp_directory]
 */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+struct sqlthdstate;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 int sqlite3_open(
   const char *filename,   /* Database filename (UTF-8) */
   sqlite3 **ppDb          /* OUT: SQLite db handle */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  ,struct sqlthdstate *
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 );
 int sqlite3_open16(
   const void *filename,   /* Database filename (UTF-16) */
@@ -3685,6 +3747,10 @@
 #define SQLITE_PREPARE_PERSISTENT              0x01
 #define SQLITE_PREPARE_NORMALIZE               0x02
 #define SQLITE_PREPARE_NO_VTAB                 0x04
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#define SQLITE_PREPARE_ONLY                    0x08
+#define SQLITE_PREPARE_SRCLIST_ONLY            0x10
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** CAPI3REF: Compiling An SQL Statement
@@ -4131,6 +4197,11 @@
 int sqlite3_bind_pointer(sqlite3_stmt*, int, void*, const char*,void(*)(void*));
 int sqlite3_bind_zeroblob(sqlite3_stmt*, int, int n);
 int sqlite3_bind_zeroblob64(sqlite3_stmt*, int, sqlite3_uint64);
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+int sqlite3_bind_datetime(sqlite3_stmt *pStmt, int i, dttz_t *dt, char *tz);
+int sqlite3_bind_interval(sqlite3_stmt *pStmt, int i, intv_t *it);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** CAPI3REF: Number Of SQL Parameters
@@ -4478,6 +4549,16 @@
 #endif
 #define SQLITE3_TEXT     3
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#define SQLITE_DATETIME       6
+#define SQLITE_INTERVAL_YM    7
+#define SQLITE_INTERVAL_DS    8
+#define SQLITE_DATETIMEUS     9
+#define SQLITE_INTERVAL_DSUS 10
+#define SQLITE_DECIMAL       11
+#define SQLITE_NEXTSEQ       (SQLITE_MAX_U32-2)
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** CAPI3REF: Result Values From A Query
 ** KEYWORDS: {column access functions}
@@ -4698,6 +4779,15 @@
 int sqlite3_column_bytes16(sqlite3_stmt*, int iCol);
 int sqlite3_column_type(sqlite3_stmt*, int iCol);
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+const dttz_t *sqlite3_column_datetime(sqlite3_stmt *pStmt, int i);
+const intv_t *sqlite3_column_interval(sqlite3_stmt *pStmt, int i, int type);
+
+int sqlite3_hasResultSet(sqlite3_stmt*);
+int sqlite3_hasNColumns(sqlite3_stmt*, int iCol);
+int sqlite3_isColumnNullType(sqlite3_stmt*, int iCol);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** CAPI3REF: Destroy A Prepared Statement Object
 ** DESTRUCTOR: sqlite3_stmt
@@ -4752,6 +4842,24 @@
 ** of any [sqlite3_bind_blob|bindings] on the [prepared statement] S.
 */
 int sqlite3_reset(sqlite3_stmt *pStmt);
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/* Reset tspec of Prepared Statement. */
+int sqlite3_resetclock(sqlite3_stmt *pStmt);
+char *stmt_tzname(sqlite3_stmt *);
+void stmt_set_dtprec(sqlite3_stmt *, int);
+
+int stmt_cached_column_count(sqlite3_stmt *);
+char *stmt_cached_column_name(sqlite3_stmt *, int);
+char *stmt_column_name(sqlite3_stmt *, int);
+char *stmt_column_decltype(sqlite3_stmt *pStmt, int index);
+char *stmt_cached_column_decltype(sqlite3_stmt *pStmt, int index);
+void stmt_set_cached_columns(sqlite3_stmt *, char **, char **, int);
+void stmt_set_vlock_tables(sqlite3_stmt *, char **, int, int, int);
+void stmt_set_has_scalar_func(sqlite3_stmt *, int);
+int stmt_do_column_names_match(sqlite3_stmt *);
+int stmt_do_column_decltypes_match(sqlite3_stmt *pStmt);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** CAPI3REF: Create Or Redefine SQL Functions
@@ -5093,6 +5201,11 @@
 int sqlite3_value_numeric_type(sqlite3_value*);
 int sqlite3_value_nochange(sqlite3_value*);
 int sqlite3_value_frombind(sqlite3_value*);
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+const dttz_t *sqlite3_value_datetime(sqlite3_value*);
+const intv_t *sqlite3_value_interval(sqlite3_value*, int type);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** CAPI3REF: Finding The Subtype Of SQL Values
@@ -5423,6 +5536,11 @@
 void sqlite3_result_zeroblob(sqlite3_context*, int n);
 int sqlite3_result_zeroblob64(sqlite3_context*, sqlite3_uint64 n);
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+void sqlite3_result_datetime(sqlite3_context*, dttz_t*, const char *tz);
+void sqlite3_result_interval(sqlite3_context*, intv_t*);
+void sqlite3_result_decimal(sqlite3_context*, decQuad*);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** CAPI3REF: Setting The Subtype Of An SQL Function
@@ -6180,6 +6298,10 @@
   int *pPrimaryKey,           /* OUTPUT: True if column part of PK */
   int *pAutoinc               /* OUTPUT: True if column is auto-increment */
 );
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+char* sqlite3_prepare_plan(sqlite3_stmt *stmt);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** CAPI3REF: Load An Extension
@@ -6316,6 +6438,10 @@
 ** registered using [sqlite3_auto_extension()].
 */
 void sqlite3_reset_auto_extension(void);
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+int sqlite3_stmt_has_remotes(sqlite3_stmt *stmt);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** The interface to the virtual-table mechanism is currently considered
@@ -6386,7 +6512,19 @@
   /* The methods above are in versions 1 and 2 of the sqlite_module object.
   ** Those below are for version 3 and greater. */
   int (*xShadowName)(const char*);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  int access_flag;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+};
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+enum {
+  CDB2_ALLOW_ALL  = 1<<0, /* Allow access to all */
+  CDB2_ALLOW_USER = 1<<1, /* Limit access only to permitted users */
+  CDB2_HIDDEN     = 1<<2, /* Remove it from system table list
+                             (comdb2_systables) */
 };
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** CAPI3REF: Virtual Table Indexing Information
@@ -6514,6 +6652,9 @@
   int idxFlags;              /* Mask of SQLITE_INDEX_SCAN_* flags */
   /* Fields below are only available in SQLite 3.10.0 and later */
   sqlite3_uint64 colUsed;    /* Input: Mask of columns used by statement */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  char *zTable;
+#endif /* SQLITE_BUILDING_FOR_COMDB2 */
 };
 
 /*
