--- sqlite-3.28.0/src/sqliteLimit.h	2026-01-09 16:11:01
+++ src/sqliteLimit.h	2025-09-24 15:21:58
@@ -67,7 +67,11 @@
 ** at all times.
 */
 #ifndef SQLITE_MAX_EXPR_DEPTH
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+# define SQLITE_MAX_EXPR_DEPTH 3000
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 # define SQLITE_MAX_EXPR_DEPTH 1000
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
 
 /*
@@ -125,7 +129,11 @@
 ** and we have to allow 2 extra counts for the "main" and "temp" databases.
 */
 #ifndef SQLITE_MAX_ATTACHED
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+# define SQLITE_MAX_ATTACHED 100
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 # define SQLITE_MAX_ATTACHED 10
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
 
 
@@ -133,7 +141,12 @@
 ** The maximum value of a ?nnn wildcard that the parser will accept.
 */
 #ifndef SQLITE_MAX_VARIABLE_NUMBER
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/* COMPAT: Needs 1024 vars because Oracle has 1000 and DB2 has 1012. */
+# define SQLITE_MAX_VARIABLE_NUMBER 2048
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 # define SQLITE_MAX_VARIABLE_NUMBER 999
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
 
 /* Maximum page size.  The upper bound on this value is 65536.  This a limit
