--- sqlite-3.28.0/src/test_vfs.c	2026-01-09 16:11:01
+++ src/test_vfs.c	2025-09-24 15:21:58
@@ -84,6 +84,10 @@
   sqlite3_vfs *pVfs;              /* The testvfs registered with SQLite */
   Tcl_Interp *interp;             /* Interpreter to run script in */
   Tcl_Obj *pScript;               /* Script to execute */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  int nScript;                    /* Number of elements in array apScript */
+  Tcl_Obj **apScript;             /* Array version of pScript */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   TestvfsBuffer *pBuffer;         /* List of shared buffers */
   int isNoshm;
   int isFullshm;
@@ -286,7 +290,14 @@
   Tcl_Obj *arg4
 ){
   int rc;                         /* Return code from Tcl_EvalObj() */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  int nArg;                       /* Elements in eval'd list */
+  int nScript;
+  Tcl_Obj ** ap;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   Tcl_Obj *pEval;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
   assert( p->pScript );
 
   assert( zMethod );
@@ -294,6 +305,29 @@
   assert( arg2==0 || arg1!=0 );
   assert( arg3==0 || arg2!=0 );
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( !p->apScript ){
+    int nByte;
+    int i;
+    if( TCL_OK!=Tcl_ListObjGetElements(p->interp, p->pScript, &nScript, &ap) ){
+      Tcl_BackgroundError(p->interp);
+      Tcl_ResetResult(p->interp);
+      return;
+    }
+    p->nScript = nScript;
+    nByte = (nScript+TESTVFS_MAX_ARGS)*sizeof(Tcl_Obj *);
+    p->apScript = (Tcl_Obj **)ckalloc(nByte);
+    memset(p->apScript, 0, nByte);
+    for(i=0; i<nScript; i++){
+      p->apScript[i] = ap[i];
+    }
+  }
+  p->apScript[p->nScript] = Tcl_NewStringObj(zMethod, -1);
+  p->apScript[p->nScript+1] = arg1;
+  p->apScript[p->nScript+2] = arg2;
+  p->apScript[p->nScript+3] = arg3;
+  /* TODO: p->apScript[p->nScript+4] = arg4; */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pEval = Tcl_DuplicateObj(p->pScript);
   Tcl_IncrRefCount(p->pScript);
   Tcl_ListObjAppendElement(p->interp, pEval, Tcl_NewStringObj(zMethod, -1));
@@ -302,11 +336,24 @@
   if( arg3 ) Tcl_ListObjAppendElement(p->interp, pEval, arg3);
   if( arg4 ) Tcl_ListObjAppendElement(p->interp, pEval, arg4);
 
+  for(nArg=p->nScript; p->apScript[nArg]; nArg++){
+    Tcl_IncrRefCount(p->apScript[nArg]);
+  }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  rc = Tcl_EvalObjv(p->interp, nArg, p->apScript, TCL_EVAL_GLOBAL);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   rc = Tcl_EvalObjEx(p->interp, pEval, TCL_EVAL_GLOBAL);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( rc!=TCL_OK ){
     Tcl_BackgroundError(p->interp);
     Tcl_ResetResult(p->interp);
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  for(nArg=p->nScript; p->apScript[nArg]; nArg++){
+    Tcl_DecrRefCount(p->apScript[nArg]);
+    p->apScript[nArg] = 0;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
 
@@ -1233,6 +1280,11 @@
         int nByte;
         if( p->pScript ){
           Tcl_DecrRefCount(p->pScript);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          ckfree((char *)p->apScript);
+          p->apScript = 0;
+          p->nScript = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           p->pScript = 0;
         }
         Tcl_GetStringFromObj(objv[2], &nByte);
@@ -1388,6 +1440,10 @@
   Testvfs *p = (Testvfs *)cd;
   if( p->pScript ) Tcl_DecrRefCount(p->pScript);
   sqlite3_vfs_unregister(p->pVfs);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+  ckfree((char *)p->apScript);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   ckfree((char *)p->pVfs);
   ckfree((char *)p);
 }
