--- sqlite-3.28.0/src/threads.c	2026-01-09 16:05:49
+++ src/threads.c	2025-09-24 15:21:58
@@ -72,13 +72,10 @@
   ** forces worker threads to run sequentially and deterministically 
   ** for testing purposes. */
   if( sqlite3FaultSim(200) ){
-    rc = 1;
-  }else{    
-    rc = pthread_create(&p->tid, 0, xTask, pIn);
-  }
-  if( rc ){
     p->done = 1;
     p->pOut = xTask(pIn);
+  }else{    
+    Pthread_create(&p->tid, 0, xTask, pIn);
   }
   *ppThread = p;
   return SQLITE_OK;
