--- sqlite-3.28.0/src/tokenize.c	2026-01-09 16:11:01
+++ src/tokenize.c	2026-01-09 15:00:03
@@ -55,6 +55,10 @@
 #define CC_DOT       26    /* '.' */
 #define CC_ILLEGAL   27    /* Illegal character */
 #define CC_NUL       28    /* 0x00 */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#define CC_LB        29    /* '{' */
+#define CC_RB        30    /* '}' */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 static const unsigned char aiClass[] = {
 #ifdef SQLITE_ASCII
@@ -66,7 +70,11 @@
 /* 4x */    5,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
 /* 5x */    1,  1,  1,  1,  1,  1,  1,  1,  0,  1,  1,  9, 27, 27, 27,  1,
 /* 6x */    8,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/* 7x */    1,  1,  1,  1,  1,  1,  1,  1,  0,  1,  1, 29, 10, 30, 25, 27,
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 /* 7x */    1,  1,  1,  1,  1,  1,  1,  1,  0,  1,  1, 27, 10, 27, 25, 27,
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 /* 8x */    2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
 /* 9x */    2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
 /* Ax */    2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,  2,
@@ -90,8 +98,13 @@
 /* 9x */   27,  1,  1,  1,  1,  1,  1,  1,  1,  1, 27, 27, 27, 27, 27, 27,
 /* Ax */   27, 25,  1,  1,  1,  1,  1,  0,  1,  1, 27, 27, 27, 27, 27, 27,
 /* Bx */   27, 27, 27, 27, 27, 27, 27, 27, 27, 27,  9, 27, 27, 27, 27, 27,
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/* Cx */   29,  1,  1,  1,  1,  1,  1,  1,  1,  1, 27, 27, 27, 27, 27, 27,
+/* Dx */   30,  1,  1,  1,  1,  1,  1,  1,  1,  1, 27, 27, 27, 27, 27, 27,
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 /* Cx */   27,  1,  1,  1,  1,  1,  1,  1,  1,  1, 27, 27, 27, 27, 27, 27,
 /* Dx */   27,  1,  1,  1,  1,  1,  1,  1,  1,  1, 27, 27, 27, 27, 27, 27,
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 /* Ex */   27, 27,  1,  1,  1,  1,  1,  0,  1,  1, 27, 27, 27, 27, 27, 27,
 /* Fx */    3,  3,  3,  3,  3,  3,  3,  3,  3,  3, 27, 27, 27, 27, 27, 27,
 #endif
@@ -504,12 +517,37 @@
         /* This token started out using characters that can appear in keywords,
         ** but z[i] is a character not allowed within keywords, so this must
         ** be an identifier instead */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        /* Special cases for COMDB2 */
+        if( i==5 && sqlite3StrNICmp((char*)z,"genid",5)==0 && z[i]=='4'
+            && z[i+1]=='8' && !IdChar(z[i+2]) ){
+          *tokenType = TK_GENID48;
+          return 7;
+        }else if( i==2 && sqlite3StrNICmp((char*)z,"lz",2)==0 && z[i]=='4'
+            && !IdChar(z[i+1]) ){
+          *tokenType = TK_LZ4;
+          return 3;
+        }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         i++;
         break;
       }
       *tokenType = TK_ID;
       return keywordCode((char*)z, i, tokenType);
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    case CC_LB: {
+      /*
+      ** NOTE: This code assumes that the curly braced block represents the
+      **       last token in the string.  For backward compatibility, there
+      **       is no checking for unmatched curly braces.
+      */
+      for(i=1; (c=z[i])!=0; i++){}
+      testcase( z[i-1]=='}' );  testcase( z[i-1]!='}' );
+      *tokenType = z[i-1]=='}' ? TK_NOSQL : TK_ILLEGAL;
+      return i;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     case CC_X: {
 #ifndef SQLITE_OMIT_BLOB_LITERAL
       testcase( z[0]=='x' ); testcase( z[0]=='X' );
@@ -597,6 +635,9 @@
   assert( pParse->pVList==0 );
   pParse->pParentParse = db->pParse;
   db->pParse = pParse;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  int need_to_check_for_semi = pParse->prepFlags & SQLITE_PREPARE_REQUIRE_SEMI;
+#endif
   while( 1 ){
     n = sqlite3GetToken((u8*)zSql, &tokenType);
     mxSqlLen -= n;
@@ -622,16 +663,27 @@
         continue;
       }
       if( zSql[0]==0 ){
-        /* Upon reaching the end of input, call the parser two more times
-        ** with tokens TK_SEMI and 0, in that order. */
-        if( lastTokenParsed==TK_SEMI ){
-          tokenType = 0;
-        }else if( lastTokenParsed==0 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        if (need_to_check_for_semi && lastTokenParsed != TK_SEMI) {
+          pParse->rc = SQLITE_MISSING_SEMI;
           break;
-        }else{
-          tokenType = TK_SEMI;
+        } else
+#endif
+        if (lastTokenParsed == 0) {
+          break;
+        } else {
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          need_to_check_for_semi = 0;
+#endif
+          /* 
+           * Upon reaching the end of input, call the parser two more times
+           * with tokens TK_SEMI and 0, in that order.
+           * (if we're running with SQLITE_PREPARE_REQUIRE_SEMI,
+           * then the client is expected to provide the semicolon).
+           */
+          tokenType = lastTokenParsed==TK_SEMI ? 0 : TK_SEMI;
+          n = 0;
         }
-        n = 0;
 #ifndef SQLITE_OMIT_WINDOWFUNC
       }else if( tokenType==TK_WINDOW ){
         assert( n==6 );
@@ -643,7 +695,14 @@
         assert( n==6 );
         tokenType = analyzeFilterKeyword((const u8*)&zSql[6], lastTokenParsed);
 #endif /* SQLITE_OMIT_WINDOWFUNC */
-      }else{
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      } else if (need_to_check_for_semi && !sqlite3_complete(pParse->zTail)) {
+        // This check is needed for a case in which a trailing semicolon
+        // is nested within an illegal token -- ex: select ';
+        pParse->rc = SQLITE_MISSING_SEMI;
+        break;
+#endif
+      } else {
         sqlite3ErrorMsg(pParse, "unrecognized token: \"%.*s\"", n, zSql);
         break;
       }
@@ -654,6 +713,18 @@
     lastTokenParsed = tokenType;
     zSql += n;
     assert( db->mallocFailed==0 || pParse->rc!=SQLITE_OK || startedWithOom );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( pParse->prepFlags&SQLITE_PREPARE_SRCLIST_ONLY ){
+      if( pParse->rc!=SQLITE_OK ){
+        if( pParse->zErrMsg ){
+          sqlite3DbFree(db, pParse->zErrMsg);
+          pParse->zErrMsg = 0;
+        }
+        pParse->rc = SQLITE_OK;
+        break;
+      }
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( pParse->rc!=SQLITE_OK ) break;
   }
   assert( nErr==0 );
@@ -740,7 +811,209 @@
   }
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 /*
+** This function should return non-zero if the specified token type is able
+** to be the left-hand operand of a binary sub-expression that makes use of
+** either a plus or minus (binary) operator.
+*/
+static int isLeftOperand(int tokenType){
+  switch( tokenType ){
+    case TK_NULL:
+    case TK_TRUEFALSE:
+    case TK_STRING:
+    case TK_INTEGER:
+    case TK_FLOAT:
+    case TK_VARIABLE:
+    case TK_BLOB:
+    case TK_ID:       return 1;
+    default:          return 0;
+  }
+}
+
+/*
+** This function should return non-zero if the specified token type is
+** a keyword that requires the next token to be a table identifier.
+*/
+static int requiresTableId(int tokenType, int useStrId, int prevType){
+  /* Treat the next STRING token as an ID, if we encounter any of the following
+   * token types:
+   *
+   * FROM   (eg SELECT * FROM 'a'('1'), 'a' should be treated as an ID, and '1' should be treated as a STRING)
+   * JOIN   (eg SELECT * FROM a JOIN 'b', 'b' should be treated as an ID)
+   * INTO   (eg INSERT INTO 'a'('i', 'j') VALUES (1, 1), 'a', 'i', 'j' should be treated as ID)
+   * UPDATE (eg UPDATE 'a' SET i=1, 'a' should be treated as an ID)
+   * TABLE  (eg CREATE TABLE 'a' ('i' INTEGER, 'j' INTEGER), 'a', 'i', 'j' should be treated as ID)
+   * */
+  switch( tokenType ){
+    case TK_FROM:
+    case TK_JOIN:
+    case TK_INTO:
+    case TK_UPDATE:
+    case TK_TABLE:    return 1;
+    case TK_LP:
+    case TK_SPACE:    return useStrId;
+    case TK_STRING:
+    case TK_ID:       return ( prevType==TK_INTO || prevType==TK_TABLE );
+    default:          return 0;
+  }
+}
+
+char *sqlite3Normalize_alternate(
+  Vdbe *pVdbe,       /* VM being reprepared */
+  const char *zSql,  /* The original SQL string */
+  int iDefDqId       /* Zero if double quoted strings should always be
+                      * treated as identifiers when there is no Vdbe
+                      * available */
+){
+  sqlite3 *db;       /* The database connection */
+  int i;             /* Next unread byte of zSql[] */
+  int n;             /* length of current token */
+  int tokenType;     /* type of current token */
+  int prevType = 0;  /* Previous non-whitespace token */
+  int nParen;        /* Number of nested levels of parentheses */
+  int iStartIN;      /* Start of RHS of IN operator in z[] */
+  int nParenAtIN;    /* Value of nParen at start of RHS of IN operator */
+  int j;             /* Bytes of normalized SQL generated so far */
+  sqlite3_str *pStr; /* The normalized SQL string under construction */
+  int useStrId = 0;  /* Treat string token as identifier */
+
+  db = pVdbe ? sqlite3VdbeDb(pVdbe) : 0;
+  tokenType = -1;
+  nParen = iStartIN = nParenAtIN = 0;
+  pStr = sqlite3_str_new(db);
+  assert( pStr!=0 );  /* sqlite3_str_new() never returns NULL */
+  for(i=0; zSql[i] && pStr->accError==0; i+=n){
+    if( tokenType!=TK_SPACE ){
+      prevType = tokenType;
+    }
+    n = sqlite3GetToken((unsigned char*)zSql+i, &tokenType);
+    if( NEVER(n<=0) ) break;
+    switch( tokenType ){
+      case TK_SPACE: {
+        break;
+      }
+      case TK_NULL: {
+        if( prevType==TK_IS || prevType==TK_NOT ){
+          sqlite3_str_append(pStr, " NULL", 5);
+          break;
+        }
+      }
+      case TK_INTEGER:
+      case TK_FLOAT:
+      case TK_VARIABLE:
+      case TK_BLOB: {
+        sqlite3_str_append(pStr, "?", 1);
+        break;
+      }
+      case TK_LP: {
+        nParen++;
+        if( prevType==TK_IN ){
+          iStartIN = pStr->nChar;
+          nParenAtIN = nParen;
+        }
+        sqlite3_str_append(pStr, "(", 1);
+        break;
+      }
+      case TK_RP: {
+        if( iStartIN>0 && nParen==nParenAtIN ){
+          assert( pStr->nChar>=iStartIN );
+          pStr->nChar = iStartIN+1;
+          sqlite3_str_append(pStr, "?,?,?", 5);
+          iStartIN = 0;
+        }
+        nParen--;
+        sqlite3_str_append(pStr, ")", 1);
+        break;
+      }
+      case TK_STRING:
+      case TK_ID: {
+        j = pStr->nChar;
+        if( sqlite3Isquote(zSql[i]) ){
+          /* Dequote the token to get just the identifier. */
+          char *zId = sqlite3_mprintf("%.*s", n, zSql+i);
+          if( zId==0 ) break;
+          sqlite3Dequote(zId);
+
+          /* If it is a string literal, replace it with a placeholder. */
+          int isSingleQuotedLiteral = ( zSql[i]=='\'' );
+          int isDoubleQuotedLiteral = ( zSql[i]=='"'  && sqlite3VdbeUsesDoubleQuotedString(pVdbe, zId, iDefDqId) );
+          int isLiteral = ( isSingleQuotedLiteral || isDoubleQuotedLiteral ) && !useStrId;
+
+          if ( isLiteral ){
+            sqlite3_str_append(pStr, "?", 1);
+            sqlite3DbFree(db, zId);
+            break;
+          }
+
+          /* This is definitely an identifier. Re-tokenize the dequoted
+          ** string to check whether it contains any internal double quotes.
+          ** If it does, re-quote it and escape the internal double quotes. */
+          int eType = 0;
+          int nId = sqlite3Strlen30(zId);
+
+          if( sqlite3GetToken((u8*)zId, &eType)==nId && eType==TK_ID ){
+            addSpaceSeparator(pStr);
+            sqlite3_str_append(pStr, zId, nId);
+          }else{
+            sqlite3_str_appendf(pStr, "\"%w\"", zId);
+          }
+
+          sqlite3DbFree(db, zId);
+        }else{
+          /* Not quoted, this is definitely an identifier. */
+          addSpaceSeparator(pStr);
+          sqlite3_str_append(pStr, zSql+i, n);
+        }
+
+        /* At this point, we know for sure we have an identifier and not a
+        ** string literal. If we are in an IN clause, prevent the identifier
+        ** from being clobbered with '?,?,?' later. */
+        iStartIN = 0;
+
+        /* Identifiers are case insensitive, "select a from t" and
+        ** "select A from t" should have the same normalized SQL. */
+        while( j<pStr->nChar ){
+          pStr->zText[j] = sqlite3Tolower(pStr->zText[j]);
+          j++;
+        }
+        break;
+      }
+      case TK_NOSQL: {
+        addSpaceSeparator(pStr);
+        sqlite3_str_append(pStr, zSql+i, n);
+        break;
+      }
+      case TK_PLUS:
+      case TK_MINUS: {
+        if( isLeftOperand(prevType) ){
+          sqlite3_str_append(pStr, zSql+i, n);
+        }
+        break;
+      }
+      case TK_SELECT: {
+        iStartIN = 0;
+        /* fall through */
+      }
+      default: {
+        if( sqlite3IsIdChar(zSql[i]) ) addSpaceSeparator(pStr);
+        j = pStr->nChar;
+        sqlite3_str_append(pStr, zSql+i, n);
+        while( j<pStr->nChar ){
+          pStr->zText[j] = sqlite3Toupper(pStr->zText[j]);
+          j++;
+        }
+        break;
+      }
+    }
+    useStrId = requiresTableId(tokenType, useStrId, prevType);
+  }
+  if( tokenType!=TK_SEMI ) sqlite3_str_append(pStr, ";", 1);
+  return sqlite3_str_finish(pStr);
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
+/*
 ** Compute a normalization of the SQL given by zSql[0..nSql-1].  Return
 ** the normalization in space obtained from sqlite3DbMalloc().  Or return
 ** NULL if anything goes wrong or if zSql is NULL.
@@ -748,6 +1021,11 @@
 char *sqlite3Normalize(
   Vdbe *pVdbe,       /* VM being reprepared */
   const char *zSql   /* The original SQL string */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  ,int iDefDqId      /* Zero if double quoted strings should always be
+                      * treated as identifiers when there is no Vdbe
+                      * available */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 ){
   sqlite3 *db;       /* The database connection */
   int i;             /* Next unread byte of zSql[] */
@@ -760,7 +1038,11 @@
   int j;             /* Bytes of normalized SQL generated so far */
   sqlite3_str *pStr; /* The normalized SQL string under construction */
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  db = pVdbe ? sqlite3VdbeDb(pVdbe) : 0;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   db = sqlite3VdbeDb(pVdbe);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   tokenType = -1;
   nParen = iStartIN = nParenAtIN = 0;
   pStr = sqlite3_str_new(db);
@@ -814,12 +1096,20 @@
         iStartIN = 0;
         j = pStr->nChar;
         if( sqlite3Isquote(zSql[i]) ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          char *zId = sqlite3_mprintf("%.*s", n, zSql+i);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           char *zId = sqlite3DbStrNDup(db, zSql+i, n);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           int nId;
           int eType = 0;
           if( zId==0 ) break;
           sqlite3Dequote(zId);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          if( zSql[i]=='"' && sqlite3VdbeUsesDoubleQuotedString(pVdbe, zId, iDefDqId) ){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           if( zSql[i]=='"' && sqlite3VdbeUsesDoubleQuotedString(pVdbe, zId) ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
             sqlite3_str_append(pStr, "?", 1);
             sqlite3DbFree(db, zId);
             break;
@@ -842,6 +1132,13 @@
         }
         break;
       }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      case TK_NOSQL: {
+        addSpaceSeparator(pStr);
+        sqlite3_str_append(pStr, zSql+i, n);
+        break;
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       case TK_SELECT: {
         iStartIN = 0;
         /* fall through */
