--- sqlite-3.28.0/src/trigger.c	2026-01-09 16:11:01
+++ src/trigger.c	2025-09-24 15:21:58
@@ -103,6 +103,10 @@
   assert( pName2!=0 );
   assert( op==TK_INSERT || op==TK_UPDATE || op==TK_DELETE );
   assert( op>0 && op<0xff );
+  if(comdb2IsDryrun(pParse)){
+      sqlite3ErrorMsg(pParse, "DRYRUN not supported for this operation");
+      goto trigger_cleanup;
+  }
   if( isTemp ){
     /* If TEMP was specified, then the trigger name may not be qualified. */
     if( pName2->n>0 ){
@@ -248,13 +252,17 @@
   pTrigger->pTabSchema = pTab->pSchema;
   pTrigger->op = (u8)op;
   pTrigger->tr_tm = tr_tm==TK_BEFORE ? TRIGGER_BEFORE : TRIGGER_AFTER;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   if( IN_RENAME_OBJECT ){
     sqlite3RenameTokenRemap(pParse, pTrigger->table, pTableName->a[0].zName);
     pTrigger->pWhen = pWhen;
     pWhen = 0;
   }else{
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
     pTrigger->pWhen = sqlite3ExprDup(db, pWhen, EXPRDUP_REDUCE);
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   pTrigger->pColumns = pColumns;
   pColumns = 0;
   assert( pParse->pNewTrigger==0 );
@@ -288,6 +296,11 @@
   int iDb;                                /* Database containing the trigger */
   Token nameToken;                        /* Trigger name for error reporting */
 
+
+  if(comdb2IsDryrun(pParse)){
+      sqlite3ErrorMsg(pParse, "DRYRUN not supported for this operation");
+      goto triggerfinish_cleanup;
+  }
   pParse->pNewTrigger = 0;
   if( NEVER(pParse->nErr) || !pTrig ) goto triggerfinish_cleanup;
   zName = pTrig->zName;
@@ -327,7 +340,11 @@
     z = sqlite3DbStrNDup(db, (char*)pAll->z, pAll->n);
     testcase( z==0 );
     sqlite3NestedParse(pParse,
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+       "INSERT INTO %Q.%s VALUES('trigger',%Q,%Q,0,'CREATE TRIGGER %q', NULL)",
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
        "INSERT INTO %Q.%s VALUES('trigger',%Q,%Q,0,'CREATE TRIGGER %q')",
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
        db->aDb[iDb].zDbSName, MASTER_NAME, zName,
        pTrig->table, z);
     sqlite3DbFree(db, z);
@@ -418,9 +435,11 @@
     pTriggerStep->zTarget = z;
     pTriggerStep->op = op;
     pTriggerStep->zSpan = triggerSpanDup(db, zStart, zEnd);
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
     if( IN_RENAME_OBJECT ){
       sqlite3RenameTokenMap(pParse, pTriggerStep->zTarget, pName);
     }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   return pTriggerStep;
 }
@@ -561,6 +580,10 @@
   const char *zName;
   sqlite3 *db = pParse->db;
 
+  if(comdb2IsDryrun(pParse)){
+      sqlite3ErrorMsg(pParse, "DRYRUN not supported for this operation");
+      goto drop_trigger_cleanup;
+  }
   if( db->mallocFailed ) goto drop_trigger_cleanup;
   if( SQLITE_OK!=sqlite3ReadSchema(pParse) ){
     goto drop_trigger_cleanup;
@@ -651,7 +674,11 @@
   assert( sqlite3SchemaMutexHeld(db, iDb, 0) );
   pHash = &(db->aDb[iDb].pSchema->trigHash);
   pTrigger = sqlite3HashInsert(pHash, zName, 0);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( pTrigger ){
+#else
   if( ALWAYS(pTrigger) ){
+#endif
     if( pTrigger->pSchema==pTrigger->pTabSchema ){
       Table *pTab = tableOfTrigger(pTrigger);
       Trigger **pp;
@@ -963,6 +990,9 @@
     transferParseError(pParse, pSubParse);
     if( db->mallocFailed==0 && pParse->nErr==0 ){
       pProgram->aOp = sqlite3VdbeTakeOpArray(v, &pProgram->nOp, &pTop->nMaxArg);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      sqlite3VdbeTransferTables(pParse->pVdbe, pSubParse->pVdbe);
+#endif
     }
     pProgram->nMem = pSubParse->nMem;
     pProgram->nCsr = pSubParse->nTab;
