--- sqlite-3.28.0/src/upsert.c	2026-01-09 16:11:01
+++ src/upsert.c	2025-09-24 15:21:58
@@ -15,6 +15,11 @@
 #include "sqliteInt.h"
 
 #ifndef SQLITE_OMIT_UPSERT
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+int is_comdb2_index_unique(const char *tbl, char *idx);
+int comdb2_get_index(const char *dbname, char *idx);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Free a list of Upsert objects
 */
@@ -136,8 +141,24 @@
   /* Check for matches against other indexes */
   for(pIdx=pTab->pIndex; pIdx; pIdx=pIdx->pNext){
     int ii, jj, nn;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    int is_comdb2_unique = is_comdb2_index_unique(pIdx->pTable->zName,
+                                                  pIdx->zName);
+    if( !is_comdb2_unique && !IsUniqueIndex(pIdx) ) continue;
+
+    /* Test whether its a DATACOPY index. */
+    int nCol = pIdx->nKeyCol;
+    for(int i=0; i<pIdx->nKeyCol; ++i ){
+        if( strcmp(pIdx->azColl[i], "DATACOPY")==0 ){
+            nCol = i;
+            break;
+        }
+    }
+    if( pTarget->nExpr!=nCol ) continue;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( !IsUniqueIndex(pIdx) ) continue;
     if( pTarget->nExpr!=pIdx->nKeyCol ) continue;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( pIdx->pPartIdxWhere ){
       if( pUpsert->pUpsertTargetWhere==0 ) continue;
       if( sqlite3ExprCompare(pParse, pUpsert->pUpsertTargetWhere,
@@ -145,7 +166,11 @@
         continue;
       }
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    nn = nCol;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     nn = pIdx->nKeyCol;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     for(ii=0; ii<nn; ii++){
       Expr *pExpr;
       sCol[0].u.zToken = (char*)pIdx->azColl[ii];
@@ -178,6 +203,10 @@
       continue;
     }
     pUpsert->pUpsertIdx = pIdx;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    int idx = comdb2_get_index(pIdx->pTable->zName, pIdx->zName);
+    comdb2SetUpsertIdx(pParse->pVdbe, idx);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     return SQLITE_OK;
   }
   sqlite3ErrorMsg(pParse, "ON CONFLICT clause does not match any "
