--- sqlite-3.28.0/src/util.c	2026-01-09 16:11:01
+++ src/util.c	2025-09-24 15:21:58
@@ -291,6 +291,39 @@
   sqlite3Dequote(p->u.zToken);
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/* return true if string is correctly quoted
+ * ie first and last char are a quote...
+ * note that string should be null terminated
+ */
+int sqlite3IsCorrectlyQuoted(const char *z){
+  char quote = z[0];
+  int i = 1;
+  if( !sqlite3Isquote(quote) ) return 1;
+  while( z[i]!='\0' ) i++;
+  if( i>1 ) return z[i-1]==quote;
+  return 0;
+}
+int sqlite3IsCorrectlyBraced(const char *z){
+  int i = 1;
+  int q = 0;
+  if( z[0]=='\0' ) return 0; /* empty string? */
+  while( z[i]!='\0' ){
+    if( z[i]=='{' || (z[i]=='}' && z[i+1]!='\0') ){ q++; }
+    i++;
+  }
+  if( q>0 ) return 0; /* braces inside content? */
+  if( i>1 ){
+    if( z[0]=='{' && z[i-1]=='}' ){
+      return 1;
+    }else if( z[0]!='{' && z[0]!='}' && z[i-1]!='{' && z[i-1]!='}' ){
+      return 1;
+    }
+  }
+  return 0;
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Generate a Token object from a string
 */
@@ -1087,6 +1120,14 @@
   ** by the getVarin32() macro */
   a = *p;
   /* a: p0 (unmasked) */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if (!(a&0x80))
+  {
+    /* Values between 0 and 127 */
+    *v = a;
+    return 1;
+  }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifndef getVarint32
   if (!(a&0x80))
   {
@@ -1095,6 +1136,7 @@
     return 1;
   }
 #endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   /* The 2-byte case */
   p++;
@@ -1249,6 +1291,70 @@
 }
 
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/*
+** Convert an eight-byte integer value to a double and vice versa.
+** This method should be safe; however, it may not be the fastest.
+*/
+i64 sqlite3DoubleToInt64(double rVal){
+  i64 iVal;
+  assert( sizeof(double)==sizeof(i64) );
+  memcpy(&iVal, &rVal, sizeof(i64));
+  return iVal;
+}
+double sqlite3Int64ToDouble(i64 iVal){
+  double rVal;
+  assert( sizeof(i64)==sizeof(double) );
+  memcpy(&rVal, &iVal, sizeof(double));
+  return rVal;
+}
+/*
+** Read or write an eight-byte big-endian integer value.
+*/
+u64 sqlite3Get8byte(const u8 *p){
+#if SQLITE_BYTEORDER==4321
+  u64 x;
+  memcpy(&x,p,8);
+  return x;
+#elif SQLITE_BYTEORDER==1234 && GCC_VERSION>=4003000
+  u64 x;
+  memcpy(&x,p,8);
+  return __builtin_bswap64(x);
+#elif SQLITE_BYTEORDER==1234 && MSVC_VERSION>=1300
+  u64 x;
+  memcpy(&x,p,8);
+  return _byteswap_uint64(x);
+#else
+  testcase( p[0]&0x80 );
+  testcase( p[1]&0x80 );
+  testcase( p[2]&0x80 );
+  testcase( p[3]&0x80 );
+  testcase( p[4]&0x80 );
+  return ((u64)p[0]<<56) | ((u64)p[1]<<48) | ((u64)p[2]<<40) |
+         ((u64)p[3]<<32) | ((u64)p[4]<<24) | (p[5]<<16) | (p[6]<<8) | p[7];
+#endif
+}
+void sqlite3Put8byte(unsigned char *p, u64 v){
+#if SQLITE_BYTEORDER==4321
+  memcpy(p,&v,8);
+#elif SQLITE_BYTEORDER==1234 && GCC_VERSION>=4003000
+  u64 x = __builtin_bswap64(v);
+  memcpy(p,&x,8);
+#elif SQLITE_BYTEORDER==1234 && MSVC_VERSION>=1300
+  u64 x = _byteswap_uint64(v);
+  memcpy(p,&x,8);
+#else
+  p[0] = (u8)(v>>56);
+  p[1] = (u8)(v>>48);
+  p[2] = (u8)(v>>40);
+  p[3] = (u8)(v>>32);
+  p[4] = (u8)(v>>24);
+  p[5] = (u8)(v>>16);
+  p[6] = (u8)(v>>8);
+  p[7] = (u8)v;
+#endif
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Translate a single byte of Hex into an integer.
@@ -1502,7 +1608,7 @@
   return a[x&7] + y - 10;
 }
 
-#ifndef SQLITE_OMIT_VIRTUALTABLE
+#if defined(SQLITE_BUILDING_FOR_COMDB2) || !defined(SQLITE_OMIT_VIRTUALTABLE)
 /*
 ** Convert a double into a LogEst
 ** In other words, compute an approximation for 10*log2(x).
@@ -1517,7 +1623,7 @@
   e = (a>>52) - 1022;
   return e*10;
 }
-#endif /* SQLITE_OMIT_VIRTUALTABLE */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) || !defined(SQLITE_OMIT_VIRTUALTABLE) */
 
 #if defined(SQLITE_ENABLE_STMT_SCANSTATUS) || \
     defined(SQLITE_ENABLE_STAT3_OR_STAT4) || \
@@ -1638,13 +1744,66 @@
 */
 int sqlite3VListNameToNum(VList *pIn, const char *zName, int nName){
   int i, mx;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  int off = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( pIn==0 ) return 0;
   mx = pIn[1];
   i = 2;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( zName[0]!='@' ) off++;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   do{
     const char *z = (const char*)&pIn[i+2];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( z && strncmp(z+off,zName,nName)==0 && z[nName+off]==0 ){
+      return pIn[i];
+    }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( strncmp(z,zName,nName)==0 && z[nName]==0 ) return pIn[i];
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     i += pIn[i+1];
   }while( i<mx );
   return 0;
 }
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/*
+** Create a copy of the VList 
+*/
+VList *sqlite3VListClone(
+  const VList *pIn,      /* The input VList.  Might be NULL */
+  void *(*alloc)(size_t size)
+){
+  VList *pOut;
+  if (!pIn) 
+    return NULL;
+
+  pOut = alloc(pIn[0]);
+  if (!pOut)
+    return NULL;
+
+  memcpy(pOut, pIn, pIn[0]);
+  return pOut;
+}
+
+#include "logmsg.h"
+void sqlite3VListPrint(loglvl lvl, const VList *pIn)
+{
+  int i, mx;
+
+  if (!pIn)
+    return;
+
+  mx = pIn[1];
+  i = 2;
+  logmsg(lvl, "%p VList info start:\n", (void *)pthread_self());
+  do{
+    const char *z = (const char*)&pIn[i+2];
+    logmsg(lvl, "%p %s %d\n", (void *)pthread_self(), z, pIn[i]);
+    i += pIn[i+1];
+  }while( i<mx );
+  logmsg(lvl, "%p VList info done.\n", (void *)pthread_self());
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
