--- sqlite-3.28.0/src/vdbe.c	2026-01-09 16:11:01
+++ src/vdbe.c	2025-11-06 15:49:51
@@ -20,7 +20,37 @@
 */
 #include "sqliteInt.h"
 #include "vdbeInt.h"
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include "comdb2vdbe.h"
+#include <ctype.h>
+#include <pthread.h>
+#include <strings.h>
+#include <sql.h>
 
+/* Comdb2 routines called from vdbe */
+void set_cook_fields(BtCursor *pCur, int cols);
+void sqlite3SetConversionError(void);
+void *get_lastkey(BtCursor *pCur);
+void print_cooked_access(BtCursor *pCur, int col);
+void comdb2SetWriteFlag(int wrflag);
+int is_datacopy(BtCursor *pCur, int *fnum);
+int get_datacopy(BtCursor *pCur, int fnum, Mem *m);
+int comdb2_is_idx_uniqnulls(BtCursor *);
+extern void comdb2_handle_limit(Vdbe*,Mem*);
+extern void sqlite3BtreeCursorSetFieldUsed(BtCursor *, unsigned long long);
+extern i64 sqlite3BtreeNewRowid(BtCursor *pCur);
+extern int sqlite3MakeRecordForComdb2(BtCursor *pCur, Mem *m, int nf, int *optimized);
+
+#define cur_is_raw(pCur)                               \
+    (pCur ?                                            \
+       (pCur->cursor_class == CURSORCLASS_TABLE  ||    \
+        pCur->cursor_class == CURSORCLASS_INDEX  ||    \
+        cur_is_remote(pCur)                      ||    \
+        pCur->is_sampled_idx)                          \
+     :  0)                                             \
+
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Invoke this macro on memory cells just prior to changing the
 ** value of the cell.  This macro verifies that shallow copies are
@@ -36,6 +66,47 @@
 # define memAboutToChange(P,M)
 #endif
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/* Removing duplicate code between OP_Rowid and OP_IdxRowid */
+void getRowid(BtCursor *pCursor, i64 rowId, u8 p3, Mem *pOut)
+{
+  if( p3==0 ){
+    pOut->u.i = rowId;
+    MemSetTypeFlag(pOut, MEM_Int);
+    return;
+  }
+  if( pCursor==NULL ){
+    MemSetTypeFlag(pOut, MEM_Null);
+    return;
+  }
+  if( p3==1 ){
+    char *zRowId = 0;
+    int nRowId = 0;
+    if( sqlite3BtreeGetGenId(rowId, 0, &zRowId, &nRowId)!=SQLITE_OK ){
+      assert( zRowId==0 );
+      assert( nRowId==0 );
+      MemSetTypeFlag(pOut, MEM_Null);
+      return;
+    }
+    assert( nRowId>=4 ); /* MIN: "2:0\0" */
+    assert( zRowId[0]=='2' ); /* REQD: For R6 compat. */
+    assert( zRowId[1]==':' ); /* REQD: For R6 compat. */
+    sqlite3VdbeMemSetStr(pOut, zRowId, nRowId, SQLITE_UTF8, sqlite3_free);
+    return;
+  }
+  if( p3==2 ){
+    unsigned long long genId = 0;
+    if( sqlite3BtreeGetGenId(rowId, &genId, 0, 0)!=SQLITE_OK ){
+      MemSetTypeFlag(pOut, MEM_Null);
+      return;
+    }
+    pOut->u.i = (genId & 0xffffffff00000000ull) >> 32;
+    MemSetTypeFlag(pOut, MEM_Int);
+    sqlite3VdbeMemDatetimefy(pOut);
+  }
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** The following global variable is incremented every time a cursor
 ** moves, either by the OP_SeekXX, OP_Next, or OP_Prev opcodes.  The test
@@ -221,7 +292,25 @@
 /* Return true if the cursor was opened using the OP_OpenSorter opcode. */
 #define isSorter(x) ((x)->eCurType==CURTYPE_SORTER)
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 /*
+** Copy the updCols array into the vdbe machine.  Exposing this to comdb2's
+** higher layers allows us to optimize the update codepath.
+*/
+void sqlite3CreateUpdCols(
+    Vdbe *v, 
+    sqlite3 *db, 
+    int cnt, 
+    int *cols
+){
+  v->updCols = sqlite3DbRealloc(db, v->updCols, sizeof(int) * (cnt + 1));
+  if( v->updCols==0 ) return;
+  v->updCols[0] = cnt;
+  memcpy(&v->updCols[1], cols, sizeof(int) * cnt);
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
+/*
 ** Allocate VdbeCursor number iCur.  Return a pointer to it.  Return NULL
 ** if we run out of memory.
 */
@@ -280,6 +369,9 @@
           &pMem->z[ROUND8(sizeof(VdbeCursor))+2*sizeof(u32)*nField];
       sqlite3BtreeCursorZero(pCx->uc.pCursor);
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    pCx->nCookFields = -1;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   return pCx;
 }
@@ -343,13 +435,40 @@
   char affinity,      /* The affinity to be applied */
   u8 enc              /* Use this text encoding */
 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* FIXME TODO XXX: Remove this test when done */
+  if (islower(affinity))
+  {
+      logmsg(LOGMSG_FATAL, "PANIC: lower case affinity: %c\n", affinity);
+      cheap_stack_trace();
+      exit(99);
+  }
+  if( affinity==SQLITE_AFF_DECIMAL ){
+    sqlite3VdbeMemDecimalfy(pRec);
+    return;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( affinity>=SQLITE_AFF_NUMERIC ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
     assert( affinity==SQLITE_AFF_INTEGER || affinity==SQLITE_AFF_REAL
+             || affinity==SQLITE_AFF_NUMERIC || affinity==SQLITE_AFF_SMALL);
+    if( affinity==SQLITE_AFF_SMALL ){
+      pRec->flags |= MEM_Small;
+    }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+    assert( affinity==SQLITE_AFF_INTEGER || affinity==SQLITE_AFF_REAL
              || affinity==SQLITE_AFF_NUMERIC );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( (pRec->flags & MEM_Int)==0 ){ /*OPTIMIZATION-IF-FALSE*/
       if( (pRec->flags & MEM_Real)==0 ){
         if( pRec->flags & MEM_Str ) applyNumericAffinity(pRec,1);
       }else{
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        /* Mem is MEM_Real (double) and we want float */
+        if( affinity==SQLITE_AFF_SMALL ){
+          pRec->u.r = (float)pRec->u.r;
+        }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         sqlite3VdbeIntegerAffinity(pRec);
       }
     }
@@ -362,9 +481,39 @@
     if( 0==(pRec->flags&MEM_Str) ){ /*OPTIMIZATION-IF-FALSE*/
       if( (pRec->flags&(MEM_Real|MEM_Int)) ){
         sqlite3VdbeMemStringify(pRec, enc, 1);
+      }
+    }
+    pRec->flags &= ~(MEM_Real|MEM_Int);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( 0==(pRec->flags&MEM_Blob) &&
+              ((pRec->flags&MEM_Datetime) || (pRec->flags&MEM_Interval)) ) 
+    {
+      if( pRec->flags&MEM_Str ){
+        /* get rid of the string;
+           this will add the cost of datetime to string conversion (acceptable)
+           AND this will correct the partial specified user string input
+           and return a fully formatted string result
+        */
+        sqlite3DbFree(pRec->db, pRec->z);
+        if(pRec->z == pRec->zMalloc)
+            pRec->zMalloc = NULL;
+        pRec->z = NULL;
+        pRec->n = 0;
+        pRec->flags &= ~MEM_Str; /* unset MEM_Str bit */
       }
+      sqlite3VdbeMemStringify(pRec, enc, 1);
     }
     pRec->flags &= ~(MEM_Real|MEM_Int);
+  }else if(affinity==SQLITE_AFF_DATETIME){
+      sqlite3VdbeMemDatetimefy(pRec);
+  }else if( (affinity==SQLITE_AFF_INTV_YE) ||
+          (affinity==SQLITE_AFF_INTV_MO) ||
+          (affinity==SQLITE_AFF_INTV_DY) ||
+          (affinity==SQLITE_AFF_INTV_HO) ||
+          (affinity==SQLITE_AFF_INTV_MI) ||
+          (affinity==SQLITE_AFF_INTV_SE) ){
+      sqlite3VdbeMemIntervalfy(pRec, affinity);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
 }
 
@@ -396,6 +545,15 @@
   applyAffinity((Mem *)pVal, affinity, enc);
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+static inline void setCookCol(VdbeCursor *pC, int nCol){
+  if( nCol > pC->nCookFields || nCol < 0 ){
+    pC->nCookFields = nCol;
+    set_cook_fields(pC->uc.pCursor, nCol);
+  }
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** pMem currently only holds a string type (or maybe a BLOB that we can
 ** interpret as a string if we want to).  Compute its corresponding
@@ -405,7 +563,13 @@
 static u16 SQLITE_NOINLINE computeNumericType(Mem *pMem){
   assert( (pMem->flags & (MEM_Int|MEM_Real))==0 );
   assert( (pMem->flags & (MEM_Str|MEM_Blob))!=0 );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#ifndef SQLITE_OMIT_INCRBLOB
   ExpandBlob(pMem);
+#endif
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+  ExpandBlob(pMem);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( sqlite3AtoF(pMem->z, &pMem->u.r, pMem->n, pMem->enc)==0 ){
     return 0;
   }
@@ -432,7 +596,7 @@
   return 0;
 }
 
-#ifdef SQLITE_DEBUG
+#if defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_DEBUG)
 /*
 ** Write a nice string representation of the contents of cell pMem
 ** into buffer zBuf, length nBuf.
@@ -509,41 +673,83 @@
     zBuf[k++] = 0;
   }
 }
-#endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_DEBUG) */
 
-#ifdef SQLITE_DEBUG
+#if defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_DEBUG)
 /*
 ** Print the value of a register for tracing purposes:
 */
 static void memTracePrint(Mem *p){
   if( p->flags & MEM_Undefined ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_USER, " undefined");
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     printf(" undefined");
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }else if( p->flags & MEM_Null ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_USER, p->flags & MEM_Zero ? " NULL-nochng" : " NULL");
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     printf(p->flags & MEM_Zero ? " NULL-nochng" : " NULL");
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }else if( (p->flags & (MEM_Int|MEM_Str))==(MEM_Int|MEM_Str) ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_USER, " si:%lld", p->u.i);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     printf(" si:%lld", p->u.i);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }else if( p->flags & MEM_Int ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_USER, " i:%lld", p->u.i);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     printf(" i:%lld", p->u.i);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifndef SQLITE_OMIT_FLOATING_POINT
   }else if( p->flags & MEM_Real ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_USER, " r:%g", p->u.r);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     printf(" r:%g", p->u.r);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
   }else if( sqlite3VdbeMemIsRowSet(p) ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_USER, " (rowset)");
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     printf(" (rowset)");
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }else{
     char zBuf[200];
     sqlite3VdbeMemPrettyPrint(p, zBuf);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_USER, " %s", zBuf);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     printf(" %s", zBuf);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( p->flags & MEM_Subtype ) logmsg(LOGMSG_USER, " subtype=0x%02x", p->eSubtype);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( p->flags & MEM_Subtype ) printf(" subtype=0x%02x", p->eSubtype);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 static void registerTrace(int iReg, Mem *p){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  logmsg(LOGMSG_USER, "REG[%d] = ", iReg);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   printf("REG[%d] = ", iReg);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   memTracePrint(p);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  logmsg(LOGMSG_USER, "\n");
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   printf("\n");
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_DEBUG)
   sqlite3VdbeCheckMemInvariants(p);
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_DEBUG) */
 }
-#endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_DEBUG) */
 
 #ifdef SQLITE_DEBUG
 #  define REGISTER_TRACE(R,M) if(db->flags&SQLITE_VdbeTrace)registerTrace(R,M)
@@ -579,9 +785,123 @@
   for(p=db->pSavepoint; p; p=p->pNext) n++;
   assert( n==(db->nSavepoint + db->isTransactionSavepoint) );
   return 1;
+}
+#endif
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+void dump_sqlite_mem(Mem *m){
+  logmsg(LOGMSG_USER, "flags %08x ", m->flags);
+  //TODO: old: printf("flags %08x type %d ", m->flags, m->type);
+  if( m->flags & MEM_Null) {
+    logmsg(LOGMSG_USER, "null ");
+  }
+  if( m->flags & MEM_Int) {
+    logmsg(LOGMSG_USER, "int %lld %016llx", m->u.i, m->u.i);
+  }else if( m->flags & MEM_Real ){
+    logmsg(LOGMSG_USER, "real %f", m->u.r);
+  }else if( m->flags & MEM_Str ){
+    logmsg(LOGMSG_USER, "string \"%.*s\"", m->n, m->z);
+  }else if( m->flags & MEM_Datetime ){
+    if (m->du.dt.dttz_prec == DTTZ_PREC_MSEC)
+      logmsg(LOGMSG_USER, "datetime ");
+    else
+      logmsg(LOGMSG_USER, "datetimeus ");
+  }else if( m->flags & MEM_Interval ){
+    switch( m->du.tv.type ){
+      case INTV_YM_TYPE: {
+        logmsg(LOGMSG_USER, "intervalym ");
+        break;
+      }
+      case INTV_DS_TYPE: {
+        logmsg(LOGMSG_USER, "intervalds ");
+        break;
+      }
+      case INTV_DSUS_TYPE: {
+        logmsg(LOGMSG_USER, "intervaldsus ");
+        break;
+      }
+      case INTV_DECIMAL_TYPE: {
+        logmsg(LOGMSG_USER, "decimal ");
+        break;
+      }
+      default: {
+        logmsg(LOGMSG_USER, "???interval??? ");
+        break;
+      }
+    }
+  }else if( m->flags & MEM_Blob ){
+    logmsg(LOGMSG_USER, "blob: \n");
+  }
+  logmsg(LOGMSG_USER, "\n");
+}
+
+int gbl_debug_sql_opcodes = 0;
+
+void dump_vdbe_mem(FILE *f, Mem *m){
+  //TODO: old: fprintf(f, "type %d ", m->type);
+  if( m->flags & MEM_Null ){
+    logmsgf(LOGMSG_USER, f, "NULL");
+  }else if( m->flags & MEM_Str ){
+    logmsgf(LOGMSG_USER, f, "string: \"%.*s\"", m->n, m->z);
+  }else if( m->flags & MEM_Int ){
+    logmsgf(LOGMSG_USER, f, "int: %lld", m->u.i);
+  }else if( m->flags & MEM_Real ){
+    logmsgf(LOGMSG_USER, f, "double: %f", m->u.r);
+  }else if( m->flags & MEM_Datetime){
+    char ctimebuf[30] = "";
+    time_t ts;
+    ts = (int) m->du.dt.dttz_sec;
+    ctime_r(&ts, ctimebuf);
+    if( m->du.dt.dttz_prec == DTTZ_PREC_MSEC ){
+      logmsgf(LOGMSG_USER, f, "datetime: epoch %lld msec %u \"%s\"",
+            m->du.dt.dttz_sec, m->du.dt.dttz_frac, ctimebuf);
+    }else{
+      logmsgf(LOGMSG_USER, f, "datetime: epoch %lld usec %u \"%s\"",
+            m->du.dt.dttz_sec, m->du.dt.dttz_frac, ctimebuf);
+    }
+  }else if( m->flags & MEM_Interval ){
+    logmsgf(LOGMSG_USER, f, "interval: type %u", m->du.tv.type);
+  }else{
+    logmsgf(LOGMSG_USER, f, "unknown type, flags %x", (int) m->flags);
+  }
+  logmsgf(LOGMSG_USER, f, " encoding: ");
+  switch( m->enc ){
+    case SQLITE_UTF8: {
+      logmsgf(LOGMSG_USER, f, "UTF8");
+      break;
+    }
+    case SQLITE_UTF16LE: {
+      logmsgf(LOGMSG_USER, f, "UTF16(le)");
+      break;
+    }
+    case SQLITE_UTF16BE: {
+      logmsgf(LOGMSG_USER, f, "UTF16(be)");
+      break;
+    }
+    default: {
+      logmsgf(LOGMSG_USER, f, "unknown");
+      break;
+    }
+  }
+  logmsgf(LOGMSG_USER, f, " cleanupfunc: 0x%p", m->xDel);
+  logmsgf(LOGMSG_USER, f, " tz: \"%s\"", m->tz);
+  logmsgf(LOGMSG_USER, f, "\n");
 }
+
+
+#ifdef SQLITE_DEBUG
+/*
+** Return true if the cursor has a hint specified.  This routine is
+** only used from within assert() statements
+*/
+int sqlite3BtreeCursorHasHint(BtCursor *pCsr, unsigned int mask){
+  //return (pCsr->hints & mask)!=0;
+  return 0;
+}
 #endif
 
+#include <vdbecompare.c>
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 /*
 ** Return the register of pOp->p2 after first preparing it to be
 ** overwritten with an integer value.
@@ -620,6 +940,9 @@
 #endif
 #ifdef SQLITE_DEBUG
   int nExtraDelete = 0;      /* Verifies FORDELETE and AUXDELETE flags */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  UNUSED_PARAMETER(nExtraDelete);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
   int rc = SQLITE_OK;        /* Value to return */
   sqlite3 *db = p->db;       /* The database */
@@ -641,6 +964,9 @@
   /*** INSERT STACK UNION HERE ***/
 
   assert( p->magic==VDBE_MAGIC_RUN );  /* sqlite3_step() verifies this */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( p->db );  /* db should not be null */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   sqlite3VdbeEnter(p);
 #ifndef SQLITE_OMIT_PROGRESS_CALLBACK
   if( db->xProgress ){
@@ -673,7 +999,11 @@
     int once = 1;
     sqlite3VdbePrintSql(p);
     if( p->db->flags & SQLITE_VdbeListing ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      logmsg(LOGMSG_USER, "VDBE Program Listing:\n");
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       printf("VDBE Program Listing:\n");
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       for(i=0; i<p->nOp; i++){
         sqlite3VdbePrintOp(stdout, i, &aOp[i]);
       }
@@ -681,13 +1011,22 @@
     if( p->db->flags & SQLITE_VdbeEQP ){
       for(i=0; i<p->nOp; i++){
         if( aOp[i].opcode==OP_Explain ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          if( once ) logmsg(LOGMSG_USER, "VDBE Query Plan:\n");
+          logmsg(LOGMSG_USER, "%s\n", aOp[i].p4.z);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           if( once ) printf("VDBE Query Plan:\n");
           printf("%s\n", aOp[i].p4.z);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           once = 0;
         }
       }
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( p->db->flags & SQLITE_VdbeTrace )  logmsg(LOGMSG_USER, "VDBE Trace:\n");
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( p->db->flags & SQLITE_VdbeTrace )  printf("VDBE Trace:\n");
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   sqlite3EndBenignMalloc();
 #endif
@@ -726,6 +1065,12 @@
     }
 #endif
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( gbl_debug_sql_opcodes ){
+      logmsg(LOGMSG_USER, "tid 0x%p step %d pc %d op %d %s\n", (void *)pthread_self(), nVmStep,
+                (int)(pOp - aOp), pOp->opcode, sqlite3OpcodeName(pOp->opcode));
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     /* Sanity checking on other operands */
 #ifdef SQLITE_DEBUG
     {
@@ -763,10 +1108,17 @@
       }
     }
 #endif
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /* TODO: Why are we clearing the error message here? */
+    if( p->zErrMsg ){
+      sqlite3DbFree(db, p->zErrMsg);
+      p->zErrMsg = NULL;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #if defined(SQLITE_DEBUG) || defined(VDBE_PROFILE)
     pOrigOp = pOp;
 #endif
-  
+
     switch( pOp->opcode ){
 
 /*****************************************************************************
@@ -1056,10 +1408,18 @@
   assert( rc==SQLITE_BUSY || rc==SQLITE_OK || rc==SQLITE_ERROR );
   if( rc==SQLITE_BUSY ){
     p->rc = SQLITE_BUSY;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    logmsg(LOGMSG_ERROR, "%s:%d SQLITE_BUSY\n", __FILE__, __LINE__);
+    cheap_stack_trace();
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }else{
     assert( rc==SQLITE_OK || (p->rc&0xff)==SQLITE_CONSTRAINT );
     assert( rc==SQLITE_OK || db->nDeferredCons>0 || db->nDeferredImmCons>0 );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    rc = p->rc ? p->rc : SQLITE_DONE;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     rc = p->rc ? SQLITE_ERROR : SQLITE_DONE;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   goto vdbe_return;
 }
@@ -1261,6 +1621,13 @@
   pOut->flags &= ~(MEM_Dyn|MEM_Ephem);
   pOut->flags |= MEM_Static|MEM_FromBind;
   UPDATE_MAX_BLOBSIZE(pOut);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( gbl_debug_sql_opcodes ){
+    char *s = (char*) sqlite3_bind_parameter_name((sqlite3_stmt*) p, pOp->p1);
+    logmsg(LOGMSG_USER, "variable %d (%s): ", pOp->p1, s ? s : "???");
+    dump_vdbe_mem(stdout, pOut);
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   break;
 }
 
@@ -1373,6 +1740,9 @@
   assert( (pIn1->flags & MEM_Int)!=0 );
   pOut = &aMem[pOp->p2];
   sqlite3VdbeMemSetInt64(pOut, pIn1->u.i);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  comdb2_handle_limit(p, pIn1);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   break;
 }
 
@@ -1555,6 +1925,120 @@
   type2 = numericType(pIn2);
   pOut = &aMem[pOp->p3];
   flags = pIn1->flags | pIn2->flags;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( pOut->db );
+
+  /* SQLite may reuse a Mem object. If the Mem object has dynamically allocated
+     memory (zMalloc), we free it here. If we do not free it here, zMalloc will
+     be overwritten to NULL by one of the arithmetic routines below and the memory
+     will be leaked. */
+  if( pOut->szMalloc>0 ){
+    sqlite3DbFreeNN(pOut->db, pOut->zMalloc);
+    pOut->szMalloc = 0;
+    pOut->z = pOut->zMalloc = NULL;
+  }
+
+  /* Subtraction only: If one operand is a String and the other is a Datetime,
+     attempt to promote the String operand to Datetime. */
+  if( ( pOp->opcode==OP_Subtract )&&( (pIn1->flags|pIn2->flags)&MEM_TypeMask )==(MEM_Str|MEM_Datetime) ) {
+    /* If pIn1 is a Datetime, the line below is a no-op; If pIn2 isn't
+       a valid Datetime string, the 2nd Datetimefy will fail and both
+       operands' types remain intact; If pIn1 is a String, but isn't a
+       valid Datetime string, the line below will fail. Both operands'
+       types are preserved, too. */
+    const char *implicit_tz = ( pIn1->tz!=NULL ) ?  pIn1->tz : pIn2->tz;
+    int implicit_conv_rc = sqlite3VdbeMemDatetimefyTz(pIn1, implicit_tz);
+    if ( implicit_conv_rc==0 )
+      sqlite3VdbeMemDatetimefyTz(pIn2, implicit_tz);
+  }
+
+  if( (pIn1->flags & MEM_Interval) && pIn1->du.tv.type == INTV_DECIMAL_TYPE)
+  {
+    Mem res;
+    rc = sqliteVdbeMemDecimalBasicArithmetics( pIn1, pIn2, pOp->opcode, &res, 1);
+    if( rc!=SQLITE_OK ) goto abort_due_to_error;
+    *pOut = res;
+    if (rc)
+    {
+      sqlite3SetString(&p->zErrMsg, db, "failed decimal arithmetics, unsupported conversion required");
+      sqlite3SetConversionError();
+    }
+  }
+  else if( (pIn2->flags & MEM_Interval) && pIn2->du.tv.type == INTV_DECIMAL_TYPE)
+  {
+    Mem res;
+    rc = sqliteVdbeMemDecimalBasicArithmetics( pIn2, pIn1, pOp->opcode, &res, 0);
+    if( rc!=SQLITE_OK ) goto abort_due_to_error;
+    *pOut = res;
+    if (rc)
+    {
+      sqlite3SetString(&p->zErrMsg, db, "failed decimal arithmetics, unsupported conversion required");
+      sqlite3SetConversionError();
+    }
+  }
+  else if( ((pIn1->flags & pIn2->flags & MEM_Interval)==MEM_Interval) &&
+            (
+             pIn1->du.tv.type == pIn2->du.tv.type ||
+             (pIn1->du.tv.type == INTV_DS_TYPE && pIn2->du.tv.type == INTV_DSUS_TYPE) ||
+             (pIn1->du.tv.type == INTV_DSUS_TYPE && pIn2->du.tv.type == INTV_DS_TYPE)
+             ) &&
+            ((pOp->opcode == OP_Add) || (pOp->opcode == OP_Subtract))){
+  
+    /* two intervals of the same type or 1 intervalds and 1 intervaldsus */
+    Mem res;
+    rc = sqlite3VdbeMemIntervalAndInterval(pIn2, pIn1, pOp->opcode, &res);
+    if( rc!=SQLITE_OK ) goto abort_due_to_error;
+    *pOut = res; /*knowledge of _operateIntervalandInterval is assumed, no dyn alloc */
+  }else if( (pIn2->flags & MEM_Interval) && (pIn1->flags & MEM_Int) &&
+            ((pOp->opcode == OP_Multiply) || (pOp->opcode == OP_Divide))){
+  
+    /* interval and int */
+    Mem res;
+    rc = sqlite3VdbeMemIntervalAndInt(pIn2, pIn1, pOp->opcode, &res);
+    if( rc!=SQLITE_OK ) goto abort_due_to_error;
+    *pOut = res; /*knowledge of _operateIntervalandInterval is assumed, no dyn alloc */
+  }else if( (pIn2->flags & MEM_Int) && (pIn1->flags & MEM_Interval)
+   && (pOp->opcode == OP_Multiply)
+  ){
+  
+    /* int and interval */
+    Mem res;
+    rc = sqlite3VdbeMemIntAndInterval(pIn2, pIn1, pOp->opcode, &res);
+    if( rc!=SQLITE_OK ) goto abort_due_to_error;
+    *pOut = res; /*knowledge of _operateIntervalandInterval is assumed, no dyn alloc */
+  }
+  else if( (pIn2->flags & MEM_Datetime) && (pIn1->flags & MEM_Datetime)
+   && (pOp->opcode == OP_Subtract)
+  ){
+  
+    /* two datetimes*/
+    Mem res;
+    rc = sqlite3VdbeMemDatetimeAndDatetime(pIn2, pIn1, pOp->opcode, &res);
+    if( rc!=SQLITE_OK ) goto abort_due_to_error;
+    *pOut = res; /*knowledge of _operateIntervalandInterval is assumed, no dyn alloc */
+  }else if( (pIn2->flags & MEM_Datetime) && (pIn1->flags & MEM_Interval)
+   && ((pOp->opcode == OP_Add) || (pOp->opcode == OP_Subtract))
+  ){
+  
+    /* datetime and interval*/
+    Mem res;
+    rc = sqlite3VdbeMemDatetimeAndInterval(pIn2, pIn1, pOp->opcode, &res);
+    if( rc!=SQLITE_OK ) goto abort_due_to_error;
+    *pOut = res; /*knowledge of _operateIntervalandInterval is assumed, no dyn alloc */
+  }else if( (pIn2->flags & MEM_Interval) && (pIn1->flags & MEM_Datetime) ){
+
+    /* datetime and interval*/
+    if( pOp->opcode == OP_Add ){ 
+      Mem res;
+      rc = sqlite3VdbeMemIntervalAndDatetime(pIn2, pIn1, pOp->opcode, &res);
+      if( rc!=SQLITE_OK ) goto abort_due_to_error;
+      *pOut = res; /*knowledge of _operateIntervalandInterval is assumed, no dyn alloc */
+    }else{
+      rc = SQLITE_MISMATCH;
+      goto abort_due_to_error;
+    }
+  }else
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( (type1 & type2 & MEM_Int)!=0 ){
     iA = pIn1->u.i;
     iB = pIn2->u.i;
@@ -1618,6 +2102,9 @@
     }
 #endif
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( !pOut->db ) pOut->db = db; /* wiped from datetime funcs */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   break;
 
 arithmetic_result_is_null:
@@ -1808,18 +2295,34 @@
 ** A NULL value is not changed by this routine.  It remains NULL.
 */
 case OP_Cast: {                  /* in1 */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( pOp->p2>=SQLITE_AFF_BLOB && pOp->p2<=SQLITE_AFF_SMALL );
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( pOp->p2>=SQLITE_AFF_BLOB && pOp->p2<=SQLITE_AFF_REAL );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   testcase( pOp->p2==SQLITE_AFF_TEXT );
   testcase( pOp->p2==SQLITE_AFF_BLOB );
   testcase( pOp->p2==SQLITE_AFF_NUMERIC );
   testcase( pOp->p2==SQLITE_AFF_INTEGER );
   testcase( pOp->p2==SQLITE_AFF_REAL );
   pIn1 = &aMem[pOp->p1];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if(pOp->p2 == SQLITE_AFF_DATETIME) {
+    pIn1->tz = &p->tzname[0];
+    pIn1->dtprec = p->dtprec;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   memAboutToChange(p, pIn1);
   rc = ExpandBlob(pIn1);
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   sqlite3VdbeMemCast(pIn1, pOp->p2, encoding);
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   UPDATE_MAX_BLOBSIZE(pIn1);
   if( rc ) goto abort_due_to_error;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  rc = sqlite3VdbeMemCast(p, pIn1, pOp->p2, encoding);
+  if( rc ) goto abort_due_to_error;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   break;
 }
 #endif /* SQLITE_OMIT_CAST */
@@ -1975,6 +2478,17 @@
     /* Neither operand is NULL.  Do a comparison. */
     affinity = pOp->p5 & SQLITE_AFF_MASK;
     if( affinity>=SQLITE_AFF_NUMERIC ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( affinity==SQLITE_AFF_DECIMAL ){
+        applyAffinity(pIn1, affinity, encoding);
+        applyAffinity(pIn3, affinity, encoding);
+        goto val_nn_done;
+      }
+      if( affinity==SQLITE_AFF_SMALL ){
+        pIn1->flags |= MEM_Small;
+        pIn3->flags |= MEM_Small;
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       if( (flags1 | flags3)&MEM_Str ){
         if( (flags1 & (MEM_Int|MEM_Real|MEM_Str))==MEM_Str ){
           applyNumericAffinity(pIn1,0);
@@ -1998,6 +2512,13 @@
         res = 0;
         goto compare_op;
       }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    }else if( affinity>SQLITE_AFF_BLOB && affinity<SQLITE_AFF_NUMERIC ){
+      applyAffinity(pIn1, affinity, encoding);
+      applyAffinity(pIn3, affinity, encoding);
+    }
+val_nn_done:
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }else if( affinity==SQLITE_AFF_TEXT ){
       if( (flags1 & MEM_Str)==0 && (flags1 & (MEM_Int|MEM_Real))!=0 ){
         testcase( pIn1->flags & MEM_Int );
@@ -2015,8 +2536,16 @@
         flags3 = (pIn3->flags & ~MEM_TypeMask) | (flags3 & MEM_TypeMask);
       }
     }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( pOp->p4type==P4_COLLSEQ || pOp->p4.pColl==0 );
     res = sqlite3MemCompare(pIn3, pIn1, pOp->p4.pColl);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( p->rc!=SQLITE_OK ){
+      rc = p->rc;
+      p->rc = 0;
+      goto abort_due_to_error;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
 compare_op:
   /* At this point, res is negative, zero, or positive if reg[P1] is
@@ -2038,16 +2567,37 @@
     res2 = aGTb[pOp->opcode - OP_Ne];
   }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* Undo any changes made by applyAffinity() to the input registers.
+  **
+  ** However, if the operand 1) was previously a string, 2) is compared with
+  ** a non-string affinity type, 3) and has lost its `Mem.z` during the type
+  ** conversion (e.g., Decimalfy or Datetimefy), do not change back its type. */
+
+  if( (flags1 & MEM_Str)==0 || (pIn1->flags & MEM_Str)!=0 || pIn1->z!=NULL ){
+    assert( (pIn1->flags & MEM_Dyn) == (flags1 & MEM_Dyn) );
+    pIn1->flags = flags1;
+  }
+  if( (flags3 & MEM_Str)==0 || (pIn3->flags & MEM_Str)!=0 || pIn3->z!=NULL ){
+    assert( (pIn3->flags & MEM_Dyn) == (flags3 & MEM_Dyn) );
+    pIn3->flags = flags3;
+  }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   /* Undo any changes made by applyAffinity() to the input registers. */
   assert( (pIn1->flags & MEM_Dyn) == (flags1 & MEM_Dyn) );
   pIn1->flags = flags1;
   assert( (pIn3->flags & MEM_Dyn) == (flags3 & MEM_Dyn) );
   pIn3->flags = flags3;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   if( pOp->p5 & SQLITE_STOREP2 ){
     pOut = &aMem[pOp->p2];
     iCompare = res;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( (pOp->p5 & SQLITE_KEEPNULL)==SQLITE_KEEPNULL ){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( (pOp->p5 & SQLITE_KEEPNULL)!=0 ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       /* The KEEPNULL flag prevents OP_Eq from overwriting a NULL with 1
       ** and prevents OP_Ne from overwriting NULL with 0.  This flag
       ** is only used in contexts where either:
@@ -2483,6 +3033,10 @@
 ** skipped for length() and all content loading can be skipped for typeof().
 */
 case OP_Column: {
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  i64 payloadSize64; /* Number of bytes in the record */
+  int datacopy;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   int p2;            /* column number to retrieve */
   VdbeCursor *pC;    /* The VDBE cursor */
   BtCursor *pCrsr;   /* The BTree cursor */
@@ -2518,6 +3072,47 @@
   assert( pC->eCurType!=CURTYPE_PSEUDO || pC->nullRow );
   assert( pC->eCurType!=CURTYPE_SORTER );
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  pCrsr = pC->uc.pCursor;
+  if( pC->eCurType == CURTYPE_BTREE && cur_is_raw(pCrsr) && !pC->nullRow ) {
+    /* We may reuse a Mem structure.
+       So delete any previously allocated memory in pDest. */
+    sqlite3VdbeMemRelease(pDest); /* takes care of both z and zMalloc */
+    if(cur_is_remote(pCrsr)) {
+      goto cooked_access;
+    }
+    else if( pC->isTable ){
+      zData = (u8 *)sqlite3BtreeDataFetch(pCrsr, &pC->szRow);
+      assert(zData != NULL);
+      rc = get_data(pCrsr, pCrsr->sc, (u8 *) zData, p2, pDest, 0, pCrsr->clnt->tzname);
+    }else{
+      datacopy = p2;
+      if( is_datacopy(pCrsr, &datacopy) ){
+        rc = get_datacopy(pCrsr, datacopy, pDest);
+      }else if(pC->nCookFields>=0 && p2>=pC->nCookFields){
+        zData = (u8 *)get_lastkey(pCrsr);
+        rc = get_data(pCrsr, pCrsr->sc, (u8 *) zData, p2, pDest, 0, pCrsr->clnt->tzname);
+      }else{
+        goto cooked_access;
+      }
+    }
+
+    if( rc ){
+      logmsg(LOGMSG_ERROR, "get_data failed for field: %d, rc %d\n", p2, rc);
+      goto abort_due_to_error;
+    }
+
+    pDest->db = p->db;
+    pDest->enc = encoding;
+    rc = sqlite3VdbeMemMakeWriteable(pDest);
+    goto op_column_out;
+  }
+
+cooked_access:
+  #ifdef debug_raw
+  print_cooked_access(pCrsr, p2);
+  #endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( pC->cacheStatus!=p->cacheCtr ){                /*OPTIMIZATION-IF-FALSE*/
     if( pC->nullRow ){
       if( pC->eCurType==CURTYPE_PSEUDO ){
@@ -2538,10 +3133,27 @@
       assert( pC->eCurType==CURTYPE_BTREE );
       assert( pCrsr );
       assert( sqlite3BtreeCursorIsValid(pCrsr) );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( pC->isTable==0 ){
+        payloadSize64 = sqlite3BtreeIntegerKey(pCrsr);
+        /* sqlite3BtreeParseCellPtr() uses getVarint32() to extract the
+        ** payload size, so it is impossible for payloadSize64 to be
+        ** larger than 32 bits. */
+        assert( (payloadSize64 & SQLITE_MAX_U32)==(u64)payloadSize64 );
+        pC->aRow = sqlite3BtreeKeyFetch(pCrsr, &pC->szRow);
+        pC->payloadSize = (u32)payloadSize64;
+      }else{
+        pC->payloadSize = sqlite3BtreePayloadSize(pCrsr);
+        pC->aRow = sqlite3BtreeDataFetch(pCrsr, &pC->szRow);
+      }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       pC->payloadSize = sqlite3BtreePayloadSize(pCrsr);
       pC->aRow = sqlite3BtreePayloadFetch(pCrsr, &pC->szRow);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       assert( pC->szRow<=pC->payloadSize );
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
       assert( pC->szRow<=65536 );  /* Maximum page size is 64KiB */
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
       if( pC->payloadSize > (u32)db->aLimit[SQLITE_LIMIT_LENGTH] ){
         goto too_big;
       }
@@ -2684,7 +3296,11 @@
     /* This is the common case where the desired content fits on the original
     ** page - where the content is not on an overflow page */
     zData = pC->aRow + aOffset[p2];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( sqlite3IsFixedLengthSerialType(t) ){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( t<12 ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3VdbeSerialGet(zData, t, pDest);
     }else{
       /* If the column value is a string, we need a persistent value, not
@@ -2709,7 +3325,12 @@
     pDest->enc = encoding;
     /* This branch happens only when content is on overflow pages */
     if( ((pOp->p5 & (OPFLAG_LENGTHARG|OPFLAG_TYPEOFARG))!=0
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          && ((!sqlite3IsFixedLengthSerialType(t) && (t&1)==0)
+            || (pOp->p5 & OPFLAG_TYPEOFARG)!=0))
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           && ((t>=12 && (t&1)==0) || (pOp->p5 & OPFLAG_TYPEOFARG)!=0))
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
      || (len = sqlite3VdbeSerialTypeLen(t))==0
     ){
       /* Content is irrelevant for
@@ -2730,10 +3351,20 @@
       if( rc!=SQLITE_OK ) goto abort_due_to_error;
       sqlite3VdbeSerialGet((const u8*)pDest->z, t, pDest);
       pDest->flags &= ~MEM_Ephem;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+     /* we need to update the field in case this is a decimal */
+     if( pC->isTable == 0 ){
+       sqlite3UpdateMemCollAttr(pCrsr, aOffset[p2], pDest);
+     }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
   }
 
 op_column_out:
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  pDest->tz = p->tzname;
+  pDest->dtprec = p->dtprec;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   UPDATE_MAX_BLOBSIZE(pDest);
   REGISTER_TRACE(pOp->p3, pDest);
   break;
@@ -2768,6 +3399,13 @@
   do{
     assert( pIn1 <= &p->aMem[(p->nMem+1 - p->nCursor)] );
     assert( memIsValid(pIn1) );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( zAffinity[0]==SQLITE_AFF_DATETIME ){
+      /* pass the timezone before trying to convert */
+      pIn1->tz = p->tzname;
+      pIn1->dtprec = p->dtprec;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     applyAffinity(pIn1, *(zAffinity++), encoding);
     pIn1++;
   }while( zAffinity[0] );
@@ -2789,7 +3427,18 @@
 ** macros defined in sqliteInt.h.
 **
 ** If P4 is NULL then all index fields have the affinity BLOB.
+*/
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/* Opcode: MakeRecord P1 P2 P3 P4 P5 *
+**
+** Allowed P5 bits:
+** <ul>
+** <li>  <b>0x400 OPFLAG_MKREC_COMDB2</b>: P3 is a MEM_Int which points to a cursor.
+**       Use the table of this cursor, and convert fields to its row data.
+** <li>
+** </ul>
 */
+#endif
 case OP_MakeRecord: {
   u8 *zNewRecord;        /* A buffer to hold the data for the new record */
   Mem *pRec;             /* The new record */
@@ -2807,6 +3456,10 @@
   int i;                 /* Space used in zNewRecord[] header */
   int j;                 /* Space used in zNewRecord[] content */
   u32 len;               /* Length of a field */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  VdbeCursor *pC;        /* Cursor to the table for which the MakeRecord is meant */
+  int optimized;         /* 1 if the optimization is applied successfully */
+#endif
 
   /* Assuming the record contains N fields, the record format looks
   ** like this:
@@ -2845,6 +3498,14 @@
   if( zAffinity ){
     pRec = pData0;
     do{
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      /* FIXME TODO XXX: Remove this test when done */
+      if( islower(*zAffinity) ){
+        logmsg(LOGMSG_FATAL, "found a lower case affinity: %c     PANIC!!!\n", *zAffinity);
+        cheap_stack_trace();
+        exit(0);
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       applyAffinity(pRec++, *(zAffinity++), encoding);
       assert( zAffinity[0]==0 || pRec<=pLast );
     }while( zAffinity[0] );
@@ -2913,6 +3574,26 @@
     if( nVarint<sqlite3VarintLen(nHdr) ) nHdr++;
   }
   nByte = nHdr+nData;
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* Serialize directly to comdb2 row format. */
+  if( pOp->p5 & OPFLAG_MKREC_COMDB2 ){
+    /* Make sure that the final data will fit. */
+    if( nByte+nZero>db->aLimit[SQLITE_LIMIT_LENGTH] ){
+      goto too_big;
+    }
+    pC = p->apCsr[pOut->u.i];
+    rc = sqlite3MakeRecordForComdb2(pC->uc.pCursor, pData0, nField, &optimized);
+    if( rc!=0 ){
+      goto abort_due_to_error;
+    }else if( optimized ){
+      /* Make pOut a placeholder */
+      sqlite3VdbeMemRelease(pOut);
+      pOut->flags = MEM_Comdb2 | MEM_Blob;
+      break;
+    }
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   /* Make sure the output register has a buffer large enough to store 
   ** the new record. The output register (pOp->p3) is not allowed to
@@ -3268,7 +3949,14 @@
   assert( p->bIsReader );
   assert( p->readOnly==0 || pOp->p2==0 );
   assert( pOp->p1>=0 && pOp->p1<db->nDb );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* we have not set btreeMask because we dont set cookieMask
+   * which then is used to set btreeMask in sqlite3FinishCoding
+   * (modified by COMDB2) by calling sqlite3VdbeUsesBtree.
+   */
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( DbMaskTest(p->btreeMask, pOp->p1) );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( pOp->p2 && (db->flags & SQLITE_QueryOnly)!=0 ){
     rc = SQLITE_READONLY;
     goto abort_due_to_error;
@@ -3276,7 +3964,12 @@
   pBt = db->aDb[pOp->p1].pBt;
 
   if( pBt ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    rc = sqlite3BtreeBeginTrans(p, pBt, pOp->p2, &iMeta);
+    comdb2SetWriteFlag(pOp->p2);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     rc = sqlite3BtreeBeginTrans(pBt, pOp->p2, &iMeta);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     testcase( rc==SQLITE_BUSY_SNAPSHOT );
     testcase( rc==SQLITE_BUSY_RECOVERY );
     if( rc!=SQLITE_OK ){
@@ -3401,7 +4094,9 @@
   rc = sqlite3BtreeUpdateMeta(pDb->pBt, pOp->p2, pOp->p3);
   if( pOp->p2==BTREE_SCHEMA_VERSION ){
     /* When the schema cookie changes, record the new cookie internally */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
     pDb->pSchema->schema_cookie = pOp->p3;
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
     db->mDbFlags |= DBFLAG_SchemaChange;
   }else if( pOp->p2==BTREE_FILE_FORMAT ){
     /* Record changes in the file format */
@@ -3520,13 +4215,21 @@
   }
   /* If the cursor is not currently open or is open on a different
   ** index, then fall through into OP_OpenRead to force a reopen */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+case OP_OpenRead_Record:
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 case OP_OpenRead:
 case OP_OpenWrite:
 
   assert( pOp->opcode==OP_OpenWrite || pOp->p5==0 || pOp->p5==OPFLAG_SEEKEQ );
   assert( p->bIsReader );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( pOp->opcode==OP_OpenRead_Record || pOp->opcode==OP_OpenRead
+          || pOp->opcode==OP_ReopenIdx || p->readOnly==0 );
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( pOp->opcode==OP_OpenRead || pOp->opcode==OP_ReopenIdx
           || p->readOnly==0 );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   if( p->expired==1 ){
     rc = SQLITE_ABORT_ROLLBACK;
@@ -3545,12 +4248,19 @@
   if( pOp->opcode==OP_OpenWrite ){
     assert( OPFLAG_FORDELETE==BTREE_FORDELETE );
     wrFlag = BTREE_WRCSR | (pOp->p5 & OPFLAG_FORDELETE);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    wrFlag |= BTREE_CUR_WR;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( sqlite3SchemaMutexHeld(db, iDb, 0) );
     if( pDb->pSchema->file_format < p->minWriteFileFormat ){
       p->minWriteFileFormat = pDb->pSchema->file_format;
     }
   }else{
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    wrFlag = BTREE_CUR_RD;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     wrFlag = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   if( pOp->p5 & OPFLAG_P2ISREG ){
     assert( p2>0 );
@@ -3586,23 +4296,42 @@
 #ifdef SQLITE_DEBUG
   pCur->wrFlag = wrFlag;
 #endif
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  rc = sqlite3BtreeCursor(p, pX, p2, wrFlag, 1, pKeyInfo, pCur->uc.pCursor);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   rc = sqlite3BtreeCursor(pX, p2, wrFlag, pKeyInfo, pCur->uc.pCursor);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pCur->pKeyInfo = pKeyInfo;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( rc ) goto abort_due_to_error;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   /* Set the VdbeCursor.isTable variable. Previous versions of
   ** SQLite used to check if the root-page flags were sane at this point
   ** and report database corruption if they were not, but this check has
   ** since moved into the btree layer.  */  
   pCur->isTable = pOp->p4type!=P4_KEYINFO;
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* if this is a recording bit, mark the BtCursor as such */
+  if( pOp->opcode==OP_OpenRead_Record ){
+     sqlite3BtreeSetRecording(pCur->uc.pCursor, 1);
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 open_cursor_set_hints:
   assert( OPFLAG_BULKCSR==BTREE_BULKLOAD );
   assert( OPFLAG_SEEKEQ==BTREE_SEEK_EQ );
   testcase( pOp->p5 & OPFLAG_BULKCSR );
 #ifdef SQLITE_ENABLE_CURSOR_HINTS
   testcase( pOp->p2 & OPFLAG_SEEKEQ );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  sqlite3BtreeCursorHint(pCur->uc.pCursor, BTREE_HINT_FLAGS,
+                         (pOp->p5 & (OPFLAG_BULKCSR|OPFLAG_SEEKEQ)));
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   sqlite3BtreeCursorHintFlags(pCur->uc.pCursor,
                                (pOp->p5 & (OPFLAG_BULKCSR|OPFLAG_SEEKEQ)));
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( rc ) goto abort_due_to_error;
   break;
 }
@@ -3630,8 +4359,14 @@
   pCx->isTable = pOrig->isTable;
   pCx->pgnoRoot = pOrig->pgnoRoot;
   pCx->isOrdered = pOrig->isOrdered;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  rc = sqlite3BtreeCursor(p, pOrig->pBtx, pCx->pgnoRoot,
+                          BTREE_CUR_WR|BTREE_WRCSR, 1,
+                          pCx->pKeyInfo, pCx->uc.pCursor);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   rc = sqlite3BtreeCursor(pOrig->pBtx, pCx->pgnoRoot, BTREE_WRCSR,
                           pCx->pKeyInfo, pCx->uc.pCursor);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   /* The sqlite3BtreeCursor() routine can only fail for the first cursor
   ** opened for a database.  Since there is already an open cursor when this
   ** opcode is run, the sqlite3BtreeCursor() cannot fail */
@@ -3686,17 +4421,25 @@
   if( pCx ){
     /* If the ephermeral table is already open, erase all existing content
     ** so that the table is empty again, rather than creating a new table. */
-    rc = sqlite3BtreeClearTable(pCx->pBtx, pCx->pgnoRoot, 0);
+    assert( pCx->isEphemeral );
+    pCx->seqCount = 0;
+    pCx->cacheStatus = CACHE_STALE;
+    if( pCx->pBtx ){
+      rc = sqlite3BtreeClearTable(pCx->pBtx, pCx->pgnoRoot, 0);
+    }
   }else{
     pCx = allocateCursor(p, pOp->p1, pOp->p2, -1, CURTYPE_BTREE);
     if( pCx==0 ) goto no_mem;
-    pCx->nullRow = 1;
     pCx->isEphemeral = 1;
     rc = sqlite3BtreeOpen(db->pVfs, 0, db, &pCx->pBtx, 
                           BTREE_OMIT_JOURNAL | BTREE_SINGLE | pOp->p5,
                           vfsFlags);
     if( rc==SQLITE_OK ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      rc = sqlite3BtreeBeginTrans(p, pCx->pBtx, 1, 0);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       rc = sqlite3BtreeBeginTrans(pCx->pBtx, 1, 0);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
     if( rc==SQLITE_OK ){
       /* If a transient index is required, create it by calling
@@ -3712,20 +4455,38 @@
           assert( pCx->pgnoRoot==MASTER_ROOT+1 );
           assert( pKeyInfo->db==db );
           assert( pKeyInfo->enc==ENC(db) );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+          rc = sqlite3BtreeCursor(p, pCx->pBtx, pCx->pgnoRoot,
+                                  BTREE_CUR_WR|BTREE_WRCSR, 0,
+                                  pKeyInfo, pCx->uc.pCursor);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
           rc = sqlite3BtreeCursor(pCx->pBtx, pCx->pgnoRoot, BTREE_WRCSR,
                                   pKeyInfo, pCx->uc.pCursor);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         }
         pCx->isTable = 0;
       }else{
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        int pgno;
+        rc = sqlite3BtreeCreateTable(pCx->pBtx, &pgno, BTREE_INTKEY);
+        if( rc==SQLITE_OK ){
+          pCx->pgnoRoot = pgno;
+          rc = sqlite3BtreeCursor(p, pCx->pBtx, pgno,
+                                  BTREE_CUR_WR|BTREE_WRCSR, 0,
+                                  pKeyInfo, pCx->uc.pCursor);
+        }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         pCx->pgnoRoot = MASTER_ROOT;
         rc = sqlite3BtreeCursor(pCx->pBtx, MASTER_ROOT, BTREE_WRCSR,
                                 0, pCx->uc.pCursor);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
         pCx->isTable = 1;
       }
     }
     pCx->isOrdered = (pOp->p5!=BTREE_UNORDERED);
   }
   if( rc ) goto abort_due_to_error;
+  pCx->nullRow = 1;
   break;
 }
 
@@ -3835,6 +4596,9 @@
   pC = p->apCsr[pOp->p1];
   assert( pC->eCurType==CURTYPE_BTREE );
   pC->maskUsed = *(u64*)pOp->p4.pI64;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  sqlite3BtreeCursorSetFieldUsed(pC->uc.pCursor, pC->maskUsed);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   break;
 }
 #endif
@@ -3935,7 +4699,9 @@
   UnpackedRecord r;  /* The key to seek for */
   int nField;        /* Number of columns or fields in the key */
   i64 iKey;          /* The rowid we are to seek to */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   int eqOnly;        /* Only interested in == results */
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   assert( pOp->p1>=0 && pOp->p1<p->nCursor );
   assert( pOp->p2!=0 );
@@ -3948,7 +4714,9 @@
   assert( pC->isOrdered );
   assert( pC->uc.pCursor!=0 );
   oc = pOp->opcode;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   eqOnly = 0;
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   pC->nullRow = 0;
 #ifdef SQLITE_DEBUG
   pC->seekOp = pOp->opcode;
@@ -4001,7 +4769,11 @@
         if( (oc & 0x0001)==(OP_SeekLT & 0x0001) ) oc++;
       }
     } 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    rc = sqlite3BtreeMovetoUnpacked(pC->uc.pCursor,0,(u64)iKey,pOp->opcode,&res);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     rc = sqlite3BtreeMovetoUnpacked(pC->uc.pCursor, 0, (u64)iKey, 0, &res);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     pC->movetoTarget = iKey;  /* Used by OP_Delete */
     if( rc!=SQLITE_OK ){
       goto abort_due_to_error;
@@ -4011,6 +4783,7 @@
     ** OP_SeekLE opcodes are allowed, and these must be immediately followed
     ** by an OP_IdxGT or OP_IdxLT opcode, respectively, with the same key.
     */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
     if( sqlite3BtreeCursorHasHint(pC->uc.pCursor, BTREE_SEEK_EQ) ){
       eqOnly = 1;
       assert( pOp->opcode==OP_SeekGE || pOp->opcode==OP_SeekLE );
@@ -4020,13 +4793,18 @@
       assert( pOp[1].p3==pOp[0].p3 );
       assert( pOp[1].p4.i==pOp[0].p4.i );
     }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 
     nField = pOp->p4.i;
     assert( pOp->p4type==P4_INT32 );
     assert( nField>0 );
     r.pKeyInfo = pC->pKeyInfo;
     r.nField = (u16)nField;
-
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /* Reset cooked state that open cursor
+     * optimization for 'or' clause would have left us with */
+    setCookCol(pC, -1);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     /* The next line of code computes as follows, only faster:
     **   if( oc==OP_SeekGT || oc==OP_SeekLE ){
     **     r.default_rc = -1;
@@ -4045,14 +4823,27 @@
     { int i; for(i=0; i<r.nField; i++) assert( memIsValid(&r.aMem[i]) ); }
 #endif
     r.eqSeen = 0;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    rc = sqlite3BtreeMovetoUnpacked(pC->uc.pCursor, &r, 0, pOp->opcode, &res);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     rc = sqlite3BtreeMovetoUnpacked(pC->uc.pCursor, &r, 0, 0, &res);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( rc!=SQLITE_OK ){
       goto abort_due_to_error;
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /* If we're walking an index backwards as part of skip scan
+       don't use the cookFields optimization
+     */
+    if(!(oc==OP_SeekLT && (pOp->p5 & OPFLAG_SKIPSCAN)!=0)) {
+        setCookCol(pC, r.nField);
+    }
+#else //!defined(SQLITE_BUILDING_FOR_COMDB2)
     if( eqOnly && r.eqSeen==0 ){
       assert( res!=0 );
       goto seek_not_found;
     }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   pC->deferredMoveto = 0;
   pC->cacheStatus = CACHE_STALE;
@@ -4061,6 +4852,11 @@
 #endif
   if( oc>=OP_SeekGE ){  assert( oc==OP_SeekGE || oc==OP_SeekGT );
     if( res<0 || (res==0 && oc==OP_SeekGT) ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( sqlite3BtreeEof(pC->uc.pCursor) ){
+        res = 1;
+      }else{
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       res = 0;
       rc = sqlite3BtreeNext(pC->uc.pCursor, 0);
       if( rc!=SQLITE_OK ){
@@ -4071,6 +4867,9 @@
           goto abort_due_to_error;
         }
       }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }else{
       res = 0;
     }
@@ -4094,37 +4893,209 @@
       res = sqlite3BtreeEof(pC->uc.pCursor);
     }
   }
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
 seek_not_found:
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( pOp->p2>0 );
   VdbeBranchTaken(res!=0,2);
   if( res ){
     goto jump_to_p2;
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   }else if( eqOnly ){
     assert( pOp[1].opcode==OP_IdxLT || pOp[1].opcode==OP_IdxGT );
     pOp++; /* Skip the OP_IdxLt or OP_IdxGT that follows */
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   break;
 }
 
-/* Opcode: SeekHit P1 P2 * * *
-** Synopsis: seekHit=P2
+
+/* Opcode: SeekScan  P1 P2 * * *
+** Synopsis: Scan-ahead up to P1 rows
 **
-** Set the seekHit flag on cursor P1 to the value in P2.
-** The seekHit flag is used by the IfNoHope opcode.
+** This opcode is a prefix opcode to OP_SeekGE.  In other words, this
+** opcode must be immediately followed by OP_SeekGE. This constraint is
+** checked by assert() statements.
 **
-** P1 must be a valid b-tree cursor.  P2 must be a boolean value,
-** either 0 or 1.
+** This opcode uses the P1 through P4 operands of the subsequent
+** OP_SeekGE.  In the text that follows, the operands of the subsequent
+** OP_SeekGE opcode are denoted as SeekOP.P1 through SeekOP.P4.   Only
+** the P1 and P2 operands of this opcode are also used, and  are called
+** This.P1 and This.P2.
+**
+** This opcode helps to optimize IN operators on a multi-column index
+** where the IN operator is on the later terms of the index by avoiding
+** unnecessary seeks on the btree, substituting steps to the next row
+** of the b-tree instead.  A correct answer is obtained if this opcode
+** is omitted or is a no-op.
+**
+** The SeekGE.P3 and SeekGE.P4 operands identify an unpacked key which
+** is the desired entry that we want the cursor SeekGE.P1 to be pointing
+** to.  Call this SeekGE.P4/P5 row the "target".
+**
+** If the SeekGE.P1 cursor is not currently pointing to a valid row,
+** then this opcode is a no-op and control passes through into the OP_SeekGE.
+**
+** If the SeekGE.P1 cursor is pointing to a valid row, then that row
+** might be the target row, or it might be near and slightly before the
+** target row.  This opcode attempts to position the cursor on the target
+** row by, perhaps by invoking sqlite3BtreeStep() on the cursor
+** between 0 and This.P1 times.
+**
+** There are three possible outcomes from this opcode:<ol>
+**
+** <li> If after This.P1 steps, the cursor is still point to a place that
+**      is earlier in the btree than the target row,
+**      then fall through into the subsquence OP_SeekGE opcode.
+**
+** <li> If the cursor is successfully moved to the target row by 0 or more
+**      sqlite3BtreeNext() calls, then jump to This.P2, which will land just
+**      past the OP_IdxGT opcode that follows the OP_SeekGE.
+**
+** <li> If the cursor ends up past the target row (indicating the the target
+**      row does not exist in the btree) then jump to SeekOP.P2. 
+** </ol>
+*/
+case OP_SeekScan: {
+  VdbeCursor *pC;
+  int res;
+  int nStep;
+  UnpackedRecord r;
+
+  assert( pOp[1].opcode==OP_SeekGE );
+
+  /* pOp->p2 points to the first instruction past the OP_IdxGT that
+  ** follows the OP_SeekGE.  */
+  assert( pOp->p2>=(int)(pOp-aOp)+2 );
+  assert( aOp[pOp->p2-1].opcode==OP_IdxGT );
+  assert( pOp[1].p1==aOp[pOp->p2-1].p1 );
+  assert( pOp[1].p2==aOp[pOp->p2-1].p2 );
+  assert( pOp[1].p3==aOp[pOp->p2-1].p3 );
+
+  assert( pOp->p1>0 );
+  pC = p->apCsr[pOp[1].p1];
+  assert( pC!=0 );
+  assert( pC->eCurType==CURTYPE_BTREE );
+  assert( !pC->isTable );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( pC->nullRow || !sqlite3BtreeCursorIsValidNN(pC->uc.pCursor) ){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+  if( !sqlite3BtreeCursorIsValidNN(pC->uc.pCursor) ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+#ifdef SQLITE_DEBUG
+     if( db->flags&SQLITE_VdbeTrace ){
+       printf("... cursor not valid - fall through\n");
+     }        
+#endif
+    break;
+  }
+  nStep = pOp->p1;
+  assert( nStep>=1 );
+  r.pKeyInfo = pC->pKeyInfo;
+  r.nField = (u16)pOp[1].p4.i;
+  r.default_rc = 0;
+  r.aMem = &aMem[pOp[1].p3];
+#ifdef SQLITE_DEBUG
+  {
+    int i;
+    for(i=0; i<r.nField; i++){
+      assert( memIsValid(&r.aMem[i]) );
+      REGISTER_TRACE(pOp[1].p3+i, &aMem[pOp[1].p3+i]);
+    }
+  }
+#endif
+  res = 0;  /* Not needed.  Only used to silence a warning. */
+  while(1){
+    rc = sqlite3VdbeIdxKeyCompare(db, pC, &r, &res);
+    if( rc ) goto abort_due_to_error;
+    if( res>0 ){
+      seekscan_search_fail:
+#ifdef SQLITE_DEBUG
+      if( db->flags&SQLITE_VdbeTrace ){
+        printf("... %d steps and then skip\n", pOp->p1 - nStep);
+      }        
+#endif
+      VdbeBranchTaken(1,3);
+      pOp++;
+      goto jump_to_p2;
+    }
+    if( res==0 ){
+#ifdef SQLITE_DEBUG
+      if( db->flags&SQLITE_VdbeTrace ){
+        printf("... %d steps and then success\n", pOp->p1 - nStep);
+      }        
+#endif
+      VdbeBranchTaken(2,3);
+      goto jump_to_p2;
+      break;
+    }
+    if( nStep<=0 ){
+#ifdef SQLITE_DEBUG
+      if( db->flags&SQLITE_VdbeTrace ){
+        printf("... fall through after %d steps\n", pOp->p1);
+      }        
+#endif
+      VdbeBranchTaken(0,3);
+      break;
+    }
+    nStep--;
+    rc = sqlite3BtreeNext(pC->uc.pCursor, 0);
+    if( rc ){
+      if( rc==SQLITE_DONE ){
+        rc = SQLITE_OK;
+        goto seekscan_search_fail;
+      }else{
+        goto abort_due_to_error;
+      }
+    }
+  }
+  
+  break;
+}
+
+
+/* Opcode: SeekHit P1 P2 P3 * *
+** Synopsis: set P2<=seekHit<=P3
+**
+** Increase or decrease the seekHit value for cursor P1, if necessary,
+** so that it is no less than P2 and no greater than P3.
+**
+** The seekHit integer represents the maximum of terms in an index for which
+** there is known to be at least one match.  If the seekHit value is smaller
+** than the total number of equality terms in an index lookup, then the
+** OP_IfNoHope opcode might run to see if the IN loop can be abandoned
+** early, thus saving work.  This is part of the IN-early-out optimization.
+**
+** P1 must be a valid b-tree cursor.
 */
 case OP_SeekHit: {
   VdbeCursor *pC;
   assert( pOp->p1>=0 && pOp->p1<p->nCursor );
   pC = p->apCsr[pOp->p1];
   assert( pC!=0 );
-  assert( pOp->p2==0 || pOp->p2==1 );
-  pC->seekHit = pOp->p2 & 1;
+  assert( pOp->p3>=pOp->p2 );
+  if( pC->seekHit<pOp->p2 ){
+    pC->seekHit = pOp->p2;
+  }else if( pC->seekHit>pOp->p3 ){
+    pC->seekHit = pOp->p3;
+  }
   break;
 }
 
+/* Opcode: IfNotOpen P1 P2 * * *
+** Synopsis: if( !csr[P1] ) goto P2
+**
+** If cursor P1 is not open, jump to instruction P2. Otherwise, fall through.
+*/
+case OP_IfNotOpen: {        /* jump */
+  assert( pOp->p1>=0 && pOp->p1<p->nCursor );
+  if( !p->apCsr[pOp->p1] ){
+    goto jump_to_p2_and_check_for_interrupt;
+  }
+  break;
+}
+
+
 /* Opcode: Found P1 P2 P3 P4 *
 ** Synopsis: key=r[P3@P4]
 **
@@ -4142,7 +5113,7 @@
 **
 ** See also: NotFound, NoConflict, NotExists. SeekGe
 */
-/* Opcode: NotFound P1 P2 P3 P4 *
+/* Opcode: NotFound P1 P2 P3 P4 P5
 ** Synopsis: key=r[P3@P4]
 **
 ** If P4==0 then register P3 holds a blob constructed by MakeRecord.  If
@@ -4155,6 +5126,8 @@
 ** falls through to the next instruction and P1 is left pointing at the
 ** matching entry.
 **
+** Comdb2: Throw a verification error if P5 is non-zero.
+**
 ** This operation leaves the cursor in a state where it cannot be
 ** advanced in either direction.  In other words, the Next and Prev
 ** opcodes do not work after this operation.
@@ -4165,16 +5138,20 @@
 ** Synopsis: key=r[P3@P4]
 **
 ** Register P3 is the first of P4 registers that form an unpacked
-** record.
+** record.  Cursor P1 is an index btree.  P2 is a jump destination.
+** In other words, the operands to this opcode are the same as the
+** operands to OP_NotFound and OP_IdxGT.
 **
-** Cursor P1 is on an index btree.  If the seekHit flag is set on P1, then
-** this opcode is a no-op.  But if the seekHit flag of P1 is clear, then
-** check to see if there is any entry in P1 that matches the
-** prefix identified by P3 and P4.  If no entry matches the prefix,
-** jump to P2.  Otherwise fall through.
+** This opcode is an optimization attempt only.  If this opcode always
+** falls through, the correct answer is still obtained, but extra works
+** is performed.
 **
-** This opcode behaves like OP_NotFound if the seekHit
-** flag is clear and it behaves like OP_Noop if the seekHit flag is set.
+** A value of N in the seekHit flag of cursor P1 means that there exists
+** a key P3:N that will match some record in the index.  We want to know
+** if it is possible for a record P3:P4 to match some record in the
+** index.  If it is not possible, we can skips some work.  So if seekHit
+** is less than P4, attempt to find out if a match is possible by running
+** OP_NotFound.
 **
 ** This opcode is used in IN clause processing for a multi-column key.
 ** If an IN clause is attached to an element of the key other than the
@@ -4216,7 +5193,7 @@
   assert( pOp->p1>=0 && pOp->p1<p->nCursor );
   pC = p->apCsr[pOp->p1];
   assert( pC!=0 );
-  if( pC->seekHit ) break;
+  if( pC->seekHit>=pOp->p4.i ) break;
   /* Fall through into OP_NotFound */
 }
 case OP_NoConflict:     /* jump, in3 */
@@ -4274,15 +5251,30 @@
     /* For the OP_NoConflict opcode, take the jump if any of the
     ** input fields are NULL, since any key with a NULL will not
     ** conflict */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    int is_idx_uniqnulls = comdb2_is_idx_uniqnulls(pC->uc.pCursor);
+    for(ii=0; is_idx_uniqnulls && ii<pIdxKey->nField; ii++){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     for(ii=0; ii<pIdxKey->nField; ii++){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       if( pIdxKey->aMem[ii].flags & MEM_Null ){
         takeJump = 1;
         break;
       }
     }
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( (pOp->opcode==OP_IfNoHope || pOp->opcode==OP_NotFound) && pOp->p5 ){
+    res = -1;
+  }else{
+    res = 0; /* res = -1 triggers early verify check */
+  }
+  rc = sqlite3BtreeMovetoUnpacked(pC->uc.pCursor,pIdxKey,0,pOp->opcode,&res);
+  if( pOp->p4.i==0 ){ sqlite3DbFree(db, pFree); }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   rc = sqlite3BtreeMovetoUnpacked(pC->uc.pCursor, pIdxKey, 0, 0, &res);
   if( pFree ) sqlite3DbFreeNN(db, pFree);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( rc!=SQLITE_OK ){
     goto abort_due_to_error;
   }
@@ -4297,6 +5289,7 @@
   }else{
     VdbeBranchTaken(takeJump||alreadyExists==0,2);
     if( takeJump || !alreadyExists ) goto jump_to_p2;
+    if( pOp->opcode==OP_IfNoHope ) pC->seekHit = pOp->p4.i;
   }
   break;
 }
@@ -4325,7 +5318,7 @@
 **
 ** See also: Found, NotFound, NoConflict, SeekRowid
 */
-/* Opcode: NotExists P1 P2 P3 * *
+/* Opcode: NotExists P1 P2 P3 * P5
 ** Synopsis: intkey=r[P3]
 **
 ** P1 is the index of a cursor open on an SQL table btree (with integer
@@ -4342,6 +5335,8 @@
 ** The OP_NotFound opcode performs the same operation on index btrees
 ** (with arbitrary multi-value keys).
 **
+** Comdb2: Throw a verification error if P5 is non-zero.
+**
 ** This opcode leaves the cursor in a state where it cannot be advanced
 ** in either direction.  In other words, the Next and Prev opcodes will
 ** not work following this opcode.
@@ -4383,7 +5378,12 @@
   assert( pCrsr!=0 );
   res = 0;
   iKey = pIn3->u.i;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( pOp->p5 ) res = -1; /* res = -1 triggers early verify check */
+  rc = sqlite3BtreeMovetoUnpacked(pCrsr, 0, iKey, pOp->opcode, &res);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   rc = sqlite3BtreeMovetoUnpacked(pCrsr, 0, iKey, 0, &res);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( rc==SQLITE_OK || res==0 );
   pC->movetoTarget = iKey;  /* Used by OP_Delete */
   pC->nullRow = 0;
@@ -4392,12 +5392,19 @@
   VdbeBranchTaken(res!=0,2);
   pC->seekResult = res;
   if( res!=0 ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( pOp->p2==0 && rc==SQLITE_OK ){
+      rc = SQLITE_CORRUPT_BKPT;
+    }
+    goto jump_to_p2;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( rc==SQLITE_OK );
     if( pOp->p2==0 ){
       rc = SQLITE_CORRUPT_BKPT;
     }else{
       goto jump_to_p2;
     }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   if( rc ) goto abort_due_to_error;
   break;
@@ -4453,7 +5460,14 @@
   assert( pC->isTable );
   assert( pC->eCurType==CURTYPE_BTREE );
   assert( pC->uc.pCursor!=0 );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
   {
+    UNUSED_PARAMETER2(res, cnt);
+    UNUSED_PARAMETER2(pMem, pFrame);
+    v = sqlite3BtreeNewRowid(pC->uc.pCursor);
+  }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+  {
     /* The next rowid or record number (different terms for the same
     ** thing) is obtained in a two-step algorithm.
     **
@@ -4552,6 +5566,7 @@
     pC->deferredMoveto = 0;
     pC->cacheStatus = CACHE_STALE;
   }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pOut->u.i = v;
   break;
 }
@@ -4618,6 +5633,7 @@
   assert( pKey->flags & MEM_Int );
   assert( memIsValid(pKey) );
   REGISTER_TRACE(pOp->p3, pKey);
+  memset(&x, 0, sizeof(BtreePayload));
   x.nKey = pKey->u.i;
 
   if( pOp->p4type==P4_TABLE && HAS_UPDATE_HOOK(db) ){
@@ -4647,6 +5663,15 @@
   if( pOp->p5 & OPFLAG_NCHANGE ) p->nChange++;
   if( pOp->p5 & OPFLAG_LASTROWID ) db->lastRowid = x.nKey;
   assert( pData->flags & (MEM_Blob|MEM_Str) );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* Data is already serialized to comdb2 row format and stored in the btree cursor. */
+  if( pData->flags & MEM_Comdb2 ){
+    seekResult = ((pOp->p5 & OPFLAG_USESEEKRESULT) ? pC->seekResult : 0);
+    rc = sqlite3BtreeInsert(pC->uc.pCursor, &x,
+        (pOp->p5 & OPFLAG_ISUPDATE)!=0, seekResult, pOp->p5
+    );
+  }else{
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   x.pData = pData->z;
   x.nData = pData->n;
   seekResult = ((pOp->p5 & OPFLAG_USESEEKRESULT) ? pC->seekResult : 0);
@@ -4656,9 +5681,16 @@
     x.nZero = 0;
   }
   x.pKey = 0;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
   rc = sqlite3BtreeInsert(pC->uc.pCursor, &x,
+      (pOp->p5 & OPFLAG_ISUPDATE)!=0, seekResult, pOp->p5
+  );
+  }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+  rc = sqlite3BtreeInsert(pC->uc.pCursor, &x,
       (pOp->p5 & (OPFLAG_APPEND|OPFLAG_SAVEPOSITION)), seekResult
   );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pC->deferredMoveto = 0;
   pC->cacheStatus = CACHE_STALE;
 
@@ -4721,9 +5753,26 @@
   assert( pC!=0 );
   assert( pC->eCurType==CURTYPE_BTREE );
   assert( pC->uc.pCursor!=0 );
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* We don't generate the index deletes, and we miss 
+  ** positioninig on the data cursor in the process
+  ** (see sqlite3GenerateRowIndexDelete).
+  */
+  if( pC->isTable && pC->deferredMoveto ){
+    int p2; /* NOT USED: column number to retrieve */
+    rc = sqlite3VdbeCursorMoveto(&pC, &p2);
+    if( rc ) goto abort_due_to_error;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( pC->deferredMoveto==0 );
   sqlite3VdbeIncrWriteCounter(p, pC);
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( pOp->p5 && db->xUpdateCallback && pOp->p4.z && pC->isTable ){
+    pC->movetoTarget = sqlite3BtreeIntegerKey(pC->uc.pCursor);
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifdef SQLITE_DEBUG
   if( pOp->p4type==P4_TABLE && HasRowid(pOp->p4.pTab) && pOp->p5==0 ){
     /* If p5 is zero, the seek operation that positioned the cursor prior to
@@ -4935,10 +5984,23 @@
   if( rc!=SQLITE_OK ) goto abort_due_to_error;
 #endif
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( pC->isTable==0 ){
+    i64 n64 = sqlite3BtreeIntegerKey(pCrsr);
+    if( n64>db->aLimit[SQLITE_LIMIT_LENGTH] ){
+      goto too_big;
+    }
+    assert( n64<=0xffffffff );
+    n = (u32)n64;
+  }else{
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   n = sqlite3BtreePayloadSize(pCrsr);
   if( n>(u32)db->aLimit[SQLITE_LIMIT_LENGTH] ){
     goto too_big;
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   testcase( n==0 );
   rc = sqlite3VdbeMemFromBtree(pCrsr, 0, n, pOut);
   if( rc ) goto abort_due_to_error;
@@ -4948,7 +6010,7 @@
   break;
 }
 
-/* Opcode: Rowid P1 P2 * * *
+/* Opcode: Rowid P1 P2 P3 * *
 ** Synopsis: r[P2]=rowid
 **
 ** Store in register P2 an integer which is the key of the table entry that
@@ -4957,6 +6019,10 @@
 ** P1 can be either an ordinary table or a virtual table.  There used to
 ** be a separate OP_VRowid opcode for use with virtual tables, but this
 ** one opcode now works for both table types.
+**
+** If P3 is 1, return a string in the format rrn:genid instead of the
+** integer key.  If P3 is 2, return an integer timestamp instead of the
+** integer key.
 */
 case OP_Rowid: {                 /* out2 */
   VdbeCursor *pC;
@@ -4995,7 +6061,11 @@
     }
     v = sqlite3BtreeIntegerKey(pC->uc.pCursor);
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  getRowid(pC->uc.pCursor, v, pOp->p3, pOut);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pOut->u.i = v;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   break;
 }
 
@@ -5255,6 +6325,9 @@
   assert( pOp->p5<ArraySize(p->aCounter) );
   pC = p->apCsr[pOp->p1];
   assert( pC!=0 );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( pC->nullRow ) break;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( pC->deferredMoveto==0 );
   assert( pC->eCurType==CURTYPE_BTREE );
   assert( pOp->opcode!=OP_Next || pOp->p4.xAdvance==sqlite3BtreeNext );
@@ -5264,11 +6337,12 @@
   ** The Prev opcode is only used after SeekLT, SeekLE, and Last. */
   assert( pOp->opcode!=OP_Next
        || pC->seekOp==OP_SeekGT || pC->seekOp==OP_SeekGE
-       || pC->seekOp==OP_Rewind || pC->seekOp==OP_Found 
-       || pC->seekOp==OP_NullRow|| pC->seekOp==OP_SeekRowid);
+       || pC->seekOp==OP_Rewind || pC->seekOp==OP_Found
+       || pC->seekOp==OP_NullRow|| pC->seekOp==OP_SeekRowid
+       || pC->seekOp==OP_IfNoHope);
   assert( pOp->opcode!=OP_Prev
        || pC->seekOp==OP_SeekLT || pC->seekOp==OP_SeekLE
-       || pC->seekOp==OP_Last 
+       || pC->seekOp==OP_Last   || pC->seekOp==OP_IfNoHope
        || pC->seekOp==OP_NullRow);
 
   rc = pOp->p4.xAdvance(pC->uc.pCursor, pOp->p3);
@@ -5286,6 +6360,9 @@
   if( rc!=SQLITE_DONE ) goto abort_due_to_error;
   rc = SQLITE_OK;
   pC->nullRow = 1;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( pC->eCurType==CURTYPE_BTREE ) setCookCol(pC, 0);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   goto check_for_interrupt;
 }
 
@@ -5329,6 +6406,9 @@
   VdbeCursor *pC;
   BtreePayload x;
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  memset(&x, 0, sizeof(BtreePayload));
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( pOp->p1>=0 && pOp->p1<p->nCursor );
   pC = p->apCsr[pOp->p1];
   sqlite3VdbeIncrWriteCounter(p, pC);
@@ -5348,10 +6428,18 @@
     x.pKey = pIn2->z;
     x.aMem = aMem + pOp->p3;
     x.nMem = (u16)pOp->p4.i;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
     rc = sqlite3BtreeInsert(pC->uc.pCursor, &x,
+         (pOp->p5 & OPFLAG_ISUPDATE)!=0,
+        ((pOp->p5 & OPFLAG_USESEEKRESULT) ? pC->seekResult : 0),
+         pOp->p5
+        );
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+    rc = sqlite3BtreeInsert(pC->uc.pCursor, &x,
          (pOp->p5 & (OPFLAG_APPEND|OPFLAG_SAVEPOSITION)), 
         ((pOp->p5 & OPFLAG_USESEEKRESULT) ? pC->seekResult : 0)
         );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( pC->deferredMoveto==0 );
     pC->cacheStatus = CACHE_STALE;
   }
@@ -5386,7 +6474,12 @@
   r.nField = (u16)pOp->p3;
   r.default_rc = 0;
   r.aMem = &aMem[pOp->p2];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* NOTE: in this release updates are not supported yet */
+  rc = sqlite3BtreeMovetoUnpacked(pCrsr, &r, 0, pOp->opcode, &res);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   rc = sqlite3BtreeMovetoUnpacked(pCrsr, &r, 0, 0, &res);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( rc ) goto abort_due_to_error;
   if( res==0 ){
     rc = sqlite3BtreeDelete(pCrsr, BTREE_AUXDELETE);
@@ -5424,6 +6517,10 @@
 ** the end of the index key pointed to by cursor P1.  This integer should be
 ** the rowid of the table entry to which this index entry points.
 **
+** If P3 is 1, return a string in the format rrn:genid instead of the
+** integer key.  If P3 is 2, return an integer timestamp instead of the
+** integer key.
+**
 ** See also: Rowid, MakeRecord.
 */
 case OP_DeferredSeek:
@@ -5454,6 +6551,9 @@
     rowid = 0;  /* Not needed.  Only used to silence a warning. */
     rc = sqlite3VdbeIdxRowid(db, pC->uc.pCursor, &rowid);
     if( rc!=SQLITE_OK ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      logmsg(LOGMSG_ERROR, "abort_due_to_error\n");
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       goto abort_due_to_error;
     }
     if( pOp->opcode==OP_DeferredSeek ){
@@ -5471,7 +6571,11 @@
       pTabCur->pAltCursor = pC;
     }else{
       pOut = out2Prerelease(p, pOp);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      getRowid(pC->uc.pCursor, rowid, pOp->p3, pOut);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       pOut->u.i = rowid;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     }
   }else{
     assert( pOp->opcode==OP_IdxRowid );
@@ -5480,6 +6584,24 @@
   break;
 }
 
+/* Opcode: FinishSeek P1 * * * *
+**
+** If cursor P1 was previously moved via OP_DeferredSeek, complete that
+** seek operation now, without further delay.  If the cursor seek has
+** already occurred, this instruction is a no-op.
+*/
+case OP_FinishSeek: {
+  VdbeCursor *pC;             /* The P1 index cursor */
+
+  assert( pOp->p1>=0 && pOp->p1<p->nCursor );
+  pC = p->apCsr[pOp->p1];
+  if( pC->deferredMoveto ){
+    rc = sqlite3VdbeFinishMoveto(pC);
+    if( rc ) goto abort_due_to_error;
+  }
+  break;
+}
+
 /* Opcode: IdxGE P1 P2 P3 P4 P5
 ** Synopsis: key=r[P3@P4]
 **
@@ -5543,6 +6665,9 @@
   assert( pOp->p4type==P4_INT32 );
   r.pKeyInfo = pC->pKeyInfo;
   r.nField = (u16)pOp->p4.i;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  setCookCol(pC, r.nField);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( pOp->opcode<OP_IdxLT ){
     assert( pOp->opcode==OP_IdxLE || pOp->opcode==OP_IdxGT );
     r.default_rc = -1;
@@ -5551,6 +6676,9 @@
     r.default_rc = 0;
   }
   r.aMem = &aMem[pOp->p3];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( r.aMem->tz==NULL ) r.aMem->tz = p->tzname;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #ifdef SQLITE_DEBUG
   {
     int i;
@@ -5769,7 +6897,11 @@
 
   iDb = pOp->p1;
   assert( iDb>=0 && iDb<db->nDb );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( iDb>1 || DbHasProperty(db, iDb, DB_SchemaLoaded) );
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( DbHasProperty(db, iDb, DB_SchemaLoaded) );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 #ifndef SQLITE_OMIT_ALTERTABLE
   if( pOp->p4.z==0 ){
@@ -5828,6 +6960,9 @@
 */
 case OP_LoadAnalysis: {
   assert( pOp->p1>=0 && pOp->p1<db->nDb );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  p->numTables = 0;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   rc = sqlite3AnalysisLoad(db, pOp->p1);
   if( rc ) goto abort_due_to_error;
   break;  
@@ -6909,7 +8044,34 @@
     goto abort_due_to_error;
   }
   pModule = pVtab->pModule;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  int comdb2_check_vtab_access(sqlite3*, sqlite3_module*);
+  rc = comdb2_check_vtab_access(db, (sqlite3_module*)pModule);
+  if( rc ){
+    if (p->crtPartitionLocks>0) {
+      /* release the view lock, if any */
+      extern void views_unlock(void);
+      views_unlock();
+      p->crtPartitionLocks = 0;
+    }
+    goto abort_due_to_error;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   rc = pModule->xOpen(pVtab, &pVCur);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  extern int views_needs_comdb2_tables_lock(const char *);
+  if( p->crtPartitionLocks>0 && pModule->systable_lock &&
+      !strncasecmp(pModule->systable_lock, "comdb2_tables", strlen("comdb2_tables")+1 )){
+    p->crtPartitionLocks--;
+    if( p->crtPartitionLocks==0 ){
+        /* done collecting all schema, release the views lock for the rest
+         * of execution
+         */
+        extern void views_unlock(void);
+        views_unlock();
+    }
+  }
+#endif
   sqlite3VtabImportErrmsg(p, pVtab);
   if( rc ) goto abort_due_to_error;
 
@@ -7337,6 +8499,17 @@
   if( pCtx->pOut != pOut ){
     pCtx->pOut = pOut;
     for(i=pCtx->argc-1; i>=0; i--) pCtx->argv[i] = &aMem[pOp->p2+i];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    for(i=pCtx->argc-1; i>=0; i--){
+      /* hook the tzname just in case a following conversion will need it
+       * this is basically a hack needed by now() which has no way of
+       * setting the timezone
+       * instead, the next time a function is applied to the result of now
+       * the correct tzname is set for it */
+      pCtx->argv[i]->tz = p->tzname;
+      pCtx->argv[i]->dtprec= p->dtprec;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
 
   memAboutToChange(p, pOut);
@@ -7367,8 +8540,22 @@
     if( sqlite3VdbeMemTooBig(pOut) ) goto too_big;
   }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* above operations wipe the pOut->db sometimes */
+  if( !pCtx->pOut->db ){
+    pCtx->pOut->db = db;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   REGISTER_TRACE(pOp->p3, pOut);
   UPDATE_MAX_BLOBSIZE(pOut);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* the stack contains a synthetic Mem which could be needing a tzname
+   * the argument-less functions (as now()) are unable to receive a tzname
+   * and therefore to set it;
+   * we do it here; 11012007dh; */
+  pCtx->pOut->tz = p->tzname;
+  pCtx->pOut->dtprec = p->dtprec;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   break;
 }
 
@@ -7485,7 +8672,9 @@
   assert( pOp->p4type==P4_EXPR );
   pC = p->apCsr[pOp->p1];
   if( pC ){
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
     assert( pC->eCurType==CURTYPE_BTREE );
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3BtreeCursorHint(pC->uc.pCursor, BTREE_HINT_RANGE,
                            pOp->p4.pExpr, aMem);
   }
@@ -7508,7 +8697,117 @@
   break;
 }
 #endif
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/* OpCode: OpFuncLoad * P2 * P4 *
+** Synopsis: Load OpFunc P4 into mem[P2]
+**
+** This operator loads the custom function stored in P4 into Mem[P2].
+*/  
+case OP_OpFuncLoad: {          /*  out2-prerelease */
+  OpFunc *f = pOp->p4.comdb2func;
+
+  pIn1 = &aMem[pOp->p2]; /* TODO pOut */
+  pIn1->u.pOpFunc = f;
+  pIn1->flags = MEM_OpFunc;
+  break;
+}
+
+/* OpCode: OpFuncExec P1 * * * * 
+** Synopsis: Exec OpFunc in Mem[P1]
+**
+** This operator executes the OpFunc stored in Mem[P1].
+*/ 
+case OP_OpFuncExec: {
+  pIn1 = &aMem[pOp->p1];
+  OpFunc *f = pIn1->u.pOpFunc;
+  assert( f && f->func );
+  rc = f->func(f);
+  testcase( rc==SQLITE_OK );
+  testcase( rc==SQLITE_DONE );
+  testcase( rc!=SQLITE_OK && rc!=SQLITE_DONE );
+  testcase( f->rc==SQLITE_OK );
+  testcase( f->rc==SQLITE_DONE );
+  testcase( f->rc!=SQLITE_OK && f->rc!=SQLITE_DONE );
+  if( f->rc ){
+    rc = f->rc;
+    p->rc = rc==SQLITE_DONE ? SQLITE_OK : rc;
+    sqlite3SetString(&p->zErrMsg, db, f->errorMsg);
+  }
+  testcase( rc==SQLITE_OK );
+  testcase( rc==SQLITE_DONE );
+  testcase( rc==SQLITE_COMDB2SCHEMA );
+  testcase( rc!=SQLITE_OK && rc!=SQLITE_COMDB2SCHEMA );
+  if( rc==SQLITE_DONE || rc==SQLITE_COMDB2SCHEMA ){
+    goto vdbe_return;
+  }else if( rc ){
+    goto abort_due_to_error;
+  }
+  break;
+}
+
+/* OpCode: OpFuncInteger P1 P2 * * * 
+** Synopsis: Next integer of Mem[P1] into Mem[P2]
+**
+** Loads the next integer stored in the buffer of the OpFunc in Mem[P1]
+** into Mem[P2].
+*/ 
+case OP_OpFuncInteger: {          /* out2-prerelease */
+  pIn1 = &aMem[pOp->p1];
+  OpFunc *f = pIn1->u.pOpFunc;
+  pOut = &aMem[pOp->p2];
+  pOut->flags = MEM_Int;
+  pOut->u.i = nextInteger(f);
+  break;
+}
 
+/* OpCode: OpFuncReal P1 P2 * * * 
+** Synopsis: Next Real of Mem[P1] into Mem[P2]
+**
+** Loads the next double stored in the buffer of the OpFunc in Mem[P1]
+** into Mem[P2].
+*/ 
+case OP_OpFuncReal: {          /* out2-prerelease */
+  pIn1 = &aMem[pOp->p1];
+  OpFunc *f = pIn1->u.pOpFunc;
+  pOut = &aMem[pOp->p2];
+  pOut->flags = MEM_Real;
+  pOut->u.r = nextReal(f);
+  break;
+}
+
+/* OpCode: OpFuncString P1 P2 * * * 
+** Synopsis: Next String of Mem[P1] into Mem[P2]
+**
+** Loads the next string stored in the buffer of the OpFunc in Mem[P1]
+** into Mem[P2].
+*/
+case OP_OpFuncString: {          /* out2-prerelease */
+  pIn1 = &aMem[pOp->p1];
+  OpFunc *f = pIn1->u.pOpFunc;
+  pOut = &aMem[pOp->p2];
+  pOut->flags = MEM_Str|MEM_Static|MEM_Term;
+  nextString(f, pOut);
+  pOut->enc = encoding;
+  break;
+}
+
+/* Opcode:  OpFuncNext P1 P2 * * *
+** Synopsis: if Mem[P1] has not been fully read jump to P2
+**
+** Check if the buffer in Mem[P1] has been fully read.  If yes, continue 
+** else jump to P2.
+*/
+case OP_OpFuncNext: {            /* jump */
+  pIn1 = &aMem[pOp->p1];
+  OpFunc *f = pIn1->u.pOpFunc;
+  int more = moreRecords(f);
+  VdbeBranchTaken(more, 2);
+  if( more ) goto jump_to_p2;
+  break;
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /* Opcode: Noop * * * * *
 **
 ** Do nothing.  This instruction is often useful as a jump
@@ -7553,7 +8852,11 @@
 #ifdef SQLITE_DEBUG
     if( db->flags & SQLITE_VdbeTrace ){
       u8 opProperty = sqlite3OpcodeProperty[pOrigOp->opcode];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( rc!=0 ) logmsg(LOGMSG_ERROR, "rc=%d\n",rc);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       if( rc!=0 ) printf("rc=%d\n",rc);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       if( opProperty & (OPFLG_OUT2) ){
         registerTrace(pOrigOp->p2, &aMem[pOrigOp->p2]);
       }
@@ -7581,7 +8884,32 @@
                    (int)(pOp - aOp), p->zSql, p->zErrMsg);
   sqlite3VdbeHalt(p);
   if( rc==SQLITE_IOERR_NOMEM ) sqlite3OomFault(db);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* if it's one of our return codes, don't clobber it */
+  /* TODO: Make this into a switch (or its own function?). */
+  if( rc!=SQLITE_DEADLOCK
+   && rc!=SQLITE_CONSTRAINT
+   && rc!=SQLITE_ACCESS
+   && rc!=SQLITE_TOOBIG
+   && rc!=SQLITE_TIMEDOUT
+   && rc!=SQLITE_COST_TOO_HIGH
+   && rc!=SQLITE_NO_TEMPTABLES
+   && rc!=SQLITE_NO_TABLESCANS
+   && rc!=SQLITE_TRANTOOCOMPLEX
+   && rc!=SQLITE_CLIENT_CHANGENODE
+   && rc!=SQLITE_TRAN_CANCELLED
+   && rc!=SQLITE_TRAN_NOLOG
+   && rc!=SQLITE_TRAN_NOUNDO
+   && rc!=SQLITE_SCHEMA_REMOTE
+   && rc!=SQLITE_SCHEMA_DOHSQL
+   && rc!=SQLITE_SCHEMA_PUSH_REMOTE
+   && rc!=SQLITE_SCHEMA_PUSH_REMOTE_WRITE
+  ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   rc = SQLITE_ERROR;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( resetSchemaOnFault>0 ){
     sqlite3ResetOneSchema(db, resetSchemaOnFault-1);
   }
@@ -7602,9 +8930,11 @@
 #endif
   p->aCounter[SQLITE_STMTSTATUS_VM_STEP] += (int)nVmStep;
   sqlite3VdbeLeave(p);
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   assert( rc!=SQLITE_OK || nExtraDelete==0 
        || sqlite3_strlike("DELETE%",p->zSql,0)!=0 
   );
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   return rc;
 
   /* Jump to here if a string or blob larger than SQLITE_MAX_LENGTH
