--- sqlite-3.28.0/src/vdbemem.c	2026-01-09 16:11:01
+++ src/vdbemem.c	2025-09-24 15:21:58
@@ -18,6 +18,18 @@
 #include "sqliteInt.h"
 #include "vdbeInt.h"
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <arpa/inet.h>
+#include <inttypes.h>
+#include <flibc.h>
+#include <strings.h>
+#include <types.h>
+#include <util.h>
+#include "debug_switches.h"
+#include "logmsg.h"
+#include "str0.h"
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 #ifdef SQLITE_DEBUG
 /*
 ** Check invariants on a Mem object.
@@ -29,6 +41,7 @@
   /* If MEM_Dyn is set then Mem.xDel!=0.  
   ** Mem.xDel might not be initialized if MEM_Dyn is clear.
   */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   assert( (p->flags & MEM_Dyn)==0 || p->xDel!=0 );
 
   /* MEM_Dyn may only be set if Mem.szMalloc==0.  In this way we
@@ -36,6 +49,7 @@
   ** Mem.z = Mem.zMalloc without having to check Mem.flags&MEM_Dyn.
   ** That saves a few cycles in inner loops. */
   assert( (p->flags & MEM_Dyn)==0 || p->szMalloc==0 );
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   /* Cannot be both MEM_Int and MEM_Real at the same time */
   assert( (p->flags & (MEM_Int|MEM_Real))!=(MEM_Int|MEM_Real) );
@@ -69,8 +83,15 @@
   }
 
   /* The szMalloc field holds the correct memory allocation size */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /*
+  ** TODO: fix this assert -- either to correct spec or explain why
+  **       it is not needed.
+  */
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( p->szMalloc==0
        || p->szMalloc==sqlite3DbMallocSize(p->db,p->zMalloc) );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   /* If p holds a string or blob, the Mem.z must point to exactly
   ** one of the following:
@@ -80,6 +101,7 @@
   **   (3) An ephemeral string or blob
   **   (4) A static string or blob
   */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   if( (p->flags & (MEM_Str|MEM_Blob)) && p->n>0 ){
     assert( 
       ((p->szMalloc>0 && p->z==p->zMalloc)? 1 : 0) +
@@ -88,6 +110,7 @@
       ((p->flags&MEM_Static)!=0 ? 1 : 0) == 1
     );
   }
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   return 1;
 }
 #endif
@@ -195,8 +218,10 @@
   assert( bPreserve==0 || pMem->flags&(MEM_Blob|MEM_Str) );
   testcase( bPreserve && pMem->z==0 );
 
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
   assert( pMem->szMalloc==0
        || pMem->szMalloc==sqlite3DbMallocSize(pMem->db, pMem->zMalloc) );
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( pMem->szMalloc>0 && bPreserve && pMem->z==pMem->zMalloc ){
     pMem->z = pMem->zMalloc = sqlite3DbReallocOrFree(pMem->db, pMem->z, n);
     bPreserve = 0;
@@ -216,8 +241,18 @@
   if( bPreserve && pMem->z ){
     assert( pMem->z!=pMem->zMalloc );
     memcpy(pMem->zMalloc, pMem->z, pMem->n);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( pMem->flags&MEM_Xor ){
+      xorbufcpy(pMem->zMalloc, pMem->zMalloc, pMem->n);
+      pMem->flags &= ~MEM_Xor;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( (pMem->flags&MEM_Dyn)!=0 && pMem->xDel ){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( (pMem->flags&MEM_Dyn)!=0 ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( pMem->xDel!=0 && pMem->xDel!=SQLITE_DYNAMIC );
     pMem->xDel((void *)(pMem->z));
   }
@@ -248,7 +283,11 @@
   }
   assert( (pMem->flags & MEM_Dyn)==0 );
   pMem->z = pMem->zMalloc;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  pMem->flags &= (MEM_Null|MEM_Int|MEM_Real|MEM_Datetime|MEM_Interval);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pMem->flags &= (MEM_Null|MEM_Int|MEM_Real);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   return SQLITE_OK;
 }
 
@@ -355,7 +394,11 @@
   assert( pMem->db==0 || sqlite3_mutex_held(pMem->db->mutex) );
   assert( !(fg&MEM_Zero) );
   assert( !(fg&(MEM_Str|MEM_Blob)) );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  assert( fg&(MEM_Int|MEM_Real|MEM_Datetime|MEM_Interval) );
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( fg&(MEM_Int|MEM_Real) );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( !sqlite3VdbeMemIsRowSet(pMem) );
   assert( EIGHT_BYTE_ALIGNMENT(pMem) );
 
@@ -371,6 +414,110 @@
   ** 
   ** FIX ME: It would be better if sqlite3_snprintf() could do UTF-16.
   */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( fg&MEM_Interval ){
+    char tmp[64];
+    char *z;
+
+    if( pMem->du.tv.type==INTV_YM_TYPE ){
+      snprintf(tmp, sizeof(tmp), "%s%u-%2.2u",
+               (pMem->du.tv.sign==-1) ? "- " : "", pMem->du.tv.u.ym.years,
+               pMem->du.tv.u.ym.months);
+      tmp[sizeof(tmp) - 1] = 0;
+    }else if( pMem->du.tv.type==INTV_DS_TYPE ){
+      snprintf(tmp, sizeof(tmp), "%s%u %2.2u:%2.2u:%2.2u.%3.3u",
+               (pMem->du.tv.sign==-1)?"- ":"", pMem->du.tv.u.ds.days,
+               pMem->du.tv.u.ds.hours, pMem->du.tv.u.ds.mins,
+               pMem->du.tv.u.ds.sec, pMem->du.tv.u.ds.frac);
+      tmp[sizeof(tmp) - 1] = 0;
+    }else if( pMem->du.tv.type==INTV_DSUS_TYPE ){
+      snprintf(tmp, sizeof(tmp), "%s%u %2.2u:%2.2u:%2.2u.%6.6u",
+               (pMem->du.tv.sign==-1) ? "- " : "", pMem->du.tv.u.ds.days,
+               pMem->du.tv.u.ds.hours, pMem->du.tv.u.ds.mins,
+               pMem->du.tv.u.ds.sec, pMem->du.tv.u.ds.frac);
+      tmp[sizeof(tmp) - 1] = 0;
+    }else if( pMem->du.tv.type==INTV_DECIMAL_TYPE ){
+      /* max string is like 43 letters */
+      if( sqlite3DecimalToString(&pMem->du.tv.u.dec, tmp, sizeof(tmp)) ){
+        logmsg(LOGMSG_ERROR, "%s:%d failed to convert decimal to string\n", 
+               __FILE__, __LINE__);
+        return SQLITE_INTERNAL;
+      }
+    }else{
+      logmsg(LOGMSG_ERROR, "BUG %s:%d type=%d\n", __FILE__, __LINE__,
+             pMem->du.tv.type);
+      return SQLITE_INTERNAL;
+    }
+    if( pMem->zMalloc ){
+        sqlite3DbFree(pMem->db, pMem->zMalloc);
+        pMem->zMalloc = NULL;
+        pMem->szMalloc = 0;
+        pMem->z = NULL;
+    }
+    pMem->n = strlen(tmp);
+    z = sqlite3GlobalConfig.m.xMalloc(pMem->n + 2);
+    if( !z ) return SQLITE_NOMEM_BKPT;
+    memcpy(z, tmp, pMem->n);
+    z[pMem->n] = 0;
+    z[pMem->n+1] = 0;
+    pMem->z = z;
+    pMem->zMalloc = z;
+    pMem->szMalloc = pMem->n + 2;
+    pMem->enc = SQLITE_UTF8;
+    pMem->flags |= MEM_Str | MEM_Term | MEM_Dyn;
+    if( pMem->xDel ) pMem->xDel = 0;
+    if( bForce ) pMem->flags &= ~MEM_Interval;
+    sqlite3VdbeChangeEncoding(pMem, enc);
+    return SQLITE_OK;
+  }else if( fg&MEM_Datetime ){
+    char    tmp[64];
+    int     outdtsz;
+
+    if( pMem->zMalloc ){
+      sqlite3DbFree(pMem->db, pMem->zMalloc);
+      pMem->zMalloc = NULL;
+      pMem->szMalloc = 0;
+      pMem->z = NULL;
+    }
+
+    if( convMem2ClientDatetimeStr(pMem, tmp, sizeof(tmp), &outdtsz) ){ 
+      char *z;
+      if (pMem->db)
+          sqlite3ErrorWithMsg(pMem->db, SQLITE_CONV_ERROR,
+                              "can't convert datetime value to string");
+      pMem->n = strlen("conv_error");
+      z = sqlite3GlobalConfig.m.xMalloc(pMem->n+2);
+      if( !z ) return SQLITE_NOMEM;
+      memcpy(z, "conv_error", pMem->n);
+      z[pMem->n] = 0;
+      z[pMem->n+1] = 0;
+      pMem->z = z;
+      pMem->enc = SQLITE_UTF8;
+      pMem->flags |= MEM_Str | MEM_Term | MEM_Dyn;
+      if( pMem->xDel ) pMem->xDel = 0;
+      if( bForce ) pMem->flags &= ~MEM_Datetime;
+      sqlite3VdbeChangeEncoding(pMem, enc);
+      return SQLITE_CONV_ERROR;
+    }else{
+      char *z;
+      pMem->n = strlen(tmp);
+      z = sqlite3GlobalConfig.m.xMalloc(pMem->n+2);
+      if( !z ) return SQLITE_NOMEM;
+      memcpy(z, tmp, pMem->n);
+      z[pMem->n] = 0;
+      z[pMem->n+1] = 0;
+      pMem->z = z;
+      pMem->zMalloc = z;
+      pMem->szMalloc = pMem->n + 2;
+      pMem->enc = SQLITE_UTF8;
+      pMem->flags |= MEM_Str | MEM_Term | MEM_Dyn;
+      if( pMem->xDel ) pMem->xDel = 0;
+      if( bForce ) pMem->flags &= ~MEM_Datetime;
+      sqlite3VdbeChangeEncoding(pMem, enc);
+      return SQLITE_OK;
+    }
+  }else
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( fg & MEM_Int ){
     sqlite3_snprintf(nByte, pMem->z, "%lld", pMem->u.i);
   }else{
@@ -461,7 +608,11 @@
     assert( (p->flags & MEM_Agg)==0 );
     testcase( p->flags & MEM_Dyn );
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( p->flags&MEM_Dyn && p->xDel ){
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( p->flags&MEM_Dyn ){
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( p->xDel!=SQLITE_DYNAMIC && p->xDel!=0 );
     p->xDel((void *)p->z);
   }
@@ -559,6 +710,25 @@
     return pMem->u.i;
   }else if( flags & MEM_Real ){
     return doubleToInt64(pMem->u.r);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  }else if( pMem->flags & MEM_Interval ){
+    if( pMem->du.tv.type==INTV_YM_TYPE ){
+      return pMem->du.tv.sign * (i64)(pMem->du.tv.u.ym.years * 12 +
+                                      pMem->du.tv.u.ym.months);
+    }else if( pMem->du.tv.type==INTV_DS_TYPE ||
+              pMem->du.tv.type==INTV_DSUS_TYPE ){
+      return pMem->du.tv.sign * (i64)(pMem->du.tv.u.ds.days * 24 * 3600 +
+                                      pMem->du.tv.u.ds.hours * 3600 +
+                                      pMem->du.tv.u.ds.mins * 60 +
+                                      pMem->du.tv.u.ds.sec);
+    }else{
+      logmsg(LOGMSG_ERROR, "%s:%d: Cannot convert a decimal to integer\n",
+             __FILE__, __LINE__);
+      return 0;
+    }
+  }else if( pMem->flags & MEM_Datetime ){
+    return pMem->du.dt.dttz_sec;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }else if( flags & (MEM_Str|MEM_Blob) ){
     assert( pMem->z || pMem->n==0 );
     return memIntValue(pMem);
@@ -586,6 +756,23 @@
     return pMem->u.r;
   }else if( pMem->flags & MEM_Int ){
     return (double)pMem->u.i;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  }else if( pMem->flags & MEM_Interval ){
+    char tmp[128];
+    switch( pMem->du.tv.type ){
+      case INTV_YM_TYPE:
+      case INTV_DS_TYPE:
+      case INTV_DSUS_TYPE:
+        return interval_to_double(&pMem->du.tv);
+      default:
+        /* WARNING: lossy */
+        decQuadToString(&pMem->du.tv.u.dec, tmp);
+        return atof(tmp);
+    }
+  }else if( pMem->flags & MEM_Datetime ){
+    return pMem->du.dt.dttz_sec + pMem->du.dt.dttz_frac /
+           (pMem->du.dt.dttz_prec == DTTZ_PREC_MSEC ? 1E3 : 1E6);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }else if( pMem->flags & (MEM_Str|MEM_Blob) ){
     return memRealValue(pMem);
   }else{
@@ -682,8 +869,28 @@
 int sqlite3VdbeMemNumerify(Mem *pMem){
   if( (pMem->flags & (MEM_Int|MEM_Real|MEM_Null))==0 ){
     int rc;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    assert( (pMem->flags & (MEM_Blob|MEM_Str|MEM_Interval|MEM_Datetime))!=0 );
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( (pMem->flags & (MEM_Blob|MEM_Str))!=0 );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( pMem->db==0 || sqlite3_mutex_held(pMem->db->mutex) );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /* SQLite used to bzero `Vdbe.aVar' (was named `azVar' back then) in
+       sqlite3VdbeMakeReady(). We then checked for a NULL `Mem.z' here to see
+       whether the Mem structure had a valid string representation. However it
+       is no longer applicable in newer SQLite versions. Newer versions do not
+       bzero `aVar' in MakeReady and as a result a bound parameter's `z' may
+       point to an invalid address. We fix this by checking against `Mem.flags'
+       instead. I believe the change not only fixes the bound parameter issue
+       here, but also covers other unaccounted cases where we assume a Mem
+       struct starts with all zeros. */
+    if ( (pMem->flags & (MEM_Str|MEM_Blob))==0 ){
+      pMem->u.r = sqlite3VdbeRealValue(pMem);
+      MemSetTypeFlag(pMem, MEM_Real);
+      sqlite3VdbeIntegerAffinity(pMem);
+    }else{
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     rc = sqlite3Atoi64(pMem->z, &pMem->u.i, pMem->n, pMem->enc);
     if( rc==0 ){
       MemSetTypeFlag(pMem, MEM_Int);
@@ -697,12 +904,19 @@
         MemSetTypeFlag(pMem, MEM_Real);
       }
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
   assert( (pMem->flags & (MEM_Int|MEM_Real|MEM_Null))!=0 );
   pMem->flags &= ~(MEM_Str|MEM_Blob|MEM_Zero);
   return SQLITE_OK;
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+void sqlite3SetConversionError(void);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Cast the datatype of the value in pMem according to the affinity
 ** "aff".  Casting is different from applying affinity in that a cast
@@ -710,8 +924,15 @@
 ** affinity even if that results in loss of data.  This routine is
 ** used (for example) to implement the SQL "cast()" operator.
 */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+int sqlite3VdbeMemCast(Vdbe *p, Mem *pMem, u8 aff, u8 encoding){
+  int rc = SQLITE_OK, rc2;
+  sqlite3 *db = p->db;
+  if( pMem->flags & MEM_Null ) return rc;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 void sqlite3VdbeMemCast(Mem *pMem, u8 aff, u8 encoding){
   if( pMem->flags & MEM_Null ) return;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   switch( aff ){
     case SQLITE_AFF_BLOB: {   /* Really a cast to BLOB */
       if( (pMem->flags & MEM_Blob)==0 ){
@@ -735,16 +956,63 @@
       sqlite3VdbeMemRealify(pMem);
       break;
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    case SQLITE_AFF_SMALL: {
+      sqlite3VdbeMemRealify(pMem);
+      break;
+    }
+    case SQLITE_AFF_DATETIME: {
+      rc2 = sqlite3VdbeMemDatetimefyTz(pMem, pMem->tz);
+      if( rc2 && !debug_switch_ignore_datetime_cast_failures() ){
+        sqlite3SetString(&p->zErrMsg, db, "cast to datetime failed");
+        sqlite3SetConversionError();
+        rc = rc2;
+      }
+      break;
+    }
+    case SQLITE_AFF_INTV_YE:
+    case SQLITE_AFF_INTV_MO:
+    case SQLITE_AFF_INTV_DY:
+    case SQLITE_AFF_INTV_HO:
+    case SQLITE_AFF_INTV_MI:
+    case SQLITE_AFF_INTV_SE: {
+      rc2 = sqlite3VdbeMemIntervalfy(pMem, aff);
+      if( rc2 && !debug_switch_ignore_datetime_cast_failures() ){
+        sqlite3SetString(&p->zErrMsg, db, "cast to interval failed");
+        sqlite3SetConversionError();
+        rc = rc2;
+      }
+      break;
+    }
+    case SQLITE_AFF_DECIMAL: {
+      rc2 = sqlite3VdbeMemDecimalfy(pMem);
+      if( rc2 && !debug_switch_ignore_datetime_cast_failures() ){
+        sqlite3SetString(&p->zErrMsg, db, "cast to decimal failed");
+        sqlite3SetConversionError();
+        rc = rc2;
+      }
+      break;
+    }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     default: {
       assert( aff==SQLITE_AFF_TEXT );
       assert( MEM_Str==(MEM_Blob>>3) );
       pMem->flags |= (pMem->flags&MEM_Blob)>>3;
       sqlite3ValueApplyAffinity(pMem, SQLITE_AFF_TEXT, encoding);
       assert( pMem->flags & MEM_Str || pMem->db->mallocFailed );
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if ( (pMem->flags & MEM_Str) & !pMem->db->mallocFailed ){
+        pMem->flags = MEM_Str | (~(MEM_AffMask|MEM_Zero));
+      }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       pMem->flags &= ~(MEM_Int|MEM_Real|MEM_Blob|MEM_Zero);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       break;
     }
   }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  return rc;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 }
 
 /*
@@ -787,6 +1055,7 @@
 ** Delete any previous value and set the value to be a BLOB of length
 ** n containing all zeros.
 */
+#ifndef SQLITE_OMIT_INCRBLOB
 void sqlite3VdbeMemSetZeroBlob(Mem *pMem, int n){
   sqlite3VdbeMemRelease(pMem);
   pMem->flags = MEM_Blob|MEM_Zero;
@@ -795,7 +1064,22 @@
   pMem->u.nZero = n;
   pMem->enc = SQLITE_UTF8;
   pMem->z = 0;
+}
+#else
+int sqlite3VdbeMemSetZeroBlob(Mem *pMem, int n){
+  int nByte = n>0?n:1;
+  if( sqlite3VdbeMemGrow(pMem, nByte, 0) ){
+    return SQLITE_NOMEM_BKPT;
+  }
+  assert( pMem->z!=0 );
+  assert( sqlite3DbMallocSize(pMem->db, pMem->z)>=nByte );
+  memset(pMem->z, 0, nByte);
+  pMem->n = n>0?n:0;
+  pMem->flags = MEM_Blob;
+  pMem->enc = SQLITE_UTF8;
+  return SQLITE_OK;
 }
+#endif
 
 /*
 ** The pMem is known to contain content that needs to be destroyed prior
@@ -856,7 +1140,7 @@
 }
 #endif
 
-#ifdef SQLITE_DEBUG
+#if defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_DEBUG)
 /*
 ** Return true if the Mem holds a RowSet object.  This routine is intended
 ** for use inside of assert() statements.
@@ -865,7 +1149,7 @@
   return (pMem->flags&(MEM_Blob|MEM_Dyn))==(MEM_Blob|MEM_Dyn)
          && pMem->xDel==sqlite3RowSetDelete;
 }
-#endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_DEBUG) */
 
 /*
 ** Delete any previous value and set the value of pMem to be an
@@ -905,6 +1189,10 @@
 }
 
 #ifdef SQLITE_DEBUG
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <memcompare.c>
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** This routine prepares a memory cell for modification by breaking
 ** its link to a shallow copy and by marking any current shallow
@@ -1097,9 +1385,60 @@
   if( nByte>iLimit ){
     return SQLITE_TOOBIG;
   }
+
+  return SQLITE_OK;
+}
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+
+int sqlite3VdbeMemSetMasterResolve(
+  Mem *pMem           /* Memory cell which will be resolved on master */
+) {
+  sqlite3VdbeMemSetNull(pMem);
+  pMem->flags = MEM_Master;
+  return SQLITE_OK;
+}
+
+/*
+** Change the value of a Mem to be a datetime.
+*/
+int sqlite3VdbeMemSetDatetime(
+  Mem *pMem,          /* Memory cell to set to datetime value */
+  dttz_t * dt,        /* datetime fields */
+  const char * tz     /* tzname propagation, for conversion purposes */ 
+){
+  sqlite3VdbeMemSetNull(pMem);
+  pMem->du.dt = *dt;
+  pMem->tz = tz;
+  pMem->flags = MEM_Datetime;
+  return SQLITE_OK;
+}
+
+/*
+** Change the value of a Mem to be an interval.
+*/
+int sqlite3VdbeMemSetInterval(
+  Mem *pMem,          /* Memory cell to set to interval value */
+  intv_t * tv         /* interval fields */
+){
+  sqlite3VdbeMemSetNull(pMem);
+  pMem->du.tv = *tv;
+  pMem->flags = MEM_Interval;
+  return SQLITE_OK;
+}
 
+/*
+** Delete any previous value and set the value stored in *pMem to val,
+** manifest type Decimal.
+*/
+int sqlite3VdbeMemSetDecimal(Mem *pMem, decQuad *val){
+  sqlite3VdbeMemSetNull(pMem);
+  pMem->du.tv.u.dec = *val;
+  pMem->du.tv.type = SQLITE_DECIMAL;
+  pMem->flags = MEM_Interval;
   return SQLITE_OK;
 }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
 /*
 ** Move data out of a btree key or data field and into a Mem structure.
@@ -1194,7 +1533,13 @@
     }
     sqlite3VdbeMemNulTerminate(pVal); /* IMP: R-31275-44060 */
   }else{
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( sqlite3VdbeMemStringify(pVal, enc, 0)!=SQLITE_OK ){
+      return 0;
+    }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3VdbeMemStringify(pVal, enc, 0);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( 0==(1&SQLITE_PTR_TO_INT(pVal->z)) );
   }
   assert(pVal->enc==(enc & ~SQLITE_UTF16_ALIGNED) || pVal->db==0
@@ -1364,7 +1709,8 @@
       goto value_from_function_out;
     }
     for(i=0; i<nVal; i++){
-      rc = sqlite3ValueFromExpr(db, pList->a[i].pExpr, enc, aff, &apVal[i]);
+      rc = sqlite3Stat4ValueFromExpr(pCtx->pParse, pList->a[i].pExpr, aff,
+                                     &apVal[i]);
       if( apVal[i]==0 || rc!=SQLITE_OK ) goto value_from_function_out;
     }
   }
@@ -1379,6 +1725,9 @@
   memset(&ctx, 0, sizeof(ctx));
   ctx.pOut = pVal;
   ctx.pFunc = pFunc;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  ctx.pVdbe = db->pVdbe;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pFunc->xSFunc(&ctx, nVal, apVal);
   if( ctx.isError ){
     rc = ctx.isError;
@@ -1386,11 +1735,13 @@
   }else{
     sqlite3ValueApplyAffinity(pVal, aff, SQLITE_UTF8);
     assert( rc==SQLITE_OK );
+#if 0 /* Not reachable except after a prior failure */
     rc = sqlite3VdbeChangeEncoding(pVal, enc);
     if( rc==SQLITE_OK && sqlite3VdbeMemTooBig(pVal) ){
       rc = SQLITE_TOOBIG;
       pCtx->pParse->nErr++;
     }
+#endif
   }
   pCtx->pParse->rc = rc;
 
@@ -1437,6 +1788,10 @@
   const char *zNeg = "";
   int rc = SQLITE_OK;
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* first in list, we dont support more than one vdbe on a db */
+  Vdbe *p = db->pVdbe;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   assert( pExpr!=0 );
   while( (op = pExpr->op)==TK_UPLUS || op==TK_SPAN ) pExpr = pExpr->pLeft;
 #if defined(SQLITE_ENABLE_STAT3_OR_STAT4)
@@ -1456,7 +1811,15 @@
     rc = valueFromExpr(db, pExpr->pLeft, enc, aff, ppVal, pCtx);
     testcase( rc!=SQLITE_OK );
     if( *ppVal ){
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      if( aff==SQLITE_AFF_DATETIME ){
+        (*ppVal)->tz = p->tzname;
+        (*ppVal)->dtprec = p->dtprec;
+      }
+      sqlite3VdbeMemCast(p, *ppVal, aff, SQLITE_UTF8);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3VdbeMemCast(*ppVal, aff, SQLITE_UTF8);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3ValueApplyAffinity(*ppVal, affinity, SQLITE_UTF8);
     }
     return rc;
@@ -1511,7 +1874,7 @@
   }else if( op==TK_NULL ){
     pVal = valueNew(db, pCtx);
     if( pVal==0 ) goto no_mem;
-    sqlite3VdbeMemNumerify(pVal);
+    sqlite3VdbeMemSetNull(pVal);
   }
 #ifndef SQLITE_OMIT_BLOB_LITERAL
   else if( op==TK_BLOB ){
@@ -1628,7 +1991,83 @@
   sqlite3InsertBuiltinFuncs(aAnalyzeTableFuncs, ArraySize(aAnalyzeTableFuncs));
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
 /*
+** There is no literal notation for datetime and interval. For these columns,
+** try converting string literal to column's affinity for stat4 comparison.
+*/
+static int castExpr(
+  sqlite3 *db,
+  Expr *pExpr,
+  u8 affinity,
+  sqlite3_value **ppVal,
+  struct ValueNewStat4Ctx *pAlloc
+){
+  dttz_t dt;
+  intv_ds_t ds;
+  int type, sign;
+  uint64_t n0, n1;
+  Vdbe *v = db->pVdbe;
+  const char *z = pExpr->u.zToken;
+  switch( affinity ){
+    case SQLITE_AFF_DATETIME: {
+      if( str_to_dttz(z, strlen(z), v->tzname, &dt, v->dtprec)==0 ){
+        sqlite3_value *pVal = valueNew(db, pAlloc);
+        if( pVal==NULL ){
+          return SQLITE_NOMEM_BKPT;
+        }
+        pVal->du.dt = dt;
+        pVal->tz = v->tzname;
+        pVal->dtprec = v->dtprec;
+        pVal->flags = MEM_Datetime;
+        *ppVal = pVal;
+        return 0;
+      }
+      break;
+    }
+    case SQLITE_AFF_INTV_MO: {
+      type = INTV_YM_TYPE;
+      if( str_to_interval(z, strlen(z), &type, &n0, &n1, &ds, &sign)==0
+       && type==0 ){ /* parsed into n0,n1 */
+        sqlite3_value *pVal = valueNew(db, pAlloc);
+        if( pVal==NULL ){
+          return SQLITE_NOMEM_BKPT;
+        }
+        intv_t *tv = &pVal->du.tv;
+        tv->u.ym.years = n0;
+        tv->u.ym.months = n1;
+        tv->sign = sign;
+        tv->type = INTV_YM_TYPE;
+        pVal->flags = MEM_Interval;
+        *ppVal = pVal;
+        return 0;
+      }
+      break;
+    }
+    case SQLITE_AFF_INTV_SE: {
+      type = INTV_DSUS_TYPE;
+      if( str_to_interval(z, strlen(z), &type, &n0, &n1, &ds, &sign)==0
+       && type==1 ){ /* parsed into ds */
+        sqlite3_value *pVal = valueNew(db, pAlloc);
+        if( pVal==NULL ){
+          return SQLITE_NOMEM_BKPT;
+        }
+        intv_t *tv = &pVal->du.tv;
+        tv->u.ds = ds;
+        tv->sign = sign;
+        tv->type = ds.prec == DTTZ_PREC_MSEC ? INTV_DS_TYPE : INTV_DSUS_TYPE;
+        pVal->flags = MEM_Interval;
+        *ppVal = pVal;
+        return 0;
+      }
+      break;
+    }
+  }
+  return valueFromExpr(db, pExpr, ENC(db), affinity, ppVal, pAlloc);
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
+/*
 ** Attempt to extract a value from pExpr and use it to construct *ppVal.
 **
 ** If pAlloc is not NULL, then an UnpackedRecord object is created for
@@ -1679,7 +2118,15 @@
       }
     }
   }else{
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( pExpr->op==TK_STRING && affinity!=SQLITE_AFF_TEXT ){
+      rc = castExpr(db, pExpr, affinity, &pVal, pAlloc);
+    }else{
+      rc = valueFromExpr(db, pExpr, ENC(db), affinity, &pVal, pAlloc);
+    }
+#else
     rc = valueFromExpr(db, pExpr, ENC(db), affinity, &pVal, pAlloc);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   }
 
   assert( pVal==0 || pVal->db==db );
@@ -1775,6 +2222,10 @@
   return stat4ValueFromExpr(pParse, pExpr, affinity, 0, ppVal);
 }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+#include <memcompare.c>
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
 /*
 ** Extract the iCol-th column from the nRec-byte record in pRec.  Write
 ** the column value into *ppVal.  If *ppVal is initially NULL then a new
@@ -1819,6 +2270,11 @@
     if( pMem==0 ) return SQLITE_NOMEM_BKPT;
   }
   sqlite3VdbeSerialGet(&a[iField-szField], t, pMem);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  if( 11==t ){ /* see sqlite3VdbeSerialGet, 11 is datetime blob */
+    pMem->tz = db->pVdbe->tzname;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   pMem->enc = ENC(db);
   return SQLITE_OK;
 }
@@ -1888,4 +2344,1051 @@
   }
   if( p->flags & MEM_Null ) return 0;
   return valueBytes(pVal, enc);
+}
+
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+/*
+** Add MEM_Datetime to the set of representations for the given Mem. 
+*/
+int sqlite3VdbeMemDatetimefyTz(Mem *pMem, const char *tz){
+  int fg = pMem->flags;
+  if( fg & MEM_Null )
+    return SQLITE_OK;
+  if( fg & MEM_Interval )
+    return SQLITE_ERROR;
+  if( pMem->flags & MEM_Blob )
+    return SQLITE_ERROR; 
+  if(pMem->flags & MEM_Real) {
+    if(real_to_dttz(pMem->u.r, &pMem->du.dt, pMem->dtprec) != 0)
+      return SQLITE_ERROR;
+    pMem->flags = MEM_Datetime;
+  }else if( pMem->flags & MEM_Int ){
+    if (int_to_dttz(pMem->u.i, &pMem->du.dt, pMem->dtprec) != 0)
+      return SQLITE_ERROR;
+    pMem->flags = MEM_Datetime;
+  }else if( pMem->flags & MEM_Str ){
+    if (str_to_dttz(pMem->z, pMem->n, pMem->tz ? pMem->tz : tz,
+     &pMem->du.dt, pMem->dtprec) != 0) {
+      return SQLITE_ERROR;
+    }
+    if( pMem->szMalloc ){
+      sqlite3DbFreeNN(pMem->db, pMem->zMalloc);
+      pMem->szMalloc = 0;
+    }
+    pMem->n = 0;
+    pMem->z = 0;
+    /*no MEM_Blob here*/
+    pMem->flags = MEM_Datetime;
+  }
+  return SQLITE_OK;
 }
+
+/*
+** Add MEM_Datetime to the set of representations for the given Mem. 
+*/
+int sqlite3VdbeMemDatetimefy(Mem *pMem){
+   return sqlite3VdbeMemDatetimefyTz(pMem, NULL);
+}
+
+/*
+** Add MEM_Datetime to the set of representations for the given Mem. 
+*/
+int sqlite3VdbeMemDecimalfy(Mem *pMem)
+{
+   decContext     dfp_ctx;
+   int            rc = SQLITE_OK;
+   int            fg = pMem->flags;
+
+   if(fg & MEM_Null) return rc;
+
+   /* cannot convert a datetime to a decimal */
+   if(fg & MEM_Datetime) return SQLITE_ERROR; 
+
+   /* cannot convert a blob to a decimal */
+   if(fg & MEM_Blob) return SQLITE_ERROR; 
+
+   /* cannot convert a interval to a decimal */
+   if(fg & MEM_Interval) return SQLITE_ERROR;
+
+   /* TODO: we need a custom routine */
+   if(fg & MEM_Real) return SQLITE_ERROR;
+
+   if(fg & MEM_Int)
+   {
+      char str[32]; 
+      void *ret = NULL;
+
+      /* we can do better here, but we'll optimize later if need to */
+      snprintf(str, sizeof(str), "%lld", pMem->u.i);
+
+      dec_ctx_init(&dfp_ctx, DEC_INIT_DECQUAD, gbl_decimal_rounding);
+
+      ret = decQuadFromString( &pMem->du.tv.u.dec, str, &dfp_ctx);
+
+      if (ret)
+      {
+         if (dfp_conv_check_status( &dfp_ctx, "string", "quad"))
+         {
+            rc = SQLITE_ERROR;
+         }
+         else
+         {
+            pMem->du.tv.type = INTV_DECIMAL_TYPE;
+            /* we ignore sign, it's embedded */
+            pMem->du.tv.sign = 0;
+            rc = SQLITE_OK;
+         }
+      }
+      else
+         rc = SQLITE_ERROR;
+      if(0){
+         char dbg[1024];
+
+         decQuadToString( &pMem->du.tv.u.dec, dbg );
+         printf("Parsed to quad %s\n", dbg);
+      }
+   }
+
+   else if(fg & MEM_Str)
+   {
+      void *ret = NULL;
+      
+      dec_ctx_init(&dfp_ctx, DEC_INIT_DECQUAD, gbl_decimal_rounding);
+
+
+      sqlite3VdbeMemNulTerminate(pMem);
+
+      ret = decQuadFromString( &pMem->du.tv.u.dec, pMem->z, &dfp_ctx);
+      
+      if (ret)
+      {
+         if (dfp_conv_check_status( &dfp_ctx, "string", "quad"))
+         {
+            rc = SQLITE_ERROR;
+         }
+         else
+         {
+            pMem->du.tv.type = INTV_DECIMAL_TYPE;
+            /* we ignore sign, it's embedded */
+            pMem->du.tv.sign = 0;
+            rc = SQLITE_OK;
+         }
+      }
+      else
+         rc = SQLITE_ERROR;
+      if(0){
+         char dbg[1024];
+
+         decQuadToString( &pMem->du.tv.u.dec, dbg );
+         printf("Parsed to quad %s\n", dbg);
+      }
+   }
+
+   sqlite3VdbeMemRelease(pMem);
+   pMem->flags = MEM_Interval;
+
+   return rc;
+}
+
+
+static char * skipsp(char *str){
+    while (str && str[0] == ' ') str++;
+    return str;
+}
+
+#include <alloca.h>
+
+/*
+** Add MEM_Interval to the set of representations for the given Mem. 
+*/
+int sqlite3VdbeMemIntervalfy(Mem *pMem, int type){
+  uint64_t tmp, tmp2;
+  int fg = pMem->flags;
+
+  if( (fg & MEM_Interval) || (fg & MEM_Null) ) return SQLITE_OK;
+
+  if( type == SQLITE_AFF_DECIMAL ){
+    logmsg(LOGMSG_FATAL, "BUG, wrong function called, use Decimalfy %s:%d\n",
+          __FILE__, __LINE__);
+    abort();
+  }
+
+  bzero(&pMem->du, sizeof(pMem->du));
+  if( pMem->flags & MEM_Int ){
+    int_to_interval(pMem->u.i, &tmp, &tmp2, &pMem->du.tv.sign);
+    goto num;
+  }else if( pMem->flags & MEM_Real ){
+    double_to_interval(pMem->u.r, &tmp, &tmp2, &pMem->du.tv.sign);
+    goto num;
+  }else if( pMem->flags & MEM_Str ){
+    int intv;
+    switch( type ){
+      case SQLITE_AFF_INTV_DY:
+      case SQLITE_AFF_INTV_HO:
+      case SQLITE_AFF_INTV_MI:
+      case SQLITE_AFF_INTV_SE: {
+        intv = INTV_DSUS_TYPE;
+        break;
+      }
+      case SQLITE_AFF_INTV_YE:
+      case SQLITE_AFF_INTV_MO: {
+        intv = INTV_YM_TYPE;
+        break;
+      }       
+      default: {
+          return SQLITE_ERROR;
+      }
+    }
+    if( str_to_interval(pMem->z, pMem->n, &intv, &tmp, &tmp2,
+     &pMem->du.tv.u.ds, &pMem->du.tv.sign) != 0 ){
+      return SQLITE_ERROR;
+    }
+    if( intv == 0 ){
+      switch( type ){
+        case SQLITE_AFF_INTV_YE:
+        case SQLITE_AFF_INTV_MO: {
+          pMem->du.tv.type = INTV_YM_TYPE;
+          pMem->du.tv.u.ym.years = tmp;
+          pMem->du.tv.u.ym.months = tmp2;
+          break;
+        }
+        default: {
+          return SQLITE_ERROR;
+        }
+      }
+    }else if( intv == 1) {
+      switch( type ){
+        case SQLITE_AFF_INTV_DY:
+        case SQLITE_AFF_INTV_HO:
+        case SQLITE_AFF_INTV_MI: {
+          pMem->du.tv.type = INTV_DSUS_TYPE;
+        }
+        case SQLITE_AFF_INTV_SE: {
+          pMem->du.tv.type = (pMem->du.tv.u.ds.prec == DTTZ_PREC_MSEC) ?
+            INTV_DS_TYPE : INTV_DSUS_TYPE;
+          break;
+        }
+        default: {
+          return SQLITE_ERROR;
+        }
+      }
+    } else if( intv == 2 ){
+num:  switch( type ){
+        case SQLITE_AFF_INTV_YE: {
+          pMem->du.tv.type = INTV_YM_TYPE;
+          pMem->du.tv.u.ym.years = tmp;
+          break;
+        }
+        case SQLITE_AFF_INTV_MO: {
+          pMem->du.tv.type = INTV_YM_TYPE;
+          pMem->du.tv.u.ym.months = tmp;
+          break;
+        }
+        case SQLITE_AFF_INTV_DY: {
+          pMem->du.tv.type = INTV_DS_TYPE;
+          _setIntervalDS( &pMem->du.tv.u.ds, 24*3600*tmp, 0);
+          break;
+        }
+        case SQLITE_AFF_INTV_HO: {
+          pMem->du.tv.type = INTV_DS_TYPE;
+          _setIntervalDS( &pMem->du.tv.u.ds, 3600*tmp, 0);
+          break;
+        }
+        case SQLITE_AFF_INTV_MI: {
+          pMem->du.tv.type = INTV_DS_TYPE;
+          _setIntervalDS( &pMem->du.tv.u.ds, 60*tmp, 0);
+          break;
+        }
+        case SQLITE_AFF_INTV_SE: {
+          if ((int)(tmp2 / 1E3) * 1000 == tmp2) {
+            pMem->du.tv.type = INTV_DS_TYPE;
+            _setIntervalDS(&pMem->du.tv.u.ds, tmp, tmp2 / 1000);
+          } else {
+            pMem->du.tv.type = INTV_DSUS_TYPE;
+            _setIntervalDSUS(&pMem->du.tv.u.ds, tmp, tmp2);
+          }
+          break;
+        }
+        default: {
+          return SQLITE_ERROR;
+        }
+      }
+    }else{
+      return SQLITE_ERROR;
+    }
+  }else{
+    return SQLITE_ERROR;
+  }
+  pMem->flags &= ~(MEM_TypeMask);
+  pMem->flags |= MEM_Interval;
+  if (pMem->du.tv.type == INTV_YM_TYPE)
+    _normalizeIntervalYM(&pMem->du.tv.u.ym);
+  return SQLITE_OK;
+}
+
+/*
+**  Operate on two intervals, res = a opcode b;
+*/
+int sqlite3VdbeMemIntervalAndInterval(
+  const Mem *a,
+  const Mem *b,
+  int opcode,
+  Mem * res
+){
+  bzero(res, sizeof(Mem));
+  res->flags |= MEM_Interval;
+  long long aa = 0, bb = 0;
+
+  switch( a->du.tv.type ){
+    case INTV_YM_TYPE: {
+      aa = a->du.tv.sign*(int)(a->du.tv.u.ym.months + a->du.tv.u.ym.years*12);
+      bb = b->du.tv.sign*(int)(b->du.tv.u.ym.months + b->du.tv.u.ym.years*12);
+
+      if( opcode == OP_Add ){
+        aa += bb;
+      }else if(opcode == OP_Subtract ){
+        aa -= bb;
+      }else return SQLITE_ERROR;
+
+      res->du.tv.type          = INTV_YM_TYPE;
+      res->du.tv.sign          = (aa<0)?-1:1;
+      res->du.tv.u.ym.months   = aa*res->du.tv.sign;
+      res->du.tv.u.ym.years    = 0;
+
+      _normalizeIntervalYM(&res->du.tv.u.ym);
+      break;
+    }
+    case INTV_DS_TYPE: {
+      if( b->du.tv.type == INTV_DS_TYPE ){
+        aa = a->du.tv.sign*(long long)(a->du.tv.u.ds.frac +
+                                       1000LL*           a->du.tv.u.ds.sec +
+                                       1000LL*60*        a->du.tv.u.ds.mins +
+                                       1000LL*3600*      a->du.tv.u.ds.hours +
+                                       1000LL*3600*24*   a->du.tv.u.ds.days);
+        bb = b->du.tv.sign*(long long)(b->du.tv.u.ds.frac +
+                                       1000LL*           b->du.tv.u.ds.sec +
+                                       1000LL*60*        b->du.tv.u.ds.mins +
+                                       1000LL*3600*      b->du.tv.u.ds.hours +
+                                       1000LL*3600*24*   b->du.tv.u.ds.days);
+
+        if( opcode == OP_Add ){
+          aa += bb;
+        }else if( opcode == OP_Subtract ){
+          aa -= bb;
+        }else return SQLITE_ERROR;
+
+        res->du.tv.type          = INTV_DS_TYPE;
+        res->du.tv.sign          = (aa<0)?-1:1;
+        aa = aa*res->du.tv.sign;
+        _setIntervalDS(&res->du.tv.u.ds,
+                       (((long long)aa)*res->du.tv.sign)/1000,
+                       (int)((aa*res->du.tv.sign)%1000));
+      }else{ // promote operand a to microsecond resolution
+        aa = a->du.tv.sign*(long long)(
+                                       1000LL*            a->du.tv.u.ds.frac +
+                                       1000000LL*         a->du.tv.u.ds.sec +
+                                       1000000LL*60*      a->du.tv.u.ds.mins +
+                                       1000000LL*3600*    a->du.tv.u.ds.hours +
+                                       1000000LL*3600*24* a->du.tv.u.ds.days);
+        bb = b->du.tv.sign*(long long)(
+                                       b->du.tv.u.ds.frac +
+                                       1000000LL*         b->du.tv.u.ds.sec +
+                                       1000000LL*60*      b->du.tv.u.ds.mins +
+                                       1000000LL*3600*    b->du.tv.u.ds.hours +
+                                       1000000LL*3600*24* b->du.tv.u.ds.days);
+
+        if( opcode == OP_Add ){
+          aa += bb;
+        }else if( opcode == OP_Subtract ){
+          aa -= bb;
+        }else return SQLITE_ERROR;
+
+        // promote the result of the expression to microsecond resolution
+        res->du.tv.type          = INTV_DSUS_TYPE;
+        res->du.tv.sign          = (aa<0)?-1:1;
+        aa = aa*res->du.tv.sign;
+        _setIntervalDSUS(&res->du.tv.u.ds,
+                         (((long long)aa)*res->du.tv.sign)/1000000,
+                         (int)((aa*res->du.tv.sign)%1000000));
+      }
+
+      break;
+    }
+    case INTV_DSUS_TYPE: {
+      if( b->du.tv.type == INTV_DS_TYPE ){
+        // promote operand b to microsecond resolution
+        aa = a->du.tv.sign*(long long)(
+                                       a->du.tv.u.ds.frac +
+                                       1000000LL*         a->du.tv.u.ds.sec +
+                                       1000000LL*60*      a->du.tv.u.ds.mins +
+                                       1000000LL*3600*    a->du.tv.u.ds.hours +
+                                       1000000LL*3600*24* a->du.tv.u.ds.days);
+        bb = b->du.tv.sign*(long long)(
+                                       1000LL*            b->du.tv.u.ds.frac +
+                                       1000000LL*         b->du.tv.u.ds.sec +
+                                       1000000LL*60*      b->du.tv.u.ds.mins +
+                                       1000000LL*3600*    b->du.tv.u.ds.hours +
+                                       1000000LL*3600*24* b->du.tv.u.ds.days);
+
+        if (opcode == OP_Add ){
+          aa += bb;
+        }else if( opcode == OP_Subtract ){
+          aa -= bb;
+        }else return SQLITE_ERROR;
+
+        // promote the result of the expression to microsecond resolution
+        res->du.tv.type          = INTV_DSUS_TYPE;
+        res->du.tv.sign          = (aa<0)?-1:1;
+        aa = aa*res->du.tv.sign;
+        _setIntervalDSUS(&res->du.tv.u.ds,
+                         (((long long)aa)*res->du.tv.sign)/1000000,
+                         (int)((aa*res->du.tv.sign)%1000000));
+      }else{
+        aa = a->du.tv.sign*(long long)(
+                                       a->du.tv.u.ds.frac +
+                                       1000000LL*         a->du.tv.u.ds.sec +
+                                       1000000LL*60*      a->du.tv.u.ds.mins +
+                                       1000000LL*3600*    a->du.tv.u.ds.hours +
+                                       1000000LL*3600*24* a->du.tv.u.ds.days);
+        bb = b->du.tv.sign*(long long)(
+                                       b->du.tv.u.ds.frac +
+                                       1000000LL*         b->du.tv.u.ds.sec +
+                                       1000000LL*60*      b->du.tv.u.ds.mins +
+                                       1000000LL*3600*    b->du.tv.u.ds.hours +
+                                       1000000LL*3600*24* b->du.tv.u.ds.days);
+
+        if( opcode == OP_Add ){
+          aa += bb;
+        }else if( opcode == OP_Subtract ){
+          aa -= bb;
+        }else return SQLITE_ERROR;
+
+        res->du.tv.type          = INTV_DSUS_TYPE;
+        res->du.tv.sign          = (aa<0)?-1:1;
+        aa = aa*res->du.tv.sign;
+        _setIntervalDSUS(&res->du.tv.u.ds,
+                         (((long long)aa)*res->du.tv.sign)/1000000,
+                         (int)((aa*res->du.tv.sign)%1000000));
+      }
+
+      break;
+    }
+    case INTV_DECIMAL_TYPE: {
+        logmsg(LOGMSG_FATAL, 
+              "BUG, wrong function called, use Decimalfy %s:%d\n",
+              __FILE__, __LINE__);
+        abort();
+    }
+  }
+  return 0;
+}
+
+/*
+**  Operate on interval and int, res = a opcode b;
+*/
+int sqlite3VdbeMemIntervalAndInt(
+  const Mem *a,
+  const Mem *b,
+  int opcode,
+  Mem * res
+                                 ){
+  bzero(res, sizeof(Mem));
+  res->flags |= MEM_Interval;
+  long long aa = 0, bb = 0;
+
+  switch( a->du.tv.type ){
+    case INTV_YM_TYPE: {
+      aa = a->du.tv.sign*(int)(a->du.tv.u.ym.months + a->du.tv.u.ym.years*12);
+      bb = b->u.i;
+
+      if( opcode == OP_Multiply)  {
+        aa *= bb;
+      } else if (opcode == OP_Divide) {
+        aa /= bb;
+      } else return SQLITE_ERROR;
+
+      res->du.tv.type          = INTV_YM_TYPE;
+      res->du.tv.sign          = (aa<0)?-1:1;
+      res->du.tv.u.ym.months   = aa*res->du.tv.sign;
+      res->du.tv.u.ym.years    = 0;
+
+      _normalizeIntervalYM(&res->du.tv.u.ym);
+      break;
+    }
+    case INTV_DS_TYPE: {
+      aa = a->du.tv.sign*(long long)(a->du.tv.u.ds.frac +
+                                     1000LL*           a->du.tv.u.ds.sec +
+                                     1000LL*60*        a->du.tv.u.ds.mins +
+                                     1000LL*3600*      a->du.tv.u.ds.hours +
+                                     1000LL*3600*24*   a->du.tv.u.ds.days);
+      bb = b->u.i;
+
+      if( opcode == OP_Multiply)  {
+        aa *= bb;
+      }else if (opcode == OP_Divide) {
+        aa /= bb;
+      }else return SQLITE_ERROR;
+
+      res->du.tv.type          = INTV_DS_TYPE;
+      res->du.tv.sign          = (aa<0)?-1:1;
+      aa = aa*res->du.tv.sign;
+      _setIntervalDS(&res->du.tv.u.ds,
+                     (((long long)aa)*res->du.tv.sign)/1000,
+                     (int)((aa*res->du.tv.sign)%1000));
+      break;
+    }
+    case INTV_DSUS_TYPE: {
+      aa = a->du.tv.sign*(long long)(
+                                     a->du.tv.u.ds.frac +
+                                     1000000LL*         a->du.tv.u.ds.sec +
+                                     1000000LL*60*      a->du.tv.u.ds.mins +
+                                     1000000LL*3600*    a->du.tv.u.ds.hours +
+                                     1000000LL*3600*24* a->du.tv.u.ds.days);
+      bb = b->u.i;
+
+      if( opcode == OP_Multiply)  {
+        aa *= bb;
+      }else if (opcode == OP_Divide) {
+        aa /= bb;
+      }else return SQLITE_ERROR;
+
+      res->du.tv.type          = INTV_DSUS_TYPE;
+      res->du.tv.sign          = (aa<0)?-1:1;
+      aa = aa*res->du.tv.sign;
+      _setIntervalDSUS(&res->du.tv.u.ds,
+                       (((long long)aa)*res->du.tv.sign)/1000000,
+                       (int)((aa*res->du.tv.sign)%1000000));
+      break;
+    }
+    case INTV_DECIMAL_TYPE: {
+        logmsg(LOGMSG_FATAL, 
+              "BUG, wrong function called, use Decimalfy %s:%d\n",
+              __FILE__, __LINE__);
+        abort();
+    }
+  }
+  return 0;
+}
+
+/*
+**  Operate on interval and int, res = a opcode b;
+*/
+int sqlite3VdbeMemIntAndInterval(
+  const Mem *a,
+  const Mem *b,
+  int opcode,
+  Mem * res
+){
+  bzero(res, sizeof(Mem));
+  res->flags |= MEM_Interval;
+  long long aa = 0, bb = 0;
+
+  switch( b->du.tv.type ){
+    case INTV_YM_TYPE: {
+      aa = a->u.i;
+      bb = b->du.tv.sign*(int)(b->du.tv.u.ym.months + b->du.tv.u.ym.years*12);
+
+      if( opcode == OP_Multiply ){
+        aa *= bb;
+      }else return -1;
+
+      res->du.tv.type          = INTV_YM_TYPE;
+      res->du.tv.sign          = (aa<0)?-1:1;
+      res->du.tv.u.ym.months   = aa*res->du.tv.sign;
+      res->du.tv.u.ym.years    = 0;
+
+      _normalizeIntervalYM(&res->du.tv.u.ym);
+      break;
+    }
+    case INTV_DS_TYPE: {
+      aa = a->u.i;
+      bb = b->du.tv.sign*(long long)(b->du.tv.u.ds.frac +
+                                     1000LL*           b->du.tv.u.ds.sec +
+                                     1000LL*60*        b->du.tv.u.ds.mins +
+                                     1000LL*3600*      b->du.tv.u.ds.hours +
+                                     1000LL*3600*24*   b->du.tv.u.ds.days);
+
+      if( opcode == OP_Multiply ){
+        aa *= bb;
+      }else return -1;
+
+      res->du.tv.type          = INTV_DS_TYPE;
+      res->du.tv.sign          = (aa<0)?-1:1;
+      aa = aa*res->du.tv.sign;
+      _setIntervalDS(&res->du.tv.u.ds,
+                     (((long long)aa)*res->du.tv.sign)/1000,
+                     (int)((aa*res->du.tv.sign)%1000));
+      break;
+    }
+    case INTV_DSUS_TYPE: {
+      aa = a->u.i;
+      bb = b->du.tv.sign*(long long)(b->du.tv.u.ds.frac +
+                                     1000000LL*         b->du.tv.u.ds.sec +
+                                     1000000LL*60*      b->du.tv.u.ds.mins +
+                                     1000000LL*3600*    b->du.tv.u.ds.hours +
+                                     1000000LL*3600*24* b->du.tv.u.ds.days);
+
+      if( opcode == OP_Multiply ){
+        aa *= bb;
+      }else return -1;
+
+      res->du.tv.type          = INTV_DSUS_TYPE;
+      res->du.tv.sign          = (aa<0)?-1:1;
+      aa = aa*res->du.tv.sign;
+      _setIntervalDSUS(&res->du.tv.u.ds,
+                       (((long long)aa)*res->du.tv.sign)/1000000,
+                       (int)((aa*res->du.tv.sign)%1000000));
+      break;
+    }
+    case INTV_DECIMAL_TYPE: {
+        logmsg(LOGMSG_FATAL, 
+              "BUG, wrong function called, use Decimalfy %s:%d\n",
+              __FILE__, __LINE__);
+        abort();
+    }
+  }
+  return 0;
+}
+
+/*
+** Implement substraction of two datetimes
+** The result is a day-second interval
+*/
+int sqlite3VdbeMemDatetimeAndDatetime(
+  const Mem *a,
+  const Mem *b,
+  int opcode,
+  Mem *res
+){
+  bzero(res, sizeof(Mem));
+  res->flags |= MEM_Interval;
+  if (opcode != OP_Subtract) return SQLITE_ERROR;
+  const dttz_t *d1 = &a->du.dt;
+  const dttz_t *d2 = &b->du.dt;
+  sub_dttz_dttz(d1, d2, &res->du.tv);
+  return 0;
+}
+
+static int _dttz_to_native_datetime(cdb2_client_datetime_t * cdt, const Mem *inp) {
+
+    char    tmp[11];
+    unsigned char buf[CLIENT_DATETIME_LEN];
+    unsigned char *p_buf=buf, *p_buf_end=(p_buf+CLIENT_DATETIME_LEN);
+    int     outnull = 0;
+    int     outdtsz  = 0; 
+    struct field_conv_opts_tz tzopts;
+
+    /* provide the timezone to the conversion routines */
+    bzero(&tzopts, sizeof(tzopts));
+    tzopts.flags |= 2 /*FLD_CONV_TZONE*/;
+    if( !inp->tz ) return SQLITE_ERROR;
+    strncpy0(tzopts.tzname, inp->tz, sizeof(tzopts.tzname));
+
+    /* ugly, arghh */
+    bzero(tmp, sizeof(tmp));
+
+    /* limit range to [-9999-01-01T235959.000 GMT, 9999-12-31T000000.000 GMT] */
+    if( !debug_switch_unlimited_datetime_range()
+        && (inp->du.dt.dttz_sec < -377705030401ll
+        || inp->du.dt.dttz_sec > 253402214400ll )){
+      return -1;
+    }
+
+    tmp[0] = 8;
+    *(long long*)&tmp[1] = flibc_htonll(inp->du.dt.dttz_sec);
+    tmp[1] ^= 0x80; 
+    *(unsigned short*)&tmp[9] = htons(
+        inp->du.dt.dttz_prec == DTTZ_PREC_MSEC ?
+        inp->du.dt.dttz_frac : inp->du.dt.dttz_frac / 1000);
+
+    if(SERVER_DATETIME_to_CLIENT_DATETIME(&tmp, sizeof(tmp),  
+              NULL, NULL,
+              buf, sizeof(buf), &outnull, &outdtsz,
+              (const struct field_conv_opts*)&tzopts, NULL))
+
+            return SQLITE_ERROR;
+
+    if(!(client_datetime_get(cdt,p_buf,p_buf_end)))
+    {
+        return SQLITE_ERROR;
+    }
+    
+    return 0;
+}
+
+static int _dttz_to_native_datetimeus(cdb2_client_datetimeus_t * cdt, const Mem *inp) {
+
+    char    tmp[13];
+    unsigned char buf[CLIENT_DATETIME_LEN];
+    unsigned char *p_buf=buf, *p_buf_end=(p_buf+CLIENT_DATETIME_LEN);
+    int     outnull = 0;
+    int     outdtsz  = 0; 
+    struct field_conv_opts_tz tzopts;
+
+    /* provide the timezone to the conversion routines */
+    bzero(&tzopts, sizeof(tzopts));
+    tzopts.flags |= 2 /*FLD_CONV_TZONE*/;
+    if(!inp->tz) return SQLITE_ERROR;
+    strncpy0(tzopts.tzname, inp->tz, sizeof(tzopts.tzname));
+
+    /* ugly, arghh */
+    bzero(tmp, sizeof(tmp));
+
+    /* limit range to [-9999-01-01T235959.000 GMT, 9999-12-31T000000.000 GMT] */
+    if (!debug_switch_unlimited_datetime_range() && 
+        (inp->du.dt.dttz_sec < -377705030401ll || 
+         inp->du.dt.dttz_sec > 253402214400ll))
+       return -1;
+
+    tmp[0] = 8;
+    *(long long*)&tmp[1] = flibc_htonll(inp->du.dt.dttz_sec);
+    tmp[1] ^= 0x80; 
+    *(unsigned short*)&tmp[9] = htonl(
+        inp->du.dt.dttz_prec == DTTZ_PREC_USEC ?
+        inp->du.dt.dttz_frac : inp->du.dt.dttz_frac * 1000);
+
+    if(SERVER_DATETIMEUS_to_CLIENT_DATETIMEUS(&tmp, sizeof(tmp),  
+              NULL, NULL,
+              buf, sizeof(buf), &outnull, &outdtsz,
+              (const struct field_conv_opts*)&tzopts, NULL))
+
+            return SQLITE_ERROR;
+
+    if(!(client_datetimeus_get(cdt,p_buf,p_buf_end)))
+    {
+        return SQLITE_ERROR;
+    }
+    
+    return 0;
+}
+
+static int _native_datetime_to_dttz(cdb2_client_datetime_t * cdt, Mem * res) {
+
+    char    tmp[11] = {0};
+    unsigned char buf[CLIENT_DATETIME_LEN+sizeof(char)];
+    unsigned char *p_buf=buf;
+    unsigned char *p_buf_end=buf+CLIENT_DATETIME_LEN;
+    int     outdtsz  = 0; 
+
+    *p_buf_end = 0;
+
+    if(!(client_datetime_put(cdt,p_buf, p_buf_end)))
+    {
+        return SQLITE_ERROR;
+    }
+
+    if(CLIENT_DATETIME_to_SERVER_DATETIME(buf, sizeof(buf), 0, NULL, NULL,
+                    tmp, sizeof(tmp), &outdtsz, NULL, NULL)) return SQLITE_ERROR;
+
+    /* convert server datetime to native datetime */
+    tmp[1] ^= 0x80;
+    res->du.dt.dttz_sec   = flibc_ntohll(*(unsigned long long*)&tmp[1]);
+    res->du.dt.dttz_frac  = ntohl(*(unsigned short*)&tmp[9]);
+    res->du.dt.dttz_prec  = DTTZ_PREC_MSEC;
+
+    return 0;
+}
+
+static int _native_datetimeus_to_dttz(cdb2_client_datetimeus_t * cdt, Mem * res) {
+
+    char    tmp[13] = {0};
+    unsigned char buf[CLIENT_DATETIME_LEN+sizeof(char)];
+    unsigned char *p_buf=buf, *p_buf_end=(p_buf+CLIENT_DATETIME_LEN);
+    int     outdtsz  = 0; 
+
+    *p_buf_end = 0;
+
+    if(!(client_datetimeus_put(cdt,p_buf, p_buf_end)))
+    {
+        return SQLITE_ERROR;
+    }
+
+    if(CLIENT_DATETIMEUS_to_SERVER_DATETIMEUS(buf, sizeof(buf), 0, NULL, NULL,
+                    tmp, sizeof(tmp), &outdtsz, NULL, NULL)) return SQLITE_ERROR;
+
+    /* convert server datetime to native datetime */
+    tmp[1] ^= 0x80;
+    res->du.dt.dttz_sec   = flibc_ntohll(*(unsigned long long*)&tmp[1]);
+    res->du.dt.dttz_frac  = ntohl(*(unsigned int*)&tmp[9]);
+    res->du.dt.dttz_prec  = DTTZ_PREC_USEC;
+
+    return 0;
+}
+
+/* "b" better be normalized */
+int sqlite3VdbeMemDatetimeAndInterval(
+  const Mem *a,
+  const Mem *b,
+  int opcode,
+  Mem * res
+){
+
+  bzero(res, sizeof(Mem));
+  res->flags |= MEM_Datetime;
+  res->tz = a->tz; /* propagate tzname */
+
+  /* operations */
+  if( b->du.tv.type == INTV_YM_TYPE ){
+    if( a->du.dt.dttz_prec == DTTZ_PREC_MSEC ){
+      cdb2_client_datetime_t   cdt;
+
+      /* convert server_datetime_t to cdb2_client_datetime_t */
+      if( _dttz_to_native_datetime(&cdt, a) ) return SQLITE_ERROR;
+
+      if( (opcode == OP_Add && b->du.tv.sign >0) || 
+          (opcode == OP_Subtract && b->du.tv.sign <0) ){
+        cdt.tm.tm_year  += b->du.tv.u.ym.years;
+        cdt.tm.tm_mon += b->du.tv.u.ym.months;
+      }else if( (opcode == OP_Add && b->du.tv.sign<0) || 
+                 (opcode == OP_Subtract && b->du.tv.sign >0) ){
+        cdt.tm.tm_year  -= b->du.tv.u.ym.years;
+        cdt.tm.tm_mon -= b->du.tv.u.ym.months;
+      }else return SQLITE_ERROR;
+
+      /* convert cdb2_client_datetime_t to server_datetime_t */
+      if( _native_datetime_to_dttz(&cdt, res) ) return SQLITE_ERROR;
+    }else{
+      cdb2_client_datetimeus_t   cdt;
+
+      /* convert server_datetimeus_t to cdb2_client_datetimeus_t */
+      if( _dttz_to_native_datetimeus(&cdt, a) ) return SQLITE_ERROR;
+
+      if( (opcode == OP_Add && b->du.tv.sign >0) || 
+          (opcode == OP_Subtract && b->du.tv.sign <0) ){
+        cdt.tm.tm_year  += b->du.tv.u.ym.years;
+        cdt.tm.tm_mon += b->du.tv.u.ym.months;
+      }else if( (opcode == OP_Add && b->du.tv.sign<0) || 
+                 (opcode == OP_Subtract && b->du.tv.sign >0) ){
+        cdt.tm.tm_year  -= b->du.tv.u.ym.years;
+        cdt.tm.tm_mon -= b->du.tv.u.ym.months;
+      }else return SQLITE_ERROR;
+
+      /* convert cdb2_client_datetimeus_t to server_datetimeus_t */
+      if( _native_datetimeus_to_dttz(&cdt, res) ) return SQLITE_ERROR;
+    }
+    res->du.dt.dttz_conv = (a->du.dt.dttz_conv == DTTZ_CONV_NOW);
+  }else if( b->du.tv.type == INTV_DS_TYPE || b->du.tv.type == INTV_DSUS_TYPE ){
+    if( opcode == OP_Add ){
+      add_dttz_intvds(&a->du.dt, &b->du.tv, &res->du.dt);
+    }else if( opcode == OP_Subtract ){
+      sub_dttz_intvds(&a->du.dt, &b->du.tv, &res->du.dt);
+    }else{
+      return SQLITE_ERROR;
+    }
+  }else{
+    logmsg(LOGMSG_FATAL, 
+          "BUG, wrong function called, use Decimalfy %s:%d\n",
+          __FILE__, __LINE__);
+    abort();
+  }
+
+  return 0;
+}
+
+int sqlite3VdbeMemIntervalAndDatetime(
+  const Mem *a,
+  const Mem *b,
+  int opcode,
+  Mem * res
+){
+
+  bzero(res, sizeof(Mem));
+  res->flags |= MEM_Datetime;
+  res->tz = b->tz; /* propagate tzname */
+
+  if( a->du.tv.type == INTV_YM_TYPE ){
+    if ( b->du.dt.dttz_prec == DTTZ_PREC_MSEC ){
+      cdb2_client_datetime_t   cdt;
+
+      /* convert server_datetime_t to cdb2_client_datetime_t */
+      if( _dttz_to_native_datetime(&cdt, b) ) return SQLITE_ERROR;
+
+      /* operations */
+      if( opcode == OP_Add ){
+        cdt.tm.tm_year  += a->du.tv.u.ym.years;
+        cdt.tm.tm_mon += a->du.tv.u.ym.months;
+      }else return SQLITE_ERROR;
+
+      /* convert cdb2_client_datetime_t to server_datetime_t */
+      if(_native_datetime_to_dttz(&cdt, res)) return SQLITE_ERROR;
+    }else{
+      cdb2_client_datetimeus_t   cdt;
+
+      /* convert server_datetimeus_t to cdb2_client_datetimeus_t */
+      if( _dttz_to_native_datetimeus(&cdt, b) ) return SQLITE_ERROR;
+
+      /* operations */
+      if( opcode == OP_Add ){
+        cdt.tm.tm_year  += a->du.tv.u.ym.years;
+        cdt.tm.tm_mon += a->du.tv.u.ym.months;
+      }else return SQLITE_ERROR;
+
+      /* convert cdb2_client_datetimeus_t to server_datetimeus_t */
+      if( _native_datetimeus_to_dttz(&cdt, res) ) return SQLITE_ERROR;
+    }
+    res->du.dt.dttz_conv = (b->du.dt.dttz_conv == DTTZ_CONV_NOW);
+  }else if( a->du.tv.type == INTV_DS_TYPE || a->du.tv.type == INTV_DSUS_TYPE ){
+    if( opcode == OP_Add ){
+      add_dttz_intvds(&b->du.dt, &a->du.tv, &res->du.dt);
+    }else{
+      return SQLITE_ERROR;
+    }
+  }else{
+    logmsg(LOGMSG_DEBUG, "%s:%d: No arithmetics between decimal and datetime\n",
+          __FILE__, __LINE__);
+    return SQLITE_ERROR;
+  }
+
+  return 0;
+}
+
+/*
+   Mem is an decimal type;
+   Mem is anything;
+ */
+int sqliteVdbeMemDecimalBasicArithmetics(
+  Mem *a,
+  Mem *b,
+  int opcode,
+  Mem * res,
+  int flipped
+){
+  decContext  ctx;
+  void        *ret = NULL;
+  int         rc = 0;
+  Mem         bcopy;
+
+  bzero(&bcopy, sizeof(Mem));
+  bzero(res, sizeof(Mem));
+
+  switch( b->flags & MEM_TypeMask ){
+    case MEM_Null: {
+      sqlite3VdbeMemSetNull(res);
+      break;
+    }
+    case MEM_Int:
+    case MEM_Str: {
+      sqlite3VdbeMemCopy(&bcopy, b);
+      b = &bcopy;
+      
+      rc = sqlite3VdbeMemDecimalfy(b);
+      if( rc ){
+        goto done;
+      }
+
+      /* fall through */
+    }
+    case MEM_Interval: {
+      Mem *tmp;
+
+      if( flipped ){
+        tmp=b;
+        b=a;
+        a=tmp;
+      }
+
+      dec_ctx_init( &ctx, DEC_INIT_DECQUAD, gbl_decimal_rounding);
+      
+      switch( opcode ){
+        case OP_Add: {
+          ret = decQuadAdd((decQuad*)&res->du.tv.u.dec, 
+                           (const decQuad*)&a->du.tv.u.dec, 
+                           (const decQuad*)&b->du.tv.u.dec, 
+                           &ctx);
+          break;
+        }
+        case OP_Subtract: {
+          ret = decQuadSubtract((decQuad*)&res->du.tv.u.dec, 
+                                (const decQuad*)&a->du.tv.u.dec, 
+                                (const decQuad*)&b->du.tv.u.dec, 
+                                &ctx);
+          break;
+        }
+        case OP_Divide: {
+          ret = decQuadDivide((decQuad*)&res->du.tv.u.dec, 
+                              (const decQuad*)&a->du.tv.u.dec, 
+                              (const decQuad*)&b->du.tv.u.dec, 
+                              &ctx);
+          break;
+        }
+        case OP_Multiply: {
+          ret = decQuadMultiply((decQuad*)&res->du.tv.u.dec, 
+                                (const decQuad*)&a->du.tv.u.dec, 
+                                (const decQuad*)&b->du.tv.u.dec, 
+                                &ctx);
+          break;
+        }
+        case OP_Remainder: {
+          ret = decQuadRemainder((decQuad*)&res->du.tv.u.dec, 
+                                 (const decQuad*)&a->du.tv.u.dec, 
+                                 (const decQuad*)&b->du.tv.u.dec, 
+                                 &ctx);
+          break;
+        }
+        default: {
+          rc = SQLITE_ERROR;
+          goto done;
+        }
+      }
+
+      if( !ret ){
+        rc = SQLITE_ERROR;
+        goto done;
+      }else{
+        if( dfp_conv_check_status( &ctx, "arithmetics", "quad") ){
+          rc = SQLITE_ERROR;
+          goto done;
+        }
+      }
+
+      res->flags |= MEM_Interval;
+      res->du.tv.type = INTV_DECIMAL_TYPE;
+      res->du.tv.sign = 0;
+      break;
+    }
+
+    default: {
+      rc = SQLITE_ERROR;
+      goto done;
+    }
+  }
+
+done:
+   sqlite3VdbeMemRelease(&bcopy);
+   return rc;
+}
+
+/* Compare 2 unpacked rows */
+int sqlite3RecordCompareExprList(UnpackedRecord *rec, Mem *mems)
+{
+  int i;
+  int rc = 0;
+  
+  for(i=0;i<rec->nField;i++)
+  {
+    rc=sqlite3MemCompare(&rec->aMem[i], &mems[i], NULL);
+    if (rc)
+        return rc;
+  }
+  return rc;
+}
+
+/* Convert a SIMPLE expression to a Mem array */
+int sqlite3ExprList2MemArray(ExprList *list, Mem *mems)
+{
+  int i;
+  for(i=0;i<list->nExpr;i++)
+  {
+    switch(list->a[i].pExpr->op)
+    {
+      case TK_INTEGER:
+        mems[i].flags = MEM_Int;
+        mems[i].u.i = list->a[i].pExpr->u.iValue;
+        break; 
+      default:
+        return -1;  
+    }
+  }
+  return 0;
+}
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
