--- sqlite-3.28.0/src/vtab.c	2026-01-09 16:11:01
+++ src/vtab.c	2025-09-24 15:21:58
@@ -335,6 +335,10 @@
   Table *pTable;        /* The new virtual table */
   sqlite3 *db;          /* Database connection */
 
+  if(comdb2IsDryrun(pParse)){
+      sqlite3ErrorMsg(pParse,"DRYRUN not supported for this operation");
+      return;
+  }
   sqlite3StartTable(pParse, pName1, pName2, 0, 0, 1, ifNotExists);
   pTable = pParse->pNewTable;
   if( pTable==0 ) return;
