--- sqlite-3.28.0/src/whereexpr.c	2026-01-09 16:11:01
+++ src/whereexpr.c	2025-09-24 15:21:58
@@ -213,6 +213,23 @@
 #endif
   pList = pExpr->x.pList;
   pLeft = pList->a[1].pExpr;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* NC: Restore the old logic. This fixes a regression caused by LIKE
+   * optimization (implemented here) particularly for BLOB types. This
+   * optimization essentially converts LIKE expression into inequality
+   * operations, triggering index lookups, which would fail for BLOB
+   * types as Comdb2 does not provide string->blob converter (db/types.c).
+   * e.g. select * from tab where blob_field like 'a%'
+   */
+  if( pLeft->op!=TK_COLUMN
+   || sqlite3ExprAffinity(pLeft)!=SQLITE_AFF_TEXT
+   || IsVirtual(pLeft->y.pTab)  /* Value might be numeric */
+  ){
+    /* IMP: R-02065-49465 The left-hand side of the LIKE or GLOB operator must
+    ** be the name of an indexed column with TEXT affinity. */
+    return 0;
+  }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
   pRight = sqlite3ExprSkipCollate(pList->a[0].pExpr);
   op = pRight->op;
@@ -972,6 +989,7 @@
     if( pIdx->aColExpr==0 ) continue;
     for(i=0; i<pIdx->nKeyCol; i++){
       if( pIdx->aiColumn[i]!=XN_EXPR ) continue;
+      assert( pIdx->bHasExpr );
       if( sqlite3ExprCompareSkip(pExpr, pIdx->aColExpr->a[i].pExpr, iCur)==0 ){
         aiCurCol[0] = iCur;
         aiCurCol[1] = XN_EXPR;
@@ -1140,6 +1158,12 @@
       testcase( (prereqLeft | extraRight) != prereqLeft );
       pNew->prereqRight = prereqLeft | extraRight;
       pNew->prereqAll = prereqAll;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      /* Lingzhi: upgrade to ISNULL?? */
+      if ( pLeft->op==TK_NULL && op==TK_IS )
+        pNew->eOperator = WO_ISNULL & opMask;
+      else
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       pNew->eOperator = (operatorMask(pDup->op) + eExtraOp) & opMask;
     }
   }
@@ -1599,6 +1623,9 @@
     pRhs = sqlite3PExpr(pParse, TK_UPLUS, 
         sqlite3ExprDup(pParse->db, pArgs->a[j].pExpr, 0), 0);
     pTerm = sqlite3PExpr(pParse, TK_EQ, pColRef, pRhs);
+    if( pItem->fg.jointype & JT_LEFT ){
+      sqlite3SetJoinExpr(pTerm, pItem->iCursor);
+    }
     whereClauseInsert(pWC, pTerm, TERM_DYNAMIC);
   }
 }
