--- sqlite-3.28.0/src/wherecode.c	2026-01-09 16:11:01
+++ src/wherecode.c	2025-09-24 15:21:58
@@ -568,7 +568,16 @@
     if( pLevel->u.in.nIn==0 ){
       pLevel->addrNxt = sqlite3VdbeMakeLabel(pParse);
     }
+    if( iEq>0 ){
+      pLoop->wsFlags |= WHERE_IN_EARLYOUT;
+    }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    //if( iEq>0 && (pLoop->wsFlags && WHERE_IN_SEEKSCAN)==0 ){
+    //  pLoop->wsFlags |= WHERE_IN_EARLYOUT;
+    //}
+#endif /* SQLITE_BUILDING_FOR_COMDB2 */
+
     i = pLevel->u.in.nIn;
     pLevel->u.in.nIn += nEq;
     pLevel->u.in.aInLoop =
@@ -591,10 +600,9 @@
           if( i==iEq ){
             pIn->iCur = iTab;
             pIn->eEndLoopOp = bRev ? OP_Prev : OP_Next;
-            if( iEq>0 && (pLoop->wsFlags & WHERE_VIRTUALTABLE)==0 ){
+            if( iEq>0 ){
               pIn->iBase = iReg - i;
               pIn->nPrefix = i;
-              pLoop->wsFlags |= WHERE_IN_EARLYOUT;
             }else{
               pIn->nPrefix = 0;
             }
@@ -604,6 +612,9 @@
           pIn++;
         }
       }
+      if( iEq>0 ){
+        sqlite3VdbeAddOp3(v, OP_SeekHit, pLevel->iIdxCur, 0, iEq);
+      }
     }else{
       pLevel->u.in.nIn = 0;
     }
@@ -703,6 +714,9 @@
     j = sqlite3VdbeAddOp0(v, OP_Goto);
     pLevel->addrSkip = sqlite3VdbeAddOp4Int(v, (bRev?OP_SeekLT:OP_SeekGT),
                             iIdxCur, 0, regBase, nSkip);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    sqlite3VdbeChangeP5(v, OPFLAG_SKIPSCAN);
+#endif
     VdbeCoverageIf(v, bRev==0);
     VdbeCoverageIf(v, bRev!=0);
     sqlite3VdbeJumpHere(v, j);
@@ -761,7 +775,7 @@
   return regBase;
 }
 
-#ifndef SQLITE_LIKE_DOESNT_MATCH_BLOBS
+#if !defined(SQLITE_BUILDING_FOR_COMDB2) && !defined(SQLITE_LIKE_DOESNT_MATCH_BLOBS)
 /*
 ** If the most recently coded instruction is a constant range constraint
 ** (a string literal) that originated from the LIKE optimization, then 
@@ -795,9 +809,9 @@
     pOp->p5 = (u8)(pLevel->iLikeRepCntr&1);    /* ASC or DESC */
   }
 }
-#else
+#else /* !defined(SQLITE_BUILDING_FOR_COMDB2) && !defined(SQLITE_LIKE_DOESNT_MATCH_BLOBS) */
 # define whereLikeOptimizationStringFixup(A,B,C)
-#endif
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) && !defined(SQLITE_LIKE_DOESNT_MATCH_BLOBS) */
 
 #ifdef SQLITE_ENABLE_CURSOR_HINTS
 /*
@@ -890,7 +904,11 @@
       pExpr->iTable = reg;
     }else if( pHint->pIdx!=0 ){
       pExpr->iTable = pHint->iIdxCur;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      /* NOTE: this breaks remote sql; discussion in progress to push this upstream */
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       pExpr->iColumn = sqlite3ColumnOfIndex(pHint->pIdx, pExpr->iColumn);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       assert( pExpr->iColumn>=0 );
     }
   }else if( pExpr->op==TK_AGG_FUNCTION ){
@@ -914,6 +932,9 @@
   WhereInfo *pWInfo,    /* The where clause */
   WhereLevel *pLevel,   /* Which loop to provide hints for */
   WhereTerm *pEndRange  /* Hint this end-of-scan boundary term if not NULL */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  ,int iLevel            /* Needed to skip non-remote cursors from hinting */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 ){
   Parse *pParse = pWInfo->pParse;
   sqlite3 *db = pParse->db;
@@ -926,8 +947,41 @@
   int i, j;
   struct CCurHint sHint;
   Walker sWalker;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /*
+    This is from conversation with sqlite. I need this mask
+    "....
+    I did not port the last sqlite patch related to cursor hints to comdb2.  
+    This last patch relates to how = predicates are coded. Adding that patch 
+    would require more restructuring in our code than I felt comfortable doing
+    at the time (or creating redundant predicates in hint expressions, which I
+    disliked as well).
 
+    Instead, I did two hacky things that achieve a similar effect and kept the
+    code working without the last patch,and forgot all about it:
+
+    1) commented out TERM_CODED term
+    2) put back logic to filter expressions that don't apply to hinted cursor
+    (which would skip coded loops, but regenerate some = predicates comdb2 
+    relied on --therefore the comment--, and missing after sqlite changes)
+    ..."
+   */
+  Bitmask msk;
+  WhereLoop *pWLoop;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   if( OptimizationDisabled(db, SQLITE_CursorHints) ) return;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  /* Really need to run this only for remote cursors */
+  /* hack, at this point only remcurs have it */
+  if( pWInfo->pTabList->a[iLevel].zDatabase==NULL )
+    return;
+
+  /* Need this Mask since the code lower ignores TERM_CODED !!!!*/
+  iCur = pLevel->iIdxCur;
+  msk = sqlite3WhereGetMask(&pWInfo->sMaskSet,
+                            pWInfo->pTabList->a[pLevel->iFrom].iCursor);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+
   iCur = pLevel->iTabCur;
   assert( iCur==pWInfo->pTabList->a[pLevel->iFrom].iCursor );
   sHint.iTabCur = iCur;
@@ -937,9 +991,23 @@
   sWalker.pParse = pParse;
   sWalker.u.pCCurHint = &sHint;
   pWC = &pWInfo->sWC;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+  pWLoop = pLevel->pWLoop;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   for(i=0; i<pWC->nTerm; i++){
     pTerm = &pWC->a[i];
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    /* msk makes sure I consider only terms that
+       apply to the cursor in question */
+    if( !(pTerm->prereqAll & msk)) continue;
+    /* I need TERM_CODED commented out. Comes hand in hand with  msk,
+       i.e. msk avoid coded predicates for unrelated terms, while 
+       TERM_CODED commented allows me still encode equality operations
+       properly*/
+    if( pTerm->wtFlags & (TERM_VIRTUAL/*|TERM_CODED*/) ) continue;
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( pTerm->wtFlags & (TERM_VIRTUAL|TERM_CODED) ) continue;
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( pTerm->prereqAll & pLevel->notReady ) continue;
 
     /* Any terms specified as part of the ON(...) clause for any LEFT 
@@ -978,13 +1046,22 @@
       if( ExprHasProperty(pTerm->pExpr, EP_FromJoin) ) continue;
     }
 
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
     /* All terms in pWLoop->aLTerm[] except pEndRange are used to initialize
+    ** the cursor.  No need to hint initialization terms. */
+    if( pTerm!=pEndRange ){
+      for(j=0; j<pWLoop->nLTerm && pWLoop->aLTerm[j]!=pTerm; j++){}
+      if( j<pWLoop->nLTerm ) continue;
+    }
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+    /* All terms in pWLoop->aLTerm[] except pEndRange are used to initialize
     ** the cursor.  These terms are not needed as hints for a pure range
     ** scan (that has no == terms) so omit them. */
     if( pLoop->u.btree.nEq==0 && pTerm!=pEndRange ){
       for(j=0; j<pLoop->nLTerm && pLoop->aLTerm[j]!=pTerm; j++){}
       if( j<pLoop->nLTerm ) continue;
     }
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 
     /* No subqueries or non-deterministic functions allowed */
     if( sqlite3ExprContainsSubquery(pTerm->pExpr) ) continue;
@@ -1010,7 +1087,11 @@
   }
 }
 #else
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+# define codeCursorHint(A,B,C,D,E)  /* No-op */
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 # define codeCursorHint(A,B,C,D)  /* No-op */
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
 #endif /* SQLITE_ENABLE_CURSOR_HINTS */
 
 /*
@@ -1042,10 +1123,15 @@
 
   assert( iIdxCur>0 );
   assert( pIdx->aiColumn[pIdx->nColumn-1]==-1 );
-  
+
+  pWInfo->bDeferredSeek = 1;
   sqlite3VdbeAddOp3(v, OP_DeferredSeek, iIdxCur, 0, iCur);
   if( (pWInfo->wctrlFlags & WHERE_OR_SUBCLAUSE)
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+   && DbMaskAllZero(sqlite3ParseToplevel(pParse)->writeMask, 0)
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
    && DbMaskAllZero(sqlite3ParseToplevel(pParse)->writeMask)
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
   ){
     int i;
     Table *pTab = pIdx->pTable;
@@ -1187,6 +1273,10 @@
   }
 }
 
+#ifdef SQLITE_BUILDING_FOR_COMDB2
+int gbl_seekscan_maxsteps = -1;
+#endif
+
 /*
 ** Generate code for the start of the iLevel-th loop in the WHERE clause
 ** implementation described by pWInfo.
@@ -1369,6 +1459,10 @@
     if( iRowidReg!=iReleaseReg ) sqlite3ReleaseTempReg(pParse, iReleaseReg);
     addrNxt = pLevel->addrNxt;
     sqlite3VdbeAddOp3(v, OP_SeekRowid, iCur, addrNxt, iRowidReg);
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    if( (pLoop->wsFlags & WHERE_ONEROW) == 0 )
+        sqlite3VdbeChangeP5(v, 1);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     VdbeCoverage(v);
     pLevel->op = OP_Noop;
     if( (pTerm->prereqAll & pLevel->notReady)==0 ){
@@ -1394,7 +1488,11 @@
       pStart = pEnd;
       pEnd = pTerm;
     }
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    codeCursorHint(pTabItem, pWInfo, pLevel, pEnd, iLevel);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     codeCursorHint(pTabItem, pWInfo, pLevel, pEnd);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     if( pStart ){
       Expr *pX;             /* The expression that defines the start bound */
       int r1, rTemp;        /* Registers for holding the start boundary */
@@ -1548,6 +1646,7 @@
     u8 bSeekPastNull = 0;        /* True to seek past initial nulls */
     u8 bStopAtNull = 0;          /* Add condition to terminate at NULLs */
     int omitTable;               /* True if we use the index only */
+    int addrSeekScan = 0;        /* Opcode of the OP_SeekScan, if any */
 
 
     pIdx = pLoop->u.btree.pIndex;
@@ -1588,7 +1687,7 @@
     if( pLoop->wsFlags & WHERE_TOP_LIMIT ){
       pRangeEnd = pLoop->aLTerm[j++];
       nExtraReg = MAX(nExtraReg, pLoop->u.btree.nTop);
-#ifndef SQLITE_LIKE_DOESNT_MATCH_BLOBS
+#if !defined(SQLITE_BUILDING_FOR_COMDB2) && !defined(SQLITE_LIKE_DOESNT_MATCH_BLOBS)
       if( (pRangeEnd->wtFlags & TERM_LIKEOPT)!=0 ){
         assert( pRangeStart!=0 );                     /* LIKE opt constraints */
         assert( pRangeStart->wtFlags & TERM_LIKEOPT );   /* occur in pairs */
@@ -1604,7 +1703,7 @@
         pLevel->iLikeRepCntr <<=1;
         pLevel->iLikeRepCntr |= bRev ^ (pIdx->aSortOrder[nEq]==SQLITE_SO_DESC);
       }
-#endif
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) && !defined(SQLITE_LIKE_DOESNT_MATCH_BLOBS) */
       if( pRangeStart==0 ){
         j = pIdx->aiColumn[nEq];
         if( (j>=0 && pIdx->pTable->aCol[j].notNull==0) || j==XN_EXPR ){
@@ -1624,13 +1723,21 @@
       SWAP(WhereTerm *, pRangeEnd, pRangeStart);
       SWAP(u8, bSeekPastNull, bStopAtNull);
       SWAP(u8, nBtm, nTop);
+    }
+
+    if( iLevel>0 && (pLoop->wsFlags & WHERE_IN_SEEKSCAN)!=0 ){
+      /* In case OP_SeekScan is used, ensure that the index cursor does not
+      ** point to a valid row for the first iteration of this loop. */
+      sqlite3VdbeAddOp1(v, OP_NullRow, iIdxCur);
     }
 
     /* Generate code to evaluate all constraint terms using == or IN
     ** and store the values of those terms in an array of registers
     ** starting at regBase.
     */
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
     codeCursorHint(pTabItem, pWInfo, pLevel, pRangeEnd);
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
     regBase = codeAllEqualityTerms(pParse,pLevel,bRev,nExtraReg,&zStartAff);
     assert( zStartAff==0 || sqlite3Strlen30(zStartAff)>=nEq );
     if( zStartAff && nTop ){
@@ -1647,6 +1754,9 @@
     start_constraints = pRangeStart || nEq>0;
 
     /* Seek the index cursor to the start of the range. */
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+    codeCursorHint(pTabItem, pWInfo, pLevel, pRangeEnd, iLevel);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     nConstraint = nEq;
     if( pRangeStart ){
       Expr *pRight = pRangeStart->pExpr->pRight;
@@ -1681,11 +1791,53 @@
       ** above has already left the cursor sitting on the correct row,
       ** so no further seeking is needed */
     }else{
-      if( pLoop->wsFlags & WHERE_IN_EARLYOUT ){
-        sqlite3VdbeAddOp1(v, OP_SeekHit, iIdxCur);
-      }
       op = aStartOp[(start_constraints<<2) + (startEq<<1) + bRev];
       assert( op!=0 );
+      if( (pLoop->wsFlags & WHERE_IN_SEEKSCAN)!=0 && op==OP_SeekGE ) {
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+        // NC: Patch that introduced Bignull has not been backported yet
+        //assert( regBignull==0 );
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+        assert( regBignull==0 );
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
+        /* TUNING:  The OP_SeekScan opcode seeks to reduce the number
+        ** of expensive seek operations by replacing a single seek with
+        ** 1 or more step operations.  The question is, how many steps
+        ** should we try before giving up and going with a seek.  The cost
+        ** of a seek is proportional to the logarithm of the of the number
+        ** of entries in the tree, so basing the number of steps to try
+        ** on the estimated number of rows in the btree seems like a good
+        ** guess. */
+#if defined SQLITE_BUILDING_FOR_COMDB2
+        /* gbl_disable_seekscan_optimization turns off seekscans completely.  This
+        ** is a bit more lenient - we can keep them on, but get rid of, or adjust,
+        ** the maximum number of steps that we try to do on the index before doing
+        ** the next find.  This is a three-way switch. The default setting of -1
+        ** lets SQLite keep its setting that estimates the max number of steps by
+        ** max number of rows it expects to traverse.  0 disables the seekscan
+        ** optimization (OP_SeekScan), but still does a seekscan, doing a find for
+        ** each next record.  A positive value caps the scan to that many
+        ** operations. */
+        if (gbl_seekscan_maxsteps == 0) {
+          addrSeekScan = 0;
+          sqlite3VdbeAddOp0(v, OP_Noop);
+        } else if (gbl_seekscan_maxsteps > 0) {
+          addrSeekScan = sqlite3VdbeAddOp1(v, OP_SeekScan, gbl_seekscan_maxsteps);
+        } else {
+#endif
+          addrSeekScan = sqlite3VdbeAddOp1(v, OP_SeekScan,
+                                           (pIdx->aiRowLogEst[0] + 9) / 10);
+#if defined SQLITE_BUILDING_FOR_COMDB2
+        }
+
+        if (pRangeStart || pRangeEnd) {
+            sqlite3VdbeChangeP5(v, 1);
+            sqlite3VdbeChangeP2(v, addrSeekScan, sqlite3VdbeCurrentAddr(v)+1);
+            addrSeekScan = 0;
+        }
+#endif
+        VdbeCoverage(v);
+      }
       sqlite3VdbeAddOp4Int(v, op, iIdxCur, addrNxt, regBase, nConstraint);
       VdbeCoverage(v);
       VdbeCoverageIf(v, op==OP_Rewind);  testcase( op==OP_Rewind );
@@ -1700,8 +1852,10 @@
     ** range (if any).
     */
     nConstraint = nEq;
+    assert( pLevel->p2==0 );
     if( pRangeEnd ){
       Expr *pRight = pRangeEnd->pExpr->pRight;
+      assert( addrSeekScan == 0 );
       codeExprOrVector(pParse, pRight, regBase+nEq, nTop);
       whereLikeOptimizationStringFixup(v, pLevel, pRangeEnd);
       if( (pRangeEnd->wtFlags & TERM_VNULL)==0
@@ -1733,7 +1887,7 @@
     sqlite3DbFree(db, zEndAff);
 
     /* Top of the loop body */
-    pLevel->p2 = sqlite3VdbeCurrentAddr(v);
+    if( pLevel->p2==0 ) pLevel->p2 = sqlite3VdbeCurrentAddr(v);
 
     /* Check if the index cursor is past the end of the range. */
     if( nConstraint ){
@@ -1743,10 +1897,11 @@
       testcase( op==OP_IdxGE );  VdbeCoverageIf(v, op==OP_IdxGE );
       testcase( op==OP_IdxLT );  VdbeCoverageIf(v, op==OP_IdxLT );
       testcase( op==OP_IdxLE );  VdbeCoverageIf(v, op==OP_IdxLE );
+      if( addrSeekScan ) sqlite3VdbeJumpHere(v, addrSeekScan);
     }
 
     if( pLoop->wsFlags & WHERE_IN_EARLYOUT ){
-      sqlite3VdbeAddOp2(v, OP_SeekHit, iIdxCur, 1);
+      sqlite3VdbeAddOp3(v, OP_SeekHit, iIdxCur, nEq, nEq);
     }
 
     /* Seek the table cursor, if required */
@@ -1755,17 +1910,7 @@
     if( omitTable ){
       /* pIdx is a covering index.  No need to access the main table. */
     }else if( HasRowid(pIdx->pTable) ){
-      if( (pWInfo->wctrlFlags & WHERE_SEEK_TABLE) || (
-          (pWInfo->wctrlFlags & WHERE_SEEK_UNIQ_TABLE) 
-       && (pWInfo->eOnePass==ONEPASS_SINGLE)
-      )){
-        iRowidReg = ++pParse->nMem;
-        sqlite3VdbeAddOp2(v, OP_IdxRowid, iIdxCur, iRowidReg);
-        sqlite3VdbeAddOp3(v, OP_NotExists, iCur, 0, iRowidReg);
-        VdbeCoverage(v);
-      }else{
-        codeDeferredSeek(pWInfo, pIdx, iCur, iIdxCur);
-      }
+      codeDeferredSeek(pWInfo, pIdx, iCur, iIdxCur);
     }else if( iCur!=iIdxCur ){
       Index *pPk = sqlite3PrimaryKeyIndex(pIdx->pTable);
       iRowidReg = sqlite3GetTempRange(pParse, pPk->nKeyCol);
@@ -1879,7 +2024,6 @@
     int iRetInit;                             /* Address of regReturn init */
     int untestedTerms = 0;             /* Some terms not completely tested */
     int ii;                            /* Loop counter */
-    u16 wctrlFlags;                    /* Flags for sub-WHERE clause */
     Expr *pAndExpr = 0;                /* An ".. AND (...)" expression */
     Table *pTab = pTabItem->pTab;
 
@@ -1980,7 +2124,6 @@
     ** eliminating duplicates from other WHERE clauses, the action for each
     ** sub-WHERE clause is to to invoke the main loop body as a subroutine.
     */
-    wctrlFlags =  WHERE_OR_SUBCLAUSE | (pWInfo->wctrlFlags & WHERE_SEEK_TABLE);
     ExplainQueryPlan((pParse, 1, "MULTI-INDEX OR"));
     for(ii=0; ii<pOrWc->nTerm; ii++){
       WhereTerm *pOrTerm = &pOrWc->a[ii];
@@ -1999,7 +2142,7 @@
         ExplainQueryPlan((pParse, 1, "INDEX %d", ii+1));
         WHERETRACE(0xffff, ("Subplan for OR-clause:\n"));
         pSubWInfo = sqlite3WhereBegin(pParse, pOrTab, pOrExpr, 0, 0,
-                                      wctrlFlags, iCovCur);
+                                      WHERE_OR_SUBCLAUSE, iCovCur);
         assert( pSubWInfo || pParse->nErr || db->mallocFailed );
         if( pSubWInfo ){
           WhereLoop *pSubLoop;
@@ -2097,6 +2240,9 @@
           }else{
             pCov = 0;
           }
+          if( sqlite3WhereUsesDeferredSeek(pSubWInfo) ){
+            pWInfo->bDeferredSeek = 1;
+          }
 
           /* Finish the loop through table entries that match term pOrTerm. */
           sqlite3WhereEnd(pSubWInfo);
@@ -2132,7 +2278,11 @@
       ** a pseudo-cursor.  No need to Rewind or Next such cursors. */
       pLevel->op = OP_Noop;
     }else{
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      codeCursorHint(pTabItem, pWInfo, pLevel, 0, iLevel);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       codeCursorHint(pTabItem, pWInfo, pLevel, 0);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       pLevel->op = aStep[bRev];
       pLevel->p1 = iCur;
       pLevel->p2 = 1 + sqlite3VdbeAddOp2(v, aStart[bRev], iCur, addrHalt);
@@ -2196,16 +2346,16 @@
         ** can skip the call to the like(A,B) function.  But this only works
         ** for strings.  So do not skip the call to the function on the pass
         ** that compares BLOBs. */
-#ifdef SQLITE_LIKE_DOESNT_MATCH_BLOBS
+#if defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_LIKE_DOESNT_MATCH_BLOBS)
         continue;
-#else
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_LIKE_DOESNT_MATCH_BLOBS) */
         u32 x = pLevel->iLikeRepCntr;
         if( x>0 ){
           skipLikeAddr = sqlite3VdbeAddOp1(v, (x&1)?OP_IfNot:OP_If,(int)(x>>1));
           VdbeCoverageIf(v, (x&1)==1);
           VdbeCoverageIf(v, (x&1)==0);
         }
-#endif
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) || defined(SQLITE_LIKE_DOESNT_MATCH_BLOBS) */
       }
 #ifdef WHERETRACE_ENABLED /* 0xffff */
       if( sqlite3WhereTrace ){
