--- sqlite-3.28.0/src/window.c	2026-01-09 16:11:01
+++ src/window.c	2025-09-24 15:21:58
@@ -972,7 +972,11 @@
     p->pSrc = sqlite3SrcListAppend(pParse, 0, 0, 0);
     if( p->pSrc ){
       p->pSrc->a[0].pSelect = pSub;
+#if defined(SQLITE_BUILDING_FOR_COMDB2)
+      sqlite3SrcListAssignCursors(pParse, p->pSrc, p->recording);
+#else /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       sqlite3SrcListAssignCursors(pParse, p->pSrc);
+#endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
       if( sqlite3ExpandSubquery(pParse, &p->pSrc->a[0]) ){
         rc = SQLITE_NOMEM;
       }else{
@@ -999,6 +1003,10 @@
 */
 void sqlite3WindowDelete(sqlite3 *db, Window *p){
   if( p ){
+    if( p->ppThis ){
+      *p->ppThis = p->pNextWin;
+      if( p->pNextWin ) p->pNextWin->ppThis = p->ppThis;
+    }
     sqlite3ExprDelete(db, p->pFilter);
     sqlite3ExprListDelete(db, p->pPartition);
     sqlite3ExprListDelete(db, p->pOrderBy);
@@ -1030,7 +1038,9 @@
 */
 static Expr *sqlite3WindowOffsetExpr(Parse *pParse, Expr *pExpr){
   if( 0==sqlite3ExprIsConstant(pExpr) ){
+#if !defined(SQLITE_BUILDING_FOR_COMDB2)
     if( IN_RENAME_OBJECT ) sqlite3RenameExprUnmap(pParse, pExpr);
+#endif /* !defined(SQLITE_BUILDING_FOR_COMDB2) */
     sqlite3ExprDelete(pParse->db, pExpr);
     pExpr = sqlite3ExprAlloc(pParse->db, TK_NULL, 0, 0);
   }
